
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://localhost:8000/Modeling%20Dense%20Multimodal%20Interactions%20Between%20Biological%20Pathways%20and%20Histology%20for%20Survival%20Prediction/">
      
      
        <link rel="prev" href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./">
      
      
        <link rel="next" href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/">
      
      
      <link rel="icon" href="../img/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction - One Last Kiss</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lxgw-wenkai-webfont@1.1.0/style.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-M0R7V8QGQ7"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-M0R7V8QGQ7",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-M0R7V8QGQ7",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#_1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="One Last Kiss" class="md-header__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            One Last Kiss
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
    
  
  🏠 Home

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../Mygo/" class="md-tabs__link">
          
  
    
  
  ⛏️ CV and Multimodels

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Session%201/" class="md-tabs__link">
          
  
    
  
  🚀 Re:0 Road to CPath & Visualization

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-tabs__link">
          
  
    
  
  💤 Re:0 Road to scRNA-seq Analysis

        </a>
      </li>
    
  

    
  

    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-tabs__link">
          
  
    
  
  🧬 scRNA Analysis

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-tabs__link">
          
  
    
  
  📐 Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-tabs__link">
          
  
    
  
  💡 Reviews

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="One Last Kiss" class="md-nav__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    One Last Kiss
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href=".." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    🏠 Home
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            🏠 Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    ⛏️ CV and Multimodels
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ⛏️ CV and Multimodels
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mygo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MyGO: Discrete Modality Information as Fine-Grained Tokens for Multi-modal Knowledge Graph Completion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GMM/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GMM
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../DINO/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    DINO
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mamba/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mamba
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ViT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ViT
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCoV2%2C3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo V2
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%80%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (一)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%20%28%E4%BA%8C%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (二)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%89%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (三)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E5%9B%9B%29%20%28TNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (四)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E7%9F%A5%E8%AF%86%E8%92%B8%E9%A6%8F%20%E9%A2%84%E8%AE%AD%E7%BB%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (知识蒸馏 预训练）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28PyramidTNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer(Pyramid TNT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%82%85%E9%87%8C%E5%8F%B6%E7%BA%A7%E6%95%B0%E5%92%8C%E5%82%85%E9%87%8C%E5%8F%B6%E5%8F%98%E6%8D%A2%E4%B8%93%E9%A2%98/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    傅里叶级数和傅里叶变换专题
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Python%20%E5%AD%98%E5%82%A8%E4%B8%8E%E8%AF%BB%E5%8F%96%20HDF5%20%E6%96%87%E4%BB%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Python 存储与读取 HDF5 文件
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Segment%20Anything%20in%20Medical%20Images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Segment Anything in Medical Images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Mamba%20Efficient%20Visual%20Representation%20Learning%20with%20Bidirectional%20State%20Space%20Model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Mamba Efficient Visual Representation Learning with Bidirectional State Space Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_18" checked>
        
          
          <label class="md-nav__link" for="__nav_2_18" id="__nav_2_18_label" tabindex="">
            
  
  <span class="md-ellipsis">
    🚑 医学影像相关
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_18_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_18">
            <span class="md-nav__icon md-icon"></span>
            🚑 医学影像相关
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Why%20is%20the%20winner%20the%20best/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Why is the winner the best
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PORPOISE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PORPOISE
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HIPT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Scaling Vision Transformers to Gigapixel Images via Hierarchical Self-Supervised Learning(HIPT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Data-efficient%20and%20weakly%20supervised%20computational%20pathology%20on%20whole-slide%20images%28CLAM%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data efficient and weakly supervised computational pathology on whole slide images(CLAM)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Towards%20a%20general-purpose%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Towards a general purpose foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20whole-slide%20foundation%20model%20for%20digital%20pathology%20from%20real-world%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A whole slide foundation model for digital pathology from real world data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20visual-language%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A visual language foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Recurrent memory with optimal polynomial projections.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      摘要
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-introduction" class="md-nav__link">
    <span class="md-ellipsis">
      1 Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      相关工作
    </span>
  </a>
  
    <nav class="md-nav" aria-label="相关工作">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 组织学全幅切片图像的生存分析
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-transformer" class="md-nav__link">
    <span class="md-ellipsis">
      2.2 多模态 Transformer 与可解释性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23" class="md-nav__link">
    <span class="md-ellipsis">
      2.3 多模态生存分析
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3" class="md-nav__link">
    <span class="md-ellipsis">
      3 方法
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3 方法">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31" class="md-nav__link">
    <span class="md-ellipsis">
      3.1 从转录组学到途径标记器
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3_1" class="md-nav__link">
    <span class="md-ellipsis">
      3 方法
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-patches" class="md-nav__link">
    <span class="md-ellipsis">
      3.2 从全幅切片图像到组织学 patches 标记器
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33" class="md-nav__link">
    <span class="md-ellipsis">
      3.3 多模态融合
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34" class="md-nav__link">
    <span class="md-ellipsis">
      3.4 生存预测
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#35" class="md-nav__link">
    <span class="md-ellipsis">
      3.5 多层次可解释性
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4" class="md-nav__link">
    <span class="md-ellipsis">
      4 实验
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4 实验">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41" class="md-nav__link">
    <span class="md-ellipsis">
      4.1 数据集与实施
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-baseline" class="md-nav__link">
    <span class="md-ellipsis">
      4.2 baseline 比较
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#43" class="md-nav__link">
    <span class="md-ellipsis">
      4.3 生存预测结果
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#44" class="md-nav__link">
    <span class="md-ellipsis">
      4.4 消融研究
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#45" class="md-nav__link">
    <span class="md-ellipsis">
      4.5 可解释性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      结论
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      补充材料
    </span>
  </a>
  
    <nav class="md-nav" aria-label="补充材料">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    <span class="md-ellipsis">
      1. 生存预测
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2" class="md-nav__link">
    <span class="md-ellipsis">
      2. 实施细节
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. 实施细节">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21_1" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 模型训练
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22" class="md-nav__link">
    <span class="md-ellipsis">
      2.2 评估指标
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3_2" class="md-nav__link">
    <span class="md-ellipsis">
      3. 额外的可解释性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4_1" class="md-nav__link">
    <span class="md-ellipsis">
      4. 额外结果
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. 额外结果">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#10" class="md-nav__link">
    <span class="md-ellipsis">
      10 倍放大下的结果
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaplan-meier" class="md-nav__link">
    <span class="md-ellipsis">
      Kaplan-Meier 分析
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      与临床协变量的比较
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      模态归因
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Prediction%20of%20recurrence%20risk%20in%20endometrial%20cancer%20with%20multimodal%20deep%20learning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction of recurrence risk in endometrial cancer with multimodal deep learning
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚀 Re:0 Road to CPath & Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            🚀 Re:0 Road to CPath & Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%201/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    理解全幅切片图像
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%202/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    在计算病理学中用于癌症诊断的弱监督深度学习
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%203/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    基于注意力的多示例学习可解释性
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../How%20to%20visualize%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to visualize Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AttentionViz%20A%20Global%20View%20of%20Transformer%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AttentionViz A Global View of Transformer Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Transformer%20Interpretability%20Beyond%20Attention%20Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transformer Interpretability Beyond Attention Visualization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualization Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Attention%20Generation%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Attention Generation Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🪛 PyTorch
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            🪛 PyTorch
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Autograd%20mechanics%20%E2%80%94%20PyTorch%202_3%20documentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Autograd mechanics — PyTorch 2 3 documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyTorch Autograd Explained   In depth Tutorial
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💤 Re:0 Road to scRNA-seq Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            💤 Re:0 Road to scRNA-seq Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏳ CHAPTER 1 简介
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ⏳ CHAPTER 1 简介
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1.The building block of life
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20Raw%20Data%20Processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 Raw data processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Analysis%20frameworks%20and%20tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Analysis frameworks and tools
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🌏 CHAPTER 2 数据准备与可视化
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            🌏 CHAPTER 2 数据准备与可视化
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20normalization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 Normalization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3_feature_selection/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 Feature selection
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.4_dimensionality_reduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.4 Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_3" >
        
          
          <label class="md-nav__link" for="__nav_4_1_3" id="__nav_4_1_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🍀 CHAPTER 3 识别细胞结构
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_3">
            <span class="md-nav__icon md-icon"></span>
            🍀 CHAPTER 3 识别细胞结构
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 Data integration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_4" >
        
          
          <label class="md-nav__link" for="__nav_4_1_4" id="__nav_4_1_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏱️ CHAPTER 4 推断细胞轨迹
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_4">
            <span class="md-nav__icon md-icon"></span>
            ⏱️ CHAPTER 4 推断细胞轨迹
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotemporal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 Pseudotemporal ordering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../rna_velocity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 RNA velocity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../lineage_tracing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 Lineage tracing
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_5" >
        
          
          <label class="md-nav__link" for="__nav_4_1_5" id="__nav_4_1_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🎁 CHAPTER 5 应对特殊情况
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_5">
            <span class="md-nav__icon md-icon"></span>
            🎁 CHAPTER 5 应对特殊情况
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../differential_gene_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 Differential gene expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compositional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 Compositional analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gsea_pathway/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 Gene set enrichment and pathway analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../perturbation_modeling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 Perturbation modeling
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_6" >
        
          
          <label class="md-nav__link" for="__nav_4_1_6" id="__nav_4_1_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⚙️ CHAPTER 6 机制建模
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_6">
            <span class="md-nav__icon md-icon"></span>
            ⚙️ CHAPTER 6 机制建模
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_regulatory_networks/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 Gene regulatory networks
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cell_cell_communication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 Cell-cell communication
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_7" >
        
          
          <label class="md-nav__link" for="__nav_4_1_7" id="__nav_4_1_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🔓 CHAPTER 7 反卷积
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_7">
            <span class="md-nav__icon md-icon"></span>
            🔓 CHAPTER 7 反卷积
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bulk_deconvolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 Bulk deconvolution
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    简明易懂的单细胞分析流程（使用 R 语言）
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            简明易懂的单细胞分析流程（使用 R 语言）
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%89%8D%E8%A8%80/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    前言
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    环境配置
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20%E6%9E%84%E5%BB%BA%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1 构建单细胞 RNA seq 分析环境
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20R%20%E7%8E%AF%E5%A2%83%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 R 环境安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Docker%20%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Docker 安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.4%20%E5%9C%A8%20Docker%20%E4%B8%AD%E5%90%AF%E5%8A%A8%20Rstudio/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.4 在 Docker 中启动 Rstudio
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.5%20%E6%9C%89%E5%85%B3%20Bioconductor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.5 有关 Bioconductor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20R%20%E7%9A%84%E5%9F%BA%E6%9C%AC%E7%94%A8%E6%B3%95/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 R 的基本用法
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20%E5%A6%82%E4%BD%95%E5%A4%84%E7%90%86%20DataFrame%EF%BC%9F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 如何处理 DataFrame？
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3%20%E4%BD%BF%E7%94%A8%20ggplot2%20%E5%8F%AF%E8%A7%86%E5%8C%96%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 使用 ggplot2 可视化数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1%20PCA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 PCA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2%20%E4%BD%BF%E7%94%A8%20clusterPlofiler%20%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 使用 clusterPlofiler 进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E5%9F%BA%E5%9B%A0%E6%9C%AC%E4%BD%93%EF%BC%88GO%EF%BC%89%E4%B8%AD%E5%90%84%20BP%E3%80%81MF%20%E5%92%8C%20CC%20%E7%9A%84%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 基因本体（GO）中各 BP、MF 和 CC 的富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E4%BD%BF%E7%94%A8%20DEseq2%20%E6%8F%90%E5%8F%96%20RNA-seq%20%E5%88%86%E6%9E%90%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%20DEG/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 使用 DEseq2 提取 RNA seq 分析数据中的 DEG
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.4%20%E4%BD%BF%E7%94%A8%E5%85%B6%E4%BB%96%E6%95%B0%E6%8D%AE%E5%BA%93%EF%BC%88KEGG%E3%80%81Reactome%20%E7%AD%89%EF%BC%89%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.4 使用其他数据库（KEGG、Reactome 等）进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.1%20scRNA-seq%20%E6%A6%82%E8%A6%81/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 scRNA seq 概要
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.2%20%E5%87%86%E5%A4%87%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 准备使用 Seurat 进行分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.3%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 使用 Seurat 进行数据预处理
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.4%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E7%9A%84%E4%B8%80%E8%88%AC%E9%A1%BA%E5%BA%8F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.4 使用 Seurat 进行数据分析的一般顺序
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.1%20%E5%A6%82%E4%BD%95%E6%9F%A5%E6%89%BE%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 如何查找单细胞 RNA seq 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.2%20%E4%B8%BA%E6%AF%8F%E4%B8%AA%E7%B0%87%E8%87%AA%E5%8A%A8%E6%B7%BB%E5%8A%A0%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E6%A0%87%E7%AD%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 为每个簇自动添加细胞注释标签
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.3%20%E4%BD%BF%E7%94%A8%20FeaturePlot%20%E6%A3%80%E6%9F%A5%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 使用 FeaturePlot 检查基因表达
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 将单细胞 RNA seq 数据集相互结合
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.5%20%E7%BA%A0%E6%AD%A3%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94%EF%BC%8C%E6%AF%94%E8%BE%83%E5%AF%B9%E7%85%A7%E7%BB%84%E5%92%8C%E5%88%BA%E6%BF%80%E7%BB%84/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.5 纠正批次效应，比较对照组和刺激组
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.6 确定每个群组的标记基因
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.1%20%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E7%BB%9F%E8%AE%A1%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E6%B0%B4%E5%B9%B3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 如何使用 Cell Ranger 统计 scRNA seq 数据中的基因表达水平
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.2%20%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E5%A4%84%E7%90%86%20NCBI%20Sequence%20Read%20Archive%20%28SRA%29%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 使用 Cell Ranger 处理 NCBI Sequence Read Archive (SRA) 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.1%20%E5%85%B3%E4%BA%8E%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 关于 scRNA seq 数据的质量控制
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.2%20%E4%BD%BF%E7%94%A8%20DoubleFinder%20%E6%A3%80%E6%B5%8B%E5%8F%8C%E9%87%8D%E7%BB%86%E8%83%9E/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.2 使用 DoubleFinder 检测双重细胞
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.3%20%E7%94%A8%20CellBender%20%E5%8E%BB%E9%99%A4%E8%83%8C%E6%99%AF%20RNA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.3 用 CellBender 去除背景 RNA
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧬 scRNA Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            🧬 scRNA Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🐍 Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            🐍 Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Large scale foundation model on single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scGPT%20toward%20building%20a%20foundation%20model%20for%20single-cell%20multi-omics%20using%20generative%20AI/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scGPT toward building a foundation model for single cell multi omics using generative AI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scBERT%20as%20a%20large-scale%20pretrained%20deep%20language%20model%20for%20cell%20type%20annotation%20of%20single-cell%20RNA-seq%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scBERT as a large scale pretrained deep language model for cell type annotation of single cell RNA seq data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RNA%20velocity%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA velocity of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TFvelo_gene%20regulation%20inspired%20RNA%20velocity%20estimation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TFvelo gene regulation inspired RNA velocity estimation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Disentanglement%20of%20single-cell%20data%20with%20biolord/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Disentanglement of single cell data with biolord
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Spatial%20transition%20tensor%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Spatial transition tensor of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRank%20infers%20drug-responsive%20cell%20types%20from%20untreated%20scRNA-seq%20data%20using%20a%20target-perturbed%20gene%20regulatory%20network/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRank infers drug responsive cell types from untreated scRNA seq data using a target perturbed gene regulatory network
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📔 文献
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            📔 文献
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ZNF689%20deficiency%20promotes%20intratumor%20heterogeneity%20and%20immunotherapy%20resistance%20in%20triple-negative%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ZNF689 deficiency promotes intratumor heterogeneity and immunotherapy resistance in triple negative breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Tumour%20vasculature%20at%20single-cell%20resolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tumour vasculature at single-cell resolution
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRNA-seq%20of%20gastric%20tumor%20shows%20complex%20intercellular%20interaction%20with%20an%20alternative%20T%20cell%20exhaustion%20trajectory%EF%BC%88%E5%8F%82%E8%80%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRNA seq of gastric tumor shows complex intercellular interaction with an alternative T cell exhaustion trajectory（参考）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20reveals%20new%20evolutionary%20complexity%20in%20uveal%20melanoma%EF%BC%88%E5%8F%82%E8%80%83%202%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis reveals new evolutionary complexity in uveal melanoma（参考 2）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20characterization%20of%20macrophages%20in%20uveal%20melanoma%20uncovers%20transcriptionally%20heterogeneous%20subsets%20conferring%20poor%20prognosis%20and%20aggressive%20behavior%EF%BC%88%E8%8C%83%E4%BE%8B%201%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell characterization of macrophages in uveal melanoma uncovers transcriptionally heterogeneous subsets conferring poor prognosis and aggressive behavior（范例 1）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ScRNA-seq%20of%20gastric%20cancer%20tissues%20reveals%20differences%20in%20the%20immune%20microenvironment%20of%20primary%20tumors%20and%20metastases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ScRNA seq of gastric cancer tissues reveals differences in the immune microenvironment of primary tumors and metastases
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Blimp-1%20and%20c-Maf%20regulate%20immune%20gene%20networks%20to%20protect%20against%20distinct%20pathways%20of%20pathobiont-induced%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Blimp 1 and c Maf regulate immune gene networks to protect against distinct pathways of pathobiont induced colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single%20cell%20analysis%20unveils%20B%20cell-dominated%20immune%20subtypes%20in%20HNSCC%20for%20enhanced%20prognostic%20and%20therapeutic%20stratification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis unveils B cell dominated immune subtypes in HNSCC for enhanced prognostic and therapeutic stratification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20spatial%20multi-omics%20and%20deep%20learning%20dissect%20enhancer-driven%20gene%20regulatory%20networks%20in%20liver%20zonation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell spatial multi omics and deep learning dissect enhancer driven gene regulatory networks in liver zonation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Neutrophil%20profiling%20illuminates%20anti-tumor%20antigen%20presenting%20potency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Neutrophil profiling illuminates anti tumor antigen presenting potency
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20blueprint%20for%20tumor-infiltrating%20B%20cells%20across%20human%20cancers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A blueprint for tumor infiltrating B cells across human cancers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20transcriptomic%20analyses%20reveal%20distinct%20immune%20cell%20contributions%20to%20epithelial%20barrier%20dysfunction%20in%20checkpoint%20inhibitor%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell transcriptomic analyses reveal distinct immune cell contributions to epithelial barrier dysfunction in checkpoint inhibitor colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Systematic%20dissection%20of%20tumor-normal%20single-cell%20ecosystems%20across%20a%20thousand%20tumors%20of%2030%20cancer%20types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Systematic dissection of tumor normal single cell ecosystems across a thousand tumors of 30 cancer types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../The%20aged%20tumor%20microenvironment%20limits%20T%20cell%20control%20of%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The aged tumor microenvironment limits T cell control of cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TIM-3%2B%20CD8%20T%20cells%20with%20a%20terminally%20exhausted%20phenotype%20retain%20functional%20capacity%20in%20hematological%20malignancies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TIM 3+ CD8 T cells with a terminally exhausted phenotype retain functional capacity in hematological malignancies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20lactate-SREBP2%20signaling%20axis%20drives%20tolerogenic%20dendritic%20cell%20maturation%20and%20promotes%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A lactate SREBP2 signaling axis drives tolerogenic dendritic cell maturation and promotes cancer progression
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20human%20omentum-specific%20mesothelial-like%20stromal%20population%20inhibits%20adipogenesis%20through%20IGFBP2%20secretion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A human omentum specific mesothelial like stromal population inhibits adipogenesis through IGFBP2 secretion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Interferon-stimulated%20neutrophils%20as%20a%20predictor%20of%20immunotherapy%20response/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interferon stimulated neutrophils as a predictor of immunotherapy response
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Repression%20of%20latent%20NF-%CE%BAB%20enhancers%20by%20PDX1%20regulates%20%CE%B2%20cell%20functional%20heterogeneity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Repression of latent NF κB enhancers by PDX1 regulates β cell functional heterogeneity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20systematic%20pan-cancer%20study%20on%20deep%20learning-based%20prediction%20of%20multi-omic%20biomarkers%20from%20routine%20pathology%20images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A systematic pan cancer study on deep learning based prediction of multi omic biomarkers from routine pathology images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20and%20spatial%20transcriptomics%20analysis%20of%20non-small%20cell%20lung%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell and spatial transcriptomics analysis of non small cell lung cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Seeing%20data%20as%20t-SNE%20and%20UMAP%20do/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Seeing data as t SNE and UMAP do
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mapping%20the%20cellular%20biogeography%20of%20human%20bone%20marrow%20niches%20using%20single-cell%20transcriptomics%20and%20proteomic%20imaging/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mapping the cellular biogeography of human bone marrow niches using single cell transcriptomics and proteomic imaging
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Onco-fetal%20Reprogramming%20of%20Endothelial%20Cells%20Drives%20Immunosuppressive%20Macrophages%20in%20Hepatocellular%20Carcinoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Onco fetal Reprogramming of Endothelial Cells Drives Immunosuppressive Macrophages in Hepatocellular Carcinoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20of%20anti-BCMA%20CAR%20T%20cell%20therapy%20in%20patients%20with%20central%20nervous%20system%20autoimmunity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis of anti BCMA CAR T cell therapy in patients with central nervous system autoimmunity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20sequencing%20depicts%20tumor%20architecture%20and%20empowers%20clinical%20decision%20in%20metastatic%20conjunctival%20melanoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell sequencing depicts tumor architecture and empowers clinical decision in metastatic conjunctival melanoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PR-SET7%20epigenetically%20restrains%20uterine%20interferon%20response%20and%20cell%20death%20governing%20proper%20postnatal%20stromal%20development/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PR SET7 epigenetically restrains uterine interferon response and cell death governing proper postnatal stromal development
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Clonal%20associations%20between%20lymphocyte%20subsets%20and%20functional%20states%20in%20rheumatoid%20arthritis%20synovium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clonal associations between lymphocyte subsets and functional states in rheumatoid arthritis synovium
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📐 Data Structures
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            📐 Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 1 章 python 入门
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%202%20%E7%AB%A0%20%E9%9D%A2%E5%90%91%E5%AF%B9%E8%B1%A1%E7%BC%96%E7%A8%8B/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 2 章 面向对象编程
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%203%20%E7%AB%A0%20%E7%AE%97%E6%B3%95%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 3 章 算法分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%204%20%E7%AB%A0%20%E9%80%92%E5%BD%92/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 4 章 递归
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%205%20%E7%AB%A0%20%E5%9F%BA%E4%BA%8E%E6%95%B0%E7%BB%84%E7%9A%84%E5%BA%8F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 5 章 基于数组的序列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%206%20%E7%AB%A0%20%E6%A0%88%E3%80%81%E9%98%9F%E5%88%97%E5%92%8C%E5%8F%8C%E7%AB%AF%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 6 章 栈、队列和双端队列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%207%20%E7%AB%A0%20%E9%93%BE%E8%A1%A8%28Linked%20Lists%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 7 章 链表(Linked Lists)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%208%20%E7%AB%A0%20%E6%A0%91%20%E2%98%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 8 章 树 ★
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%209%20%E7%AB%A0%20%E4%BC%98%E5%85%88%E7%BA%A7%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 9 章 优先级队列
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💡 Reviews
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            💡 Reviews
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_1" >
        
          
          <label class="md-nav__link" for="__nav_7_1" id="__nav_7_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧭 scRNA analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_1">
            <span class="md-nav__icon md-icon"></span>
            🧭 scRNA analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Best practices for single cell analysis across modalities
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Multi-omics%20integration%20in%20biomedical%20research%20%E2%80%93%20A%20metabolomics-centric%20review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multi omics integration in biomedical research – A metabolomics centric review
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Museum%20of%20spatial%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Museum of spatial transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20guide%20to%20artificial%20intelligence%20for%20cancer%20researchers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A guide to artificial intelligence for cancer researchers
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_2" >
        
          
          <label class="md-nav__link" for="__nav_7_2" id="__nav_7_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🦠 Metabolism
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_2">
            <span class="md-nav__icon md-icon"></span>
            🦠 Metabolism
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20analysis%20as%20a%20driver%20for%20discovery%2C%20diagnosis%2C%20and%20therapy/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic analysis as a driver for discovery, diagnosis, and therapy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20heterogeneity%20in%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic heterogeneity in cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Toward%20modeling%20metabolic%20state%20from%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Toward modeling metabolic state from single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20reprogramming%20and%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic reprogramming and cancer progression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_3" >
        
          
          <label class="md-nav__link" for="__nav_7_3" id="__nav_7_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Protein
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_3">
            <span class="md-nav__icon md-icon"></span>
            Protein
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Proteomic%20and%20interactomic%20insights%20into%20the%20molecular%20basis%20of%20cell%20functional%20diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Proteomic and interactomic insights into the molecular basis of cell functional diversity
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      摘要
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-introduction" class="md-nav__link">
    <span class="md-ellipsis">
      1 Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      相关工作
    </span>
  </a>
  
    <nav class="md-nav" aria-label="相关工作">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 组织学全幅切片图像的生存分析
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-transformer" class="md-nav__link">
    <span class="md-ellipsis">
      2.2 多模态 Transformer 与可解释性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23" class="md-nav__link">
    <span class="md-ellipsis">
      2.3 多模态生存分析
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3" class="md-nav__link">
    <span class="md-ellipsis">
      3 方法
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3 方法">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31" class="md-nav__link">
    <span class="md-ellipsis">
      3.1 从转录组学到途径标记器
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3_1" class="md-nav__link">
    <span class="md-ellipsis">
      3 方法
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-patches" class="md-nav__link">
    <span class="md-ellipsis">
      3.2 从全幅切片图像到组织学 patches 标记器
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33" class="md-nav__link">
    <span class="md-ellipsis">
      3.3 多模态融合
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34" class="md-nav__link">
    <span class="md-ellipsis">
      3.4 生存预测
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#35" class="md-nav__link">
    <span class="md-ellipsis">
      3.5 多层次可解释性
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4" class="md-nav__link">
    <span class="md-ellipsis">
      4 实验
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4 实验">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41" class="md-nav__link">
    <span class="md-ellipsis">
      4.1 数据集与实施
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-baseline" class="md-nav__link">
    <span class="md-ellipsis">
      4.2 baseline 比较
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#43" class="md-nav__link">
    <span class="md-ellipsis">
      4.3 生存预测结果
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#44" class="md-nav__link">
    <span class="md-ellipsis">
      4.4 消融研究
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#45" class="md-nav__link">
    <span class="md-ellipsis">
      4.5 可解释性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      结论
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      补充材料
    </span>
  </a>
  
    <nav class="md-nav" aria-label="补充材料">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    <span class="md-ellipsis">
      1. 生存预测
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2" class="md-nav__link">
    <span class="md-ellipsis">
      2. 实施细节
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. 实施细节">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21_1" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 模型训练
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22" class="md-nav__link">
    <span class="md-ellipsis">
      2.2 评估指标
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3_2" class="md-nav__link">
    <span class="md-ellipsis">
      3. 额外的可解释性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4_1" class="md-nav__link">
    <span class="md-ellipsis">
      4. 额外结果
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. 额外结果">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#10" class="md-nav__link">
    <span class="md-ellipsis">
      10 倍放大下的结果
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaplan-meier" class="md-nav__link">
    <span class="md-ellipsis">
      Kaplan-Meier 分析
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      与临床协变量的比较
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      模态归因
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/aletolia/aletolia.github.io.git/edit/master/docs/Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/aletolia/aletolia.github.io.git/raw/master/docs/Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


  <h1>Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction</h1>

<h2 id="_1">摘要<a class="headerlink" href="#_1" title="Permanent link">&para;</a></h2>
<p>整合全幅切片图像（WSIs）和整体转录组学以预测患者生存情况，可以增进我们对患者预后的理解。然而，这一多模态任务因数据性质不同而尤为具有挑战性：<strong>全幅切片图像为肿瘤提供了一种非常高维的空间描述，而整体转录组学则描述了该肿瘤内的全局基因表达水平</strong>。在这一背景下，我们的工作旨在解决两个关键挑战：</p>
<ul>
<li>（1）我们如何以一种语义上有意义且可解释的方式对转录组学数据进行标记化处理？及</li>
<li>（2）我们如何捕捉这两种模态之间的密集多模态交互？</li>
</ul>
<p>在此，我们提出<strong>从转录组学中学习生物途径标记，这些标记能够编码特定的细胞功能</strong>。结合编码切片形态的组织学 patches 标记，我们认为这些标记构成了适合解释性分析的基本单元。<strong>我们使用一种内存高效的多模态 Transformer 融合这两种模态，该 Transformer 能够模拟生物途径与组织学 patches 标记之间的互动</strong>。我们的模型 SurvPath 在五个来自癌症基因组图谱的数据集上的评估中达到了最新的水平，并且与单模态和多模态 baseline 相比，显示出卓越的性能。我们的解释框架确定了关键的多模态预后因素，并因此可以提供有关基因型与表型相互作用的宝贵见解。代码可在以下网址获取：https://github.com/mahmoodlab/SurvPath。</p>
<h2 id="1-introduction">1 Introduction<a class="headerlink" href="#1-introduction" title="Permanent link">&para;</a></h2>
<p><img alt="" src="https://arxiv.org/html/2304.06819v2/x1.png" /></p>
<blockquote>
<p>图 1：SurvPath 的多模态可解释性。SurvPath 通过 Transformer 的交叉注意机制，实现了生物途径和形态模式之间多模态交互的可视化，这里以高风险乳腺癌为例。弦的厚度表示注意力权重。</p>
</blockquote>
<p>预测患者预后是计算病理学（CPATH）的一个基本任务，旨在利用组织学全幅切片图像（WSIs）进行自动化风险评估、患者分层和治疗反应预测。<strong>患者的预后通常被视为一项生存任务，其目标是学习能正确排序原始诊断全幅切片生存时间的风险估计</strong>。由于全幅切片图像可能达到 100,000 x 100,000 像素，因此常用弱监督方法如多示例学习（MIL）进行生存预测。在 MIL 中，全幅切片图像被标记化为小块，<strong>从中提取特征并输入到汇聚网络（如注意力网络）进行下游分类。</strong></p>
<p>虽然组织学提供了有关细胞类型及其在组织中的组织方式的表型信息，但其他模态可以提供可能独立与预后相关的互补信号。例如，整体转录组学代表组织中的平均基因表达，可以揭示更丰富的全局细胞类型和细胞状态，并已被证实是患者生存的强预测因子。通过结合这两种模态，我们可以将整体转录组学提供的全局信息与全幅切片图像的空间信息整合起来。虽然大多数现有方法采用后期融合机制（即，融合模态级表示），但我们设计了<strong>一种早期融合方法，可以明确模拟局部形态模式和转录组学之间的精细交叉模态关系</strong>。与广泛使用的视觉 - 语言模型相比，转录组学和组织学的多模态融合呈现出两个关键技术挑战：</p>
<ol>
<li>转录组学模态的标记化：基于图像和文本的模态可以明确地标记化为对象区域和词汇标记，然而，<strong>以语义上有意义且可解释的方式标记化转录组学是具有挑战性的。由于转录组学数据已自然呈现为特征向量，许多先前研究忽略了标记化，并直接将整个特征与其他模态连接</strong>，这限制了多模态学习仅至后期融合操作。另一种方法是，可以将基因划分为代表不同基因家族的粗糙功能集（例如，肿瘤抑制基因和致癌基因），并将其用作标记。然而，这种集合提供的细胞内互动的描述是粗糙且不完整的，因为一个基因家族可以涉及不同的细胞功能。因此，它们可能缺乏与精细形态的语义对应。相反，<strong>我们提议根据已建立的生物途径对基因进行标记。途径是具有已知相互作用的基因集，与特定的细胞功能相关，例如 TGF-β 信号级联</strong>，它有助于乳腺癌中的上皮 - 间充质转化。与粗糙集相比（例如，我们工作中的 <span class="arithmatex">\(N_{\mathcal{P}}=331\)</span>），<strong>基于途径的基因分组可以产生代表独特分子过程的数百到数千个标记，我们假设这些标记是与组织学多模态融合更合适的表示</strong>。此外，由于途径代表独特的细胞功能，它们构成了适合解释性的基本推理单元（见图 1）。</li>
<li>捕获密集多模态交互：利用 Transformer 通过自注意机制捕获所有标记之间的成对相似性，可以实现组织学和途径标记的早期融合。然而，模拟大量组织学 patches 标记（例如 <span class="arithmatex">\(\left.N_{\mathscr{H}}=15,000\right)\)</span> 和途径标记 <span class="arithmatex">\(\left(N_{\mathcal{P}}=331\right)\)</span> 之间的成对互动带来了可扩展性挑战。由于 Transformer 注意力的二次复杂性，模拟所有可能的互动需要大量的计算和内存资源。为解决这一问题，我们引入了一种<strong>新的统一的、内存高效的注意力机制，能够模拟 patches 到途径、途径到 patches 以及途径之间的互动</strong>。通过以下方式实现这三种互动的建模：（1）设计查询、键和值以跨标记类型共享参数，并（2）简化注意力层以忽略 patches 之间的互动，我们通过实验发现这对生存分析的效果不太有效。</li>
</ol>
<p>总结来说，我们的贡献包括：</p>
<ul>
<li>（1）一个转录组学标记器，利用细胞生物学现有知识生成生物途径标记；</li>
<li>（2）SURVPATH，一种内存高效且不依赖分辨率的多模态 Transformer 公式，集成转录组学和 patches 标记预测患者生存；</li>
<li>（3）一个多级可解释性框架，使得派生单模态和跨模态的预测洞见成为可能；</li>
<li>（4）一系列实验和消融研究显示了 SURVPATH 的预测能力，使用了来自癌症基因组图谱计划（TCGA）的五个数据集，并与单模态和多模态融合方法进行了基准测试。</li>
</ul>
<p><img alt="" src="https://arxiv.org/html/2304.06819v2/x2.png" /></p>
<blockquote>
<p>图 2：SurvPath 的框图。(1) 我们将转录组学数据标记化为语义上有意义、可解释且可端到端学习的生物途径标记。(2) 我们进一步使用 SSL 预训练的特征提取器将相应的组织学全幅切片图像标记化为 patches 标记。(3) 我们使用一种内存高效的多模态 Transformer 将途径标记和 patches 标记结合起来，用于生存结果预测。</p>
</blockquote>
<h2 id="_2">相关工作<a class="headerlink" href="#_2" title="Permanent link">&para;</a></h2>
<h3 id="21">2.1 组织学全幅切片图像的生存分析<a class="headerlink" href="#21" title="Permanent link">&para;</a></h3>
<p>最近，已提出了几种基于组织学的生存模型【92, 88, 82, 63, 66】。大部分贡献致力于使用多实例学习（MIL）来模拟肿瘤异质性和肿瘤微环境。为此，提出了几种 MIL 池化策略，如使用图神经网络来模拟局部 patches 交互【49, 16, 38】，计算 patches 嵌入之间的差异【60】，或采用多放大倍数 patches 表示【44】。</p>
<h3 id="22-transformer">2.2 多模态 Transformer 与可解释性<a class="headerlink" href="#22-transformer" title="Permanent link">&para;</a></h3>
<p>同时，Transformer 用于多模态融合在分类和生成任务中得到了显著关注【70, 86, 61】。多模态标记可以被串联并输入到常规 Transformer【72, 18】、层次 Transformer【43】或交叉注意 Transformer【55, 46, 52】。随着模态的数量和维度增加，典型的序列长度可能变得过大，无法输入到普通 Transformer 中，因此需要低复杂性方法。已有多种模型提出了自注意力的重新表述，以减少内存和计算需求【4, 85, 31, 12, 83, 29, 15, 14】，例如，通过使用低秩分解【47, 85】来近似自注意力，使用潜在瓶颈蒸馏【31, 53, 32】，优化 GPU 读/写【15, 14】或使用稀疏注意模式【4, 56】。最近，可解释的多模态模型或事后解释方法【41, 75, 1】也已成为研究的关键领域，尤其是在医疗保健和人机交互等敏感的人工智能协作决策场景中。</p>
<h3 id="23">2.3 多模态生存分析<a class="headerlink" href="#23" title="Permanent link">&para;</a></h3>
<p>多模态整合是癌症预后的一个重要目标【64】，因为结合组织学和如基因组学或转录组学的组学数据是许多癌症类型的当前临床实践。这些工作的大多数采用后期融合机制【71, 9】，并且在模态融合的操作方式上存在差异。融合可以基于向量串联【51】、模态级对齐【7】、双线性池化（即克罗内克积）【9, 80】或分解双线性池化【39, 57】。</p>
<p>与此不同，可以使用早期融合机制，其中模拟输入各成分之间的交叉模态互动【10, 91, 17, 87】。我们的工作是在 MCAT【10】的基础上进行的，MCAT 使用交叉注意模块来模拟组织学 patches（键、值）对基因集（查询）的注意力。然而，MCAT 存在几个局限性：<strong>（1）交叉注意是单向的，只模拟 patches 到基因的互动，（2）使用粗糙的集合进行转录组学标记化，这并不反映实际的分子过程，（3）集合之间的基因重叠显著，导致冗余的交叉注意热图。</strong></p>
<h2 id="3"><strong>3 方法</strong><a class="headerlink" href="#3" title="Permanent link">&para;</a></h2>
<p>在这里，我们介绍 SurvPath，这是我们提出的基于组织学和转录组学的多模态生存预测方法。第 3.1 节介绍用于构建生物途径标记的转录组学编码器，第 3.2 节介绍用于构建 patches 标记的组织学编码器，第 3.3 节介绍我们基于 Transformer 的多模态聚合，第 3.4 节介绍其在生存预测中的应用（见图 2）。最后，第 3.5 节介绍我们的多层次可解释性框架。</p>
<h3 id="31">3.1 从转录组学到途径标记器<a class="headerlink" href="#31" title="Permanent link">&para;</a></h3>
<p><strong>构成途径</strong>：选择适当的推理单元用于转录组学分析是具有挑战性的，这归因于细胞过程的复杂和层级性质。<strong>途径由一组基因或子途径组成，这些基因或子途径参与特定的生物过程，代表了这一分析的自然推理单元</strong>。可以将这与动作识别进行比较，在动作识别中，一个动作（即生物途径）可以通过一系列由传感器捕获的动作（即一组基因的转录组学测量）来描述。</p>
<h3 id="3_1">3 方法<a class="headerlink" href="#3_1" title="Permanent link">&para;</a></h3>
<p>在此，我们介绍 SurvPath，这是我们提出的基于组织学和转录组学的多模态生存预测方法。第 3.1 节展示了如何构建生物途径标记的转录组学编码器，第 3.2 节展示了如何构建 patches 标记的组织学编码器，第 3.3 节展示了我们基于 Transformer 的多模态聚合，第 3.4 节展示了其在生存预测中的应用（参见图 2）。最后，第 3.5 节介绍了我们的多层次可解释性框架。</p>
<p>给定一组 <span class="arithmatex">\(N_{\mathcal{G}}\)</span> 基因的转录组学测量值，记为 <span class="arithmatex">\(\mathbf{g} \in{\mathbb{R} }^{N_{\mathcal{G}}}\)</span>，以及每个途径的组成，我们的目标是构建途径级标记 <span class="arithmatex">\(\mathbf{X}^{(\mathcal{P})} \in{\mathbb{R}}^{N_{\mathcal{P}} \times d}\)</span>，其中 <span class="arithmatex">\(d\)</span> 表示标记的维度。转录组学可以视为表格数据，可以通过多层感知器（MLPs）有效编码。具体来说，我们正在学习特定途径的权重 <span class="arithmatex">\(\phi_i\)</span>，即 <span class="arithmatex">\(\mathbf{x}_i^{(\mathcal{P})}=\phi_i\left(\mathbf{g}_{\mathcal{P}_i}\right)\)</span>，其中 <span class="arithmatex">\(\mathbf{g}_{\mathcal{P}_i}\)</span> 是途径 <span class="arithmatex">\(\mathcal{P}_i\)</span> 中存在的基因集。这可以视为学习一个稀疏多层感知器（S-MLP），将转录组学 <span class="arithmatex">\(\mathbf{g} \in{\mathbb{R}}^{N_{\mathcal{G}}}\)</span> 映射到标记 <span class="arithmatex">\(\mathbf{x}^{(\mathcal{P})} \in{\mathbb{R}}^{N_{\mathcal{P}} d}\)</span>。网络的稀疏性由嵌入在 S-MLP 权重中的基因到途径的连接性控制。通过简单地将 <span class="arithmatex">\(\mathbf{x}^{(\mathcal{P})} \in{ \mathbb{R}}^{N_{\mathcal{P}} d}\)</span> 重塑为 <span class="arithmatex">\(\mathbf{X}^{(\mathcal{P})} \in{\mathbb{R} }^{N_{\mathcal{P}} \times d}\)</span>，我们定义了 Transformer 可以使用的途径标记。每个途径标记对应于构成其的基因级转录组学的深层表示，这既（1）可解释，因为它编码了特定的生物功能；也（2）可在端到端的预测任务中学习。</p>
<h3 id="32-patches">3.2 从全幅切片图像到组织学 patches 标记器<a class="headerlink" href="#32-patches" title="Permanent link">&para;</a></h3>
<p>给定输入的全幅切片图像（WSI），我们的目标是衍生出定义 patches 标记的低维 patches 级嵌入。我们首先识别组织区域，以确保被忽略的背景不携带生物意义。然后，我们将识别的组织区域分解为一组 <span class="arithmatex">\(N_{\mathscr{H}}\)</span> 非重叠 patches，放大倍数为 20 倍（或分辨率约为 0.5 微米/像素），记为 <span class="arithmatex">\(\mathbf{H}=\left\{\mathbf{h}_1, \ldots, \mathbf{h}_{N_{\mathfrak{H}}}\right\}\)</span>。由于每个 WSI 的 patches 数量可能非常大（例如，可能超过 50,000 个 patches 或 78GB 的浮点数），因此需要在模型训练前提取 patches 嵌入，以减少总体内存需求。正式地，我们使用预训练的特征提取器 <span class="arithmatex">\(f(\cdot)\)</span> 将每个 patches <span class="arithmatex">\(\mathbf{h}_i\)</span> 映射到 patches 嵌入 <span class="arithmatex">\(\mathbf{x}_i^{(\mathscr{H})}=f\left(\mathbf{h}_i\right)\)</span>。在本工作中，我们使用了一个通过对超过 1500 万个泛癌症病理学 patches 的对比学习预训练的 Swin Transformer 编码器。所得的 patches 嵌入代表了 patches 的压缩表达（压缩比为 256），我们进一步通过一个可学习的线性变换将其传递，以匹配标记维度 <span class="arithmatex">\(d\)</span>，从而生成 patches 标记 <span class="arithmatex">\(\mathbf{X}^{(\mathscr{H})} \in{\mathbb{R}}^{N_{\mathscr{H}} \times d}\)</span>。</p>
<h3 id="33">3.3 多模态融合<a class="headerlink" href="#33" title="Permanent link">&para;</a></h3>
<p>我们的目标是设计一种早期融合机制，以模拟途径和 patches 标记之间的密集多模态交互。我们使用 Transformer 注意力 [72] 来测量和聚合多模态标记之间的成对交互。具体来说，我们通过连接途径和 patches 标记来定义一个多模态序列，产生 <span class="arithmatex">\(\left(N_{\mathscr{H}}+N_{\mathcal{P}}\right)\)</span> 维度为 <span class="arithmatex">\(d\)</span> 的标记，记为 <span class="arithmatex">\(\mathbf{X} \in\left(N_{\mathcal{P}}+N_{\mathscr{H}}\right) \times d\)</span>。遵循自我注意术语 [72]，我们定义三个线性投影使用可学习矩阵来提取查询 <span class="arithmatex">\((\mathbf{Q})\)</span>、键 <span class="arithmatex">\((\mathbf{K})\)</span> 和值 <span class="arithmatex">\((\mathbf{V})\)</span> 及自注意力 <span class="arithmatex">\(\mathbf{A}\)</span>，设置 <span class="arithmatex">\(d=d_k=d_q=d_v\)</span>。然后定义 Transformer 注意力为：</p>
<div class="arithmatex">\[
\mathbf{X}_{\mathrm{Att}}=\sigma\left(\frac{\mathbf{Q K}^T}{\sqrt{d}}\right) \mathbf{V}=\left(\begin{array}{cc}
\mathbf{A}_{\mathcal{P}} \rightarrow \mathcal{P} &amp; \mathbf{A}_{\mathcal{P}} \rightarrow \mathscr{H} \\
\mathbf{A}_{\mathscr{H}} \rightarrow \mathcal{P} &amp; \mathbf{A}_{\mathscr{H} \rightarrow \mathscr{H}}
\end{array}\right)\left(\begin{array}{c}
\mathbf{V}_{\mathcal{P}} \\
\mathbf{V}_{\mathscr{H}}
\end{array}\right)\tag{1}
\]</div>
<p>其中，<span class="arithmatex">\(\sigma\)</span> 是按行的 softmax。项 <span class="arithmatex">\(\mathbf{QK}^T\)</span> 的内存需求为 <span class="arithmatex">\(\mathcal{O}\left(\left(N_{\mathcal{H}}+N_{\mathcal{P}}\right)^2\right)\)</span>，对于长序列来说，计算成本会变得很高。这构成了一个主要瓶颈，因为一个全扫描图像（WSI）可能有超过50,000个补丁，这使得在大多数硬件上进行这种计算变得具有挑战性。相反，我们提议将多模态 Transformer 注意力分解为四个内部和跨模态项：（1）内模态路径自注意力编码路径到路径的相互作用 <span class="arithmatex">\(\mathbf{A}_{\mathcal{P} \rightarrow \mathcal{P}} \in \mathbb{R}^{N_{\mathcal{P}} \times N_{\mathcal{P}}}\)</span>；（2）跨模态路径引导的交叉注意力编码路径到补丁的相互作用 <span class="arithmatex">\(\mathbf{A}_{\mathcal{P} \rightarrow \mathcal{H}} \in \mathbb{R}^{N_{\mathcal{P}} \times N_{\mathcal{H}}}\)</span>；（3）跨模态组织学引导的交叉注意力编码补丁到路径的相互作用 <span class="arithmatex">\(\mathbf{A}_{\mathcal{H} \rightarrow \mathcal{P}} \in \mathbb{R}^{N_{\mathcal{H}} \times N_{\mathcal{P}}}\)</span>；以及（4）内模态完全组织学自注意力编码补丁到补丁的相互作用 <span class="arithmatex">\(\mathbf{A}_{\mathcal{H} \rightarrow \mathcal{H}} \in \mathbb{R}^{N_{\mathcal{H}} \times N_{\mathcal{H}}}\)</span>。</p>
<p>由于 patches token 的数量远大于路径的数量，即 <span class="arithmatex">\(N_{\mathcal{H}} \gg N_{\mathcal{P}}\)</span>，大部分的内存需求来自于计算和存储 <span class="arithmatex">\(\mathbf{A}_{\mathcal{H} \rightarrow \mathcal{H}}\)</span>。为了解决这个瓶颈，我们将 Transformer 注意力近似为：</p>
<div class="arithmatex">\[
\hat{\mathbf{X}}_{\mathrm{Att}}=\left(\begin{array}{c}
\mathbf{x}_{\mathrm{Att}}^{(\mathcal{P})} \\
\hat{\mathbf{X}}_{\mathrm{Att}}^{(\mathscr{H})}
\end{array}\right)=\sigma\left[\frac{1}{\sqrt{d}}\left(\begin{array}{cc}
\mathbf{Q}_{\mathcal{P}} \mathbf{K}_{\mathcal{P}}^T &amp; \mathbf{Q}_{\mathcal{P}} \mathbf{K}_{\mathscr{H}}^T \\
\mathbf{Q}_{\mathscr{H}} \mathbf{K}_{\mathcal{P}}^T &amp; -\infty
\end{array}\right)\right] \mathbf{V}\tag{2}
\]</div>
<p>其中 <span class="arithmatex">\(\mathbf{Q}_{\mathcal{P}}\)</span>（分别为 <span class="arithmatex">\(\mathbf{K}_{\mathcal{P}}\)</span>）和 <span class="arithmatex">\(\mathbf{Q}_{\mathscr{H}}\)</span>（分别为 <span class="arithmatex">\(\mathbf{K}_{\mathscr{H}}\)</span>）表示途径和组织学查询和键的子集。将 softmax 前的 patches 到 patches 交互设置为 <span class="arithmatex">\(-\infty\)</span> 相当于忽略这些交互。展开方程 10，我们得到 <span class="arithmatex">\(\mathbf{X}_{\text {Att}}^{(\mathcal{P})}=\sigma\left(\frac{\mathbf{Q}_{\mathcal{P}} \mathbf{K}^T}{\sqrt{d}}\right) \mathbf{V}_{\mathcal{P}}\)</span>，和 <span class="arithmatex">\(\hat{\mathbf{X}}_{\text {Att}}^{(\mathscr{H})}=\sigma\left(\frac{\mathbf{Q}_{\mathscr{H}} \mathbf{K}_{\mathcal{P}}^T}{\sqrt{d}}\right) \mathbf{V}_{\mathscr{H}}\)</span>。交互的数量显著减少，使得计算 <span class="arithmatex">\(\hat{\mathbf{A}}\)</span> 的内存需求有限。这种表述可以视为多模态序列上的稀疏注意模式 [4]，其中稀疏性是在 patches 标记之间强加的。这种表述在参数效率上也很高，因为为编码两种模态学习了一套独特的键、查询和值。此外，这种表述类似于图神经网络，在这个网络上，途径相互连接，并且每个途径都与所有 patches 连接。</p>
<p>通过将 <span class="arithmatex">\(\hat{\mathbf{x}}_{\text {Att}}\)</span> 通过一个带有层归一化的前馈层后，我们取得了注意力后途径和 patches 标记的平均表征，分别记为 <span class="arithmatex">\(\overline{\mathbf{x}}_{\mathrm{Att}}^{\mathcal{P}}\)</span> 和 <span class="arithmatex">\(\overline{\mathbf{x}}_{\mathrm{Att}}^{\mathfrak{H}}\)</span>。最终的表征 <span class="arithmatex">\(\overline{\mathbf{x}}_{\mathrm{Att}}\)</span>，然后通过连接 <span class="arithmatex">\(\overline{\mathbf{x}}_{\mathrm{Att}}^{\mathcal{P}}\)</span> 和 <span class="arithmatex">\(\overline{\mathbf{x}}_{\mathrm{Att}}^{\mathcal{H}}\)</span> 来定义。</p>
<h3 id="34">3.4 生存预测<a class="headerlink" href="#34" title="Permanent link">&para;</a></h3>
<p>使用多模态嵌入 <span class="arithmatex">\(\overline{\mathbf{x}}_{\text {Att }} \in{\R}^{2d}\)</span>，我们的监督目标是预测患者的生存期。根据先前的工作【90】，我们通过以下方式定义患者的生存状态：(1) 审查状态 <span class="arithmatex">\(c\)</span>，其中 <span class="arithmatex">\(c=0\)</span> 表示观察到的患者死亡，<span class="arithmatex">\(c=1\)</span> 对应于患者的最后已知随访；(2) 事件发生时间 <span class="arithmatex">\(t_i\)</span>，如果 <span class="arithmatex">\(c=0\)</span>，则对应于患者诊断和观察到的死亡之间的时间，如果 <span class="arithmatex">\(c=1\)</span>，则对应于最后一次随访。我们不直接预测事件的观测时间 <span class="arithmatex">\(t\)</span>，而是通过定义基于生存时间值四分位数的非重叠时间间隔 <span class="arithmatex">\(\left(t_{j-1}, t_j\right), j \in[1, \ldots, n]\)</span> 并表示为 <span class="arithmatex">\(y_j\)</span> 来近似它。问题简化为分类，每个患者现在由 <span class="arithmatex">\(\left(\overline{\mathbf{x}}_{\text {Att }}, y_j, c\right)\)</span> 定义。我们定义的分类器使得每个输出逻辑（经 sigmoid 激活后）<span class="arithmatex">\(\sigma\left(\hat{y}_j\right)\)</span> 表示患者在时间间隔 <span class="arithmatex">\(\left(t_{j-1}, t_j\right)\)</span> 内死亡的概率。我们进一步采取逻辑的累积乘积 <span class="arithmatex">\(\prod_{k=1}^j\left(1-\sigma\left(\hat{y}_k\right)\right)\)</span> 来表示患者在时间间隔 <span class="arithmatex">\(\left(t_{j-1}, t_j\right)\)</span> 内生存的概率。最后，通过取所有逻辑之和的负值，我们可以定义用于训练网络的患者级风险。更多信息请参见补充材料。</p>
<h3 id="35">3.5 多层次可解释性<a class="headerlink" href="#35" title="Permanent link">&para;</a></h3>
<p>我们提出了一个在多个层次上运作的可解释性框架，以推导转录组学、组织学和跨模态可解释性（见补充材料）。</p>
<p><strong>转录组学：</strong> 我们使用积分梯度（IG）【68】来识别途径和基因的影响，得出一个描述每个途径或基因对预测风险贡献程度的分数。负 IG 分数对应于与较低风险相关的途径/基因，而正 IG 分数表明与较高风险相关。非常小的分数表示影响可以忽略。这种可解释性分析有两个目的：</p>
<ul>
<li>（1）验证已知与预后相关的基因和途径；</li>
<li>（2）识别可能预测预后的新的基因和途径候选者。</li>
</ul>
<p><strong>组织学：</strong> 我们类似地使用 IG 来推导 patches 级影响，使得能够研究低风险和高风险相关 patches 的形态学。</p>
<p><strong>跨模态交互：</strong> 最后，我们可以使用学习到的 Transformer 注意力矩阵 <span class="arithmatex">\(\hat{\mathbf{A}}\)</span> 研究途径到 patches 和 patches 到途径的交互。具体地，我们定义 patches <span class="arithmatex">\(j\)</span>（分别是途径）相对于途径 <span class="arithmatex">\(i\)</span>（分别是 patches）的重要性为 <span class="arithmatex">\(\hat{\mathbf{A}}_{i j}\)</span>（分别是 <span class="arithmatex">\(\hat{\mathbf{A}}_{j i}\)</span>）。这使得可以构建相关途径和相应形态特征的热图。这种可解释性属性是我们框架独有的，使得能够研究由途径描述的特定细胞功能如何与组织学互动。</p>
<p><strong>表 1：</strong> 使用 c- 指数（在 20 倍放大下）衡量的 SURVPATH 和 baseline 在预测特定疾病患者生存方面的结果。最佳表现加粗，次佳下划线。Cat 表示串联，KP 表示克罗内克积。所有组学和多模态 baseline 都使用了 Reactome 和 Hallmark 途径集进行训练。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202404251414835.png" /></p>
<h2 id="4">4 实验<a class="headerlink" href="#4" title="Permanent link">&para;</a></h2>
<h3 id="41">4.1 数据集与实施<a class="headerlink" href="#41" title="Permanent link">&para;</a></h3>
<p>我们在来自 TCGA 的五个数据集上评估 SurvPath：膀胱尿路上皮癌（BLCA）（n=359）、乳腺浸润性癌（BRCA）（n=869）、胃腺癌（STAD）（n=317）、结直肠腺癌（COADREAD）（n=296）以及头颈部鳞状细胞癌（HNSC）（n=392）。先前的研究集中在预测总生存期（OS）【9】，然而，这种方法有过高估计癌症相关死亡比例的风险，因为患者可能因其他原因死亡。相反，我们预测疾病特异性生存期（DSS），作为更准确的表示患者疾病状态。</p>
<p>途径收集：我们使用 Xena 数据库【23】访问 TCGA 的原始转录组学数据及其 DSS 标签。我们从两个资源中提取途径：Reactome【22】和人类分子标志数据库（MSigDB）- Hallmarks【67, 42】。Reactome 和 MSigDB-Hallmarks 分别包含 1,281 和 50 个人类生物途径。我们进一步选择了至少有 90% 的转录组学数据可用的途径，结果得到 331 个途径，这些途径来源于 4,999 个不同的基因（281 个 Reactome 途径来自 1,577 个基因和 50 个 Hallmarks 途径来自 4,241 个基因）。</p>
<p>组织学收集：我们收集了所有用于初步诊断的诊断性全幅切片图像（WSIs），共得到 2,233 个 WSIs，平均每个 WSI 有 14,509 个 patches，假设使用
256×256 的 patches。总共，我们收集了超过 2.86TB 的原始图像数据，包含约 32.4 百万个 patches。</p>
<p>实施：我们使用 5 折交叉验证来训练所有模型。每个分割都根据样本位置进行分层，以减少潜在的批处理伪迹【28】。为了增加训练期间的变异性，我们从 WSI 中随机采样了 4,096 个 patches。在测试时，使用所有 patches 来得出最终预测（见补充材料）。</p>
<h3 id="42-baseline">4.2 baseline 比较<a class="headerlink" href="#42-baseline" title="Permanent link">&para;</a></h3>
<p>我们将 baseline 分为：(1) 单模态组织学方法，(2) 单模态转录组学方法，以及 (3) 多模态方法（进一步细分为早期融合与晚期融合方法）。</p>
<p><strong>组织学 baseline：</strong> 所有 baseline 使用与 SurvPath 相同的预训练特征提取器【76】。我们与 ABMIL【30】进行比较，后者使用门控注意力池化；AMISL【88】，首先使用 K-means 聚类 patches 嵌入再进行注意力处理；以及 TransMIL【62】，该方法使用 Nyström 方法【85】近似 patches 自注意力。</p>
<p><strong>转录组学 baseline：</strong> 所有 baseline 使用由 Reactome 和 Hallmarks 转录组学聚合定义的相同输入。(a) MLP【27】使用 4 层 MLP；(b) SNN【27, 9】在 MLP 基础上增加了额外的 alpha dropout 层；(c) S-MLP【25, 20】使用由 2 层稀疏途径感知 MLP 跟随的密集 2 层 MLP。这一 baseline 与我们的转录组学编码器有相似之处。</p>
<p><strong>多模态 baseline：</strong> (a) 晚期融合：我们结合了 ABMIL【30】、AMISL【88】和 TransMIL【62】与 S-MLP，并使用串联【51】表示为 ABMIL (Cat)、AMISL (Cat) 和 TransMIL (Cat)，以及使用 Kronecker 积【9, 21, 89, 81】表示为 ABMIL (KP)、AMISL (KP) 和 TransMIL (KP)。(b) 早期融合：MCAT【10】使用基因组引导的交叉注意力后跟模态特定的自注意力模块，而 MOTCat【87】使用最优运输（OT）匹配 patches 标记和基因组标记分布。</p>
<h3 id="43">4.3 生存预测结果<a class="headerlink" href="#43" title="Permanent link">&para;</a></h3>
<p>表 1 展示了在 20 倍放大下评估的 SurvPath 和 baseline 的结果（参见补充材料中 10 倍分析）。SurvPath 在 20 倍和 10 倍放大下均表现出最佳的整体性能，超越单模态和多模态 baseline。在 20 倍放大下，SurvPath 比 TransMIL 高出 +7.3%，比 MLP 高出 +3.0%，比 MCAT 高出 +3.5%。我们将 SurvPath 的高性能归因于：(1) 使用两种模态，(2) 统一、简单且参数高效的融合模型，(3) 语义意义明确的转录组学标记器。</p>
<p><strong>转录组学 vs. 组织学 vs. 多模态：</strong> 多模态 baseline 显著优于组织学 baseline。有趣的是，训练在我们的转录组学集上的简单 MLP 构成了一个强大的 baseline，其性能超过了几种多模态方法。这突出了执行稳健的特征选择和整合异质和高维数据模态的挑战。此外，相对较小的数据集大小进一步复杂化了复杂模型的学习和过拟合风险。与临床变量的比较在补充材料中提供。</p>
<p><strong>有上下文 vs. 无上下文：</strong> 尽管 TransMIL 使用 Nyström 注意力模拟路径到 patches 的交互，ABMIL 和 TransMIL 的表现相似。这一观察支持了我们忽略 patches 到 patches 交互的设计选择。此外，SurvPath 在不同放大倍数下的表现相似（在两种情况下整体 c 指数均为 0.629）。这一观察也适用于大多数组织学和多模态 baseline。</p>
<p><strong>稀疏 vs. 密集转录组学编码器：</strong> 密集的 MLP 提供了比稀疏途径感知 MLP 更好的表现。然而，当考虑的基因数量显著增加时，稀疏网络已显示出特别的参数效率，并且比常规 MLP 更具可解释性【20】。随着基因数量的增加，这一趋势可能会发展变化。</p>
<p><strong>早期融合 vs. 晚期融合：</strong> 早期融合方法（MCAT【10】、MOTCat【87】和 SurvPath）优于所有晚期融合方法。我们将这一观察归因于创建了一个可以模拟转录组学和组织学标记之间精细互动的联合特征空间。总的来说，这些发现证明了需要（1）模拟途径和 patches 标记之间的密集交互，以及（2）在单一 Transformer 注意力中统一融合的必要性。</p>
<h3 id="44">4.4 消融研究<a class="headerlink" href="#44" title="Permanent link">&para;</a></h3>
<p>为了评估我们的设计选择，我们进行了一系列消融研究，探讨了不同的标记器和融合方案。</p>
<p><strong>标记器：</strong> SURVPATH 采用 Reactome 和 Hallmarks 数据库作为生物途径的来源。我们评估了仅使用其中每个数据库的模型性能，以及将所有基因分配给一个标记（单一）和在【10】中使用的基因家族。随着转录组学标记的粒度增加，整体性能提高，表明构建具有语义的标记带来了可解释性属性并提高了性能。我们将这一观察归因于每个标记编码了越来越具体的生物功能，从而更好地实现了跨模态建模。</p>
<p><strong>融合：</strong> 我们通过进一步简化 Transformer 注意力，只考虑其左部分 <span class="arithmatex">\(A_{\mathcal{P}} \rightarrow \mathcal{P}\)</span> 和 <span class="arithmatex">\(A_{\mathscr{H} \rightarrow \mathcal{P}}\)</span>，以及其顶部部分 <span class="arithmatex">\(A_{\mathcal{P} \rightarrow \mathcal{P}}\)</span> 和 <span class="arithmatex">\(A_{\mathcal{P} \rightarrow \mathscr{H}}\)</span>（这种设计类似于 MCAT【10】中学习的单一共享多模态注意力层）来对 SURVPATH 进行消融。两个分支带来了互补信息（观察到 c 指数的下降分别为 -5.6% 和 -7.5%），这证明了需要模拟途径到 patches 和 patches 到途径的交互。我们进一步将 SURVPATH 适配为 Nyström 注意力，该注意力通过使用低秩近似简化自注意力，使得在非常长的序列上进行训练成为可能。这导致了性能显著下降 -6.9%。我们假设“真正的全注意力”具有低熵，使其更难通过低秩方法近似【8】，并且稀疏注意力模式提供了更好的近似。</p>
<p><img alt="" src="https://arxiv.org/html/2304.06819v2/x3.png" /></p>
<blockquote>
<p><strong>图 3：乳腺癌患者多层次可解释性可视化</strong>
顶部：低风险患者。底部：高风险患者。红色的基因和途径表示增加风险，蓝色的基因和途径表示降低风险。热图颜色表示重要性，红色表示高重要性，蓝色表示低重要性。在这些案例中识别为重要的途径和形态学通常与先前在侵袭性乳腺癌中描述的模式（例如，晚期雌激素反应）相符。</p>
</blockquote>
<h3 id="45">4.5 可解释性<a class="headerlink" href="#45" title="Permanent link">&para;</a></h3>
<p>对多层次可解释性的研究可以带来关于途径和组织学在决定患者风险方面的相互作用的新的生物学见解。这里，我们比较了乳腺侵袭性癌症（BRCA）的低风险（顶部）和高风险（底部）案例（图 3）以及膀胱尿路上皮癌（BLCA）（见补充材料）。</p>
<p>在分析图 3 时，我们观察到在低风险和高风险案例中，几个途径具有高绝对重要性分数，尤其是标志性上皮 - 间质转换（EMT）【79】和 COX 反应途径【50】，这两个途径已知参与乳腺癌。EMT 被认为是肿瘤细胞侵袭和转移的基础【34】，而这一途径对低风险和高风险案例的反向重要性与这一分析相符。这一发现通过研究突出 EMT 与肿瘤细胞群侵袭基质的关联的跨模态可解释性得到加强。COX 家族的环氧化酶成员，尤其是 COX-2，也与乳腺癌发生有关，并且作为治疗方案的组成部分正在研究中【26】。跨模态可解释性在两种案例中均显示了间质和免疫细胞。尽管图 3 中两种案例的重要途径有一些重叠，但大多数途径在两者之间是不同的。例如，在高风险案例中，识别了与铁代谢相关的途径（已知对乳腺癌的发生和预后有贡献【69】），显示了小团肿瘤细胞穿过密集基质侵袭的 patches。在低风险案例中，发现与细胞对雌激素反应相关的途径很重要，相应的 patches 显示了低级别侵袭性癌症或原位癌的形态，这与他人观察到的激素阳性乳腺癌往往等级较低并具有较长生存时间的观点一致【19】。有趣的是，标志性肌肉生成途径在图 3 的两种案例中都被赋予了相对较高的正面重要性。肌肉生成在乳腺癌中尚未被广泛研究，但假设肿瘤细胞自身可能表达这一途径中涉及的基因作为其上皮 - 间质转化的一部分，或它们诱导间质细胞这样做。这突出了我们的方法推动对后续研究的新的生物学见解的能力。</p>
<p>我们方法在提供单模态和跨模态可解释性方面的灵活性，使我们能够发现可能用于设计更好的癌症治疗方案的新的多模态生物标志物。随着我们对疾病分子基础的理解加深，SurvPath 的可解释性可能会激发针对特定形态和途径组合的靶向研究。</p>
<h3 id="_3">结论<a class="headerlink" href="#_3" title="Permanent link">&para;</a></h3>
<p>本文解决了转录组学和组织学多模态融合面临的两个挑战：（1）通过定义编码语义上有意义且可解释功能的生物途径标记，我们解决了转录组学标记化的挑战；（2）我们通过设计一个具有稀疏模态特定注意力模式的多模态 Transformer，克服了整合长多模态序列的计算挑战。我们的模型在测试 TCGA 的五个数据集时实现了最先进的生存性能。此外，我们的可解释性框架揭示了已知的和候选的预后特征。尽管我们的可解释性框架能够识别预后特征，但这些发现仍然是定性的。未来的工作可以关注在数据集级别推广发现的可解释性指标，例如，对特定途径的定量形态特征描述。此外，我们的发现表明，包括 patches 到 patches 的交互并没有提高性能。尽管如此，性能未提升不应被视为 patches 到 patches 交互不必要的证据，而应视为模拟这种交互是一个仍待解决的挑战性问题。</p>
<h2 id="_4">补充材料<a class="headerlink" href="#_4" title="Permanent link">&para;</a></h2>
<h3 id="1">1. 生存预测<a class="headerlink" href="#1" title="Permanent link">&para;</a></h3>
<p>根据前面引入的符号，我们的目标是通过多模态嵌入 <span class="arithmatex">\(\overline{\mathrm{x}}_{\mathrm{Att}} \in \mathbb{R}^{2d}\)</span> 来预测患者的生存情况。与之前的工作【90】一致，我们通过以下方式定义患者的生存状态：(1) 审查状态 <span class="arithmatex">\(c\)</span>，其中 <span class="arithmatex">\(c=0\)</span> 表示观测到的患者死亡，<span class="arithmatex">\(c=1\)</span> 对应于患者的最后已知随访；(2) 事件到发生时间 <span class="arithmatex">\(t_i\)</span>，如果 <span class="arithmatex">\(c=0\)</span>，对应于患者诊断和观测到的死亡之间的时间，如果 <span class="arithmatex">\(c=1\)</span>，则对应于最后一次随访。我们没有直接预测事件发生的观测时间 <span class="arithmatex">\(t\)</span>，而是通过定义基于生存时间值四分位数的非重叠时间间隔 <span class="arithmatex">\(\left(t_{j-1}, t_j\right)，j \in[1, \ldots, n]\)</span> 并表示为 <span class="arithmatex">\(y_j\)</span> 来近似它。这个问题简化为带有审查信息的分类问题，每个患者现在由 <span class="arithmatex">\(\left(\overline{\mathrm{x}}_{\mathrm{Att}}, y_j, c\right)\)</span> 定义。我们构建了一个分类器，网络预测的每个输出逻辑 <span class="arithmatex">\(\hat{y}_j\)</span> 对应一个时间间隔。由此，我们定义离散危险函数 <span class="arithmatex">\(f_{\text {hazard }}\left(y_j \mid \overline{\mathrm{x}}_{\mathrm{Att}}\right)=S\left(\hat{y}_j\right)\)</span>，其中 <span class="arithmatex">\(S\)</span> 是 sigmoid 激活函数。直观上，<span class="arithmatex">\(f_{\text {hazard }}\left(y_j \mid \overline{\mathrm{x}}_{\text {Att }}\right)\)</span> 表示患者在时间间隔 <span class="arithmatex">\(\left(t_{j-1}, t_j\right)\)</span> 内死亡的概率。此外，我们定义了离散生存函数 <span class="arithmatex">\(f_{\text {surv }}\left(y_j \mid \overline{\mathrm{x}}_{\text {Att }}\right) = \prod_{k=1}^j\left(1-f_{\text {hazard }}\left(y_k \mid \overline{\mathrm{x}}_{\text {Att }}\right)\right)\)</span>，表示患者在时间间隔 <span class="arithmatex">\(\left(t_{j-1}, t_j\right)\)</span> 内生存的概率。这使我们能够定义负对数似然（NLL）生存损失【90】，将 NLL 推广到具有审查的数据。正式地，我们表示为：</p>
<div class="arithmatex">\[
\begin{aligned}
&amp; \mathcal{L}\left(\left\{\overline{\mathrm{x}}_{\text {Att }}^{(i)}, y_j^{(i)}, c^{(i)}\right\}_{i=1}^{N_{\mathcal{D}}}\right)= \\
&amp; \quad \sum_{i=1}^{N_{\mathcal{D}}}-c^{(i)} \log \left(f_{\text {surv }}\left(y_j^{(i)} \mid \overline{\mathrm{x}}_{\text {Att }}^{(i)}\right)\right) \\
&amp; \quad+\left(1-c^{(i)}\right) \log \left(f_{\text {surv }}\left(y_j^{(i)}-1 \mid \overline{\mathrm{x}}_{\text {Att }}^{(i)}\right)\right) \\
&amp; \quad+\left(1-c^{(i)}\right) \log \left(f_{\text {hazard }}\left(y_j^{(i)} \mid \overline{\mathrm{x}}_{\text {Att }}^{(i)}\right)\right)
\end{aligned}\tag{(1)(2)(3)(4)}
\]</div>
<p>其中 <span class="arithmatex">\(N_{\mathcal{D}}\)</span> 是数据集中的样本数。直观上，方程 2 确保了在最终随访后仍然存活的患者具有高生存概率，方程 3 确保了观察到死亡的患者在死亡时间戳之前有高生存概率，方程 4 确保了观察到死亡的患者的正确时间戳被预测。更全面的数学描述可以在【90】中找到。</p>
<p>最后，通过取所有逻辑的负和，我们可以定义用于识别不同风险组和分层患者的患者级风险。</p>
<h3 id="2">2. 实施细节<a class="headerlink" href="#2" title="Permanent link">&para;</a></h3>
<h4 id="21_1">2.1 模型训练<a class="headerlink" href="#21_1" title="Permanent link">&para;</a></h4>
<p>代码使用 Python 3.9 实现，模型在 PyTorch 中实现，可解释性基于 Captum【37】。SuRvPATH、baseline 和消融研究使用 RAdam 优化器【45】进行优化，批大小为 1，学习率为 <span class="arithmatex">\(5 \times 10^{-4}\)</span>，权重衰减为 <span class="arithmatex">\(10^{-3}\)</span>。patches 编码器产生 768 维嵌入（CTransPath 输出），这些嵌入被投影到 <span class="arithmatex">\(d=256\)</span>，即标记维度。转录组学编码器由 2 层前馈网络组成，带有 alpha dropout【36】以产生途径标记。Transformer 实现为单头和单层，不使用类（CLS）标记。Transformer 后跟一个层归一化、一个前馈层和一个 2 层分类头。所有模型训练均在单个 NVIDIA RTX 3090Ti 上完成。</p>
<h4 id="22">2.2 评估指标<a class="headerlink" href="#22" title="Permanent link">&para;</a></h4>
<p>模型使用以下两个指标进行评估：(1) 一致性指数（c-index，越高越好），这一指标测量模型预测值正确预测实际生存顺序的所有可能观察对的比例（从 0.5（随机预测）到 1.0（完美预测））；(2) Kaplan-Meier（KM）曲线，用于可视化不同风险组患者在一定时间内的生存概率。我们应用 logrank 统计显著性检验来确定低风险和高风险组之间的分离是否具有统计意义（<span class="arithmatex">\(\mathrm{p}\)</span> 值 <span class="arithmatex">\(&lt;0.05\)</span>）。</p>
<h3 id="3_2">3. 额外的可解释性<a class="headerlink" href="#3_2" title="Permanent link">&para;</a></h3>
<p>图 1 展示了所提出的多层次可解释性框架的高级描述。</p>
<p>为了补充主文中提出的可解释性分析，我们进一步分析了 BLCA 中的低风险和高风险案例（见图 2）。组织学解释表明，健康膀胱肌肉的存在降低风险，而多形性肿瘤细胞和泡沫样细胞质增加风险。大多数重要途径与细胞周期控制（例如 G2M 检查点，SCF <span class="arithmatex">\(\beta \operatorname{TrCP}\)</span> 降解 em1）、代谢（例如脂肪酸代谢）和免疫相关功能（异体排斥和 IL2 STAT5 信号）相关。途径对总体风险的贡献也与以前的文献一致。例如，先前的途径表达分析发现 G2M 检查点和免疫相关途径在预测膀胱癌预后中具有重要意义【33】。通过 SURVPATH 发现的跨模态交互的定性评估在科学上是合理的。例如，异体排斥途径包括多个在免疫反应中激活的基因。在低风险案例中，异体排斥高度关注肿瘤浸润的淋巴细胞和靠近膀胱肌肉墙的淋巴细胞集合。在高风险案例中，这一途径再次关注分散在肌肉壁中的炎症细胞。SCF <span class="arithmatex">\(\beta \operatorname{TrCP}\)</span> 降解 em1 途径在控制有丝分裂的细胞分裂中起重要作用。在低风险案例中，该途径关注未受累的膀胱肌肉；而在高风险案例中，同一途径关注侵入膀胱肌肉的肿瘤细胞。尽管在低风险和高风险案例中存在途径的重叠，但 SURVPATH 也识别出仅在一种案例中出现的途径。例如，在低风险案例中，SURVPATH 发现蛋白质分泌途径高度关注肿瘤细胞而非健康的膀胱肌肉细胞。在两种情况下，G2M 检查点途径（对细胞周期的健康进程至关重要）都被认为是重要的。在高风险案例中，我们看到这一途径主要是增加风险。有趣的是，我们还发现这一途径关注大面积的坏死区，这是合理的，因为细胞周期调控的异常会导致细胞死亡。</p>
<p><img alt="" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202404251433354.png" /></p>
<blockquote>
<p><strong>图 1. 多层次可解释性框架。</strong> 从包含全幅切片图像（WSI）和转录组学测量的多模态输入以及预测的风险，我们可以在切片级、基因级和生物途径级别上进行风险归因。该框架还支持研究途径到 patches 的交互和 patches 到途径的交互，以解开两种模态之间的对应关系。</p>
</blockquote>
<h3 id="4_1">4. 额外结果<a class="headerlink" href="#4_1" title="Permanent link">&para;</a></h3>
<h4 id="10">10 倍放大下的结果<a class="headerlink" href="#10" title="Permanent link">&para;</a></h4>
<p>我们也在 10 倍放大下对 SURVPATH 及 baseline（表 1）和消融研究（表 2）进行了分析。从 20 倍放大分析的趋势仍然存在：（1）SURVPATH 实现了最佳的整体性能，（2）转录组学 baseline 仍是强有力的竞争者，（3）多模态模型提供了更好的整体性能。有趣的是，SURVPATH 在 10 倍和 20 倍放大下提供相同的性能（在五个队列中为 62.9%）。</p>
<h4 id="kaplan-meier">Kaplan-Meier 分析<a class="headerlink" href="#kaplan-meier" title="Permanent link">&para;</a></h4>
<p>图 3 展示了 20 倍放大下预测的高风险和低风险组的 Kaplan-Meier 生存曲线。所有风险高于整个队列中位数的患者被分配为高风险（红色），风险低于中位数的患者被分配为低风险（蓝色）。对于所有五种疾病，SURVPATH 相较于最佳组织学 baseline（TransMIL）、转录组学 baseline（MLP）和多模态 baseline（MCAT），在两个风险组之间展示了统计上更好的区分度。</p>
<p>我们相信 SURVPATH 能更好地区分风险组，因为简化的早期融合机制使其能够在患者风险方面找到转录组学和组织学之间更好的相关性。</p>
<h4 id="_5">与临床协变量的比较<a class="headerlink" href="#_5" title="Permanent link">&para;</a></h4>
<p>从临床上，预后可以基于患者信息如年龄，以及癌症进展评估如癌症分级。我们使用 Cox 比例风险模型从临床协变量（年龄、性别、分级）单独和组合预测生存。我们发现 SURVPATH 在所有临床协变量的生存预测中表现更佳（表 3）。</p>
<h4 id="_6">模态归因<a class="headerlink" href="#_6" title="Permanent link">&para;</a></h4>
<p>通过对每种模态的整合梯度（IG）值进行预协注意总和，我们可以得出模态归因分数（表 4）。我们发现组织学在队列中贡献了 77.2%，突显了多模态在预后中的必要性。</p>
<p><img alt="" src="https://arxiv.org/html/2304.06819v2/x5.png" /></p>
<blockquote>
<p><strong>图 2：膀胱癌患者的多层次可解释性可视化</strong>
顶部：低风险患者。底部：高风险患者。红色的基因和途径表示增加风险，蓝色的基因和途径表示降低风险。热图颜色表示重要性，红色表示高重要性，蓝色表示低重要性。在这些案例中识别为重要的途径和形态通常与先前在膀胱尿路上皮癌中描述的模式相符（例如，G2M 检查点）。</p>
</blockquote>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./" class="md-footer__link md-footer__link--prev" aria-label="Previous: Recurrent memory with optimal polynomial projections.">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Recurrent memory with optimal polynomial projections.
              </div>
            </div>
          </a>
        
        
          
          <a href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/" class="md-footer__link md-footer__link--next" aria-label="Next: Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.code.select", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.instant.prefetch", "navigation.instant.progress", "navigation.path", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "header.autohide"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.525ec568.min.js"></script>
      
        <script src="../mathjax-config.js"></script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../extra.js"></script>
      
        <script src="../js/toggle_sidebar.js"></script>
      
    
  </body>
</html>