
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://localhost:8000/Spatial%20transition%20tensor%20of%20single%20cells/">
      
      
        <link rel="prev" href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/">
      
      
        <link rel="next" href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/">
      
      
      <link rel="icon" href="../img/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>Spatial transition tensor of single cells - One Last Kiss</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lxgw-wenkai-webfont@1.1.0/style.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-M0R7V8QGQ7"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-M0R7V8QGQ7",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-M0R7V8QGQ7",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#_1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="One Last Kiss" class="md-header__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            One Last Kiss
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Spatial transition tensor of single cells
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
    
  
  🏠 Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Mygo/" class="md-tabs__link">
          
  
    
  
  ⛏️ CV and Multimodels

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Session%201/" class="md-tabs__link">
          
  
    
  
  🚀 Re:0 Road to CPath & Visualization

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-tabs__link">
          
  
    
  
  💤 Re:0 Road to scRNA-seq Analysis

        </a>
      </li>
    
  

    
  

    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-tabs__link">
          
  
    
  
  🧬 scRNA Analysis

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-tabs__link">
          
  
    
  
  📐 Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-tabs__link">
          
  
    
  
  💡 Reviews

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="One Last Kiss" class="md-nav__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    One Last Kiss
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href=".." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    🏠 Home
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            🏠 Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⛏️ CV and Multimodels
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ⛏️ CV and Multimodels
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mygo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MyGO: Discrete Modality Information as Fine-Grained Tokens for Multi-modal Knowledge Graph Completion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GMM/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GMM
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../DINO/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    DINO
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mamba/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mamba
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ViT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ViT
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCoV2%2C3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo V2
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%80%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (一)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%20%28%E4%BA%8C%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (二)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%89%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (三)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E5%9B%9B%29%20%28TNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (四)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E7%9F%A5%E8%AF%86%E8%92%B8%E9%A6%8F%20%E9%A2%84%E8%AE%AD%E7%BB%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (知识蒸馏 预训练）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28PyramidTNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer(Pyramid TNT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%82%85%E9%87%8C%E5%8F%B6%E7%BA%A7%E6%95%B0%E5%92%8C%E5%82%85%E9%87%8C%E5%8F%B6%E5%8F%98%E6%8D%A2%E4%B8%93%E9%A2%98/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    傅里叶级数和傅里叶变换专题
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Python%20%E5%AD%98%E5%82%A8%E4%B8%8E%E8%AF%BB%E5%8F%96%20HDF5%20%E6%96%87%E4%BB%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Python 存储与读取 HDF5 文件
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Segment%20Anything%20in%20Medical%20Images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Segment Anything in Medical Images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Mamba%20Efficient%20Visual%20Representation%20Learning%20with%20Bidirectional%20State%20Space%20Model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Mamba Efficient Visual Representation Learning with Bidirectional State Space Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_18" >
        
          
          <label class="md-nav__link" for="__nav_2_18" id="__nav_2_18_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚑 医学影像相关
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_18_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_18">
            <span class="md-nav__icon md-icon"></span>
            🚑 医学影像相关
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Why%20is%20the%20winner%20the%20best/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Why is the winner the best
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PORPOISE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PORPOISE
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HIPT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Scaling Vision Transformers to Gigapixel Images via Hierarchical Self-Supervised Learning(HIPT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Data-efficient%20and%20weakly%20supervised%20computational%20pathology%20on%20whole-slide%20images%28CLAM%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data efficient and weakly supervised computational pathology on whole slide images(CLAM)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Towards%20a%20general-purpose%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Towards a general purpose foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20whole-slide%20foundation%20model%20for%20digital%20pathology%20from%20real-world%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A whole slide foundation model for digital pathology from real world data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20visual-language%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A visual language foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Recurrent memory with optimal polynomial projections.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Modeling%20Dense%20Multimodal%20Interactions%20Between%20Biological%20Pathways%20and%20Histology%20for%20Survival%20Prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Prediction%20of%20recurrence%20risk%20in%20endometrial%20cancer%20with%20multimodal%20deep%20learning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction of recurrence risk in endometrial cancer with multimodal deep learning
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚀 Re:0 Road to CPath & Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            🚀 Re:0 Road to CPath & Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%201/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    理解全幅切片图像
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%202/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    在计算病理学中用于癌症诊断的弱监督深度学习
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%203/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    基于注意力的多示例学习可解释性
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../How%20to%20visualize%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to visualize Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AttentionViz%20A%20Global%20View%20of%20Transformer%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AttentionViz A Global View of Transformer Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Transformer%20Interpretability%20Beyond%20Attention%20Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transformer Interpretability Beyond Attention Visualization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualization Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Attention%20Generation%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Attention Generation Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🪛 PyTorch
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            🪛 PyTorch
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Autograd%20mechanics%20%E2%80%94%20PyTorch%202_3%20documentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Autograd mechanics — PyTorch 2 3 documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyTorch Autograd Explained   In depth Tutorial
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💤 Re:0 Road to scRNA-seq Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            💤 Re:0 Road to scRNA-seq Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏳ CHAPTER 1 简介
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ⏳ CHAPTER 1 简介
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1.The building block of life
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20Raw%20Data%20Processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 Raw data processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Analysis%20frameworks%20and%20tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Analysis frameworks and tools
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🌏 CHAPTER 2 数据准备与可视化
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            🌏 CHAPTER 2 数据准备与可视化
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20normalization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 Normalization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3_feature_selection/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 Feature selection
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.4_dimensionality_reduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.4 Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_3" >
        
          
          <label class="md-nav__link" for="__nav_4_1_3" id="__nav_4_1_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🍀 CHAPTER 3 识别细胞结构
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_3">
            <span class="md-nav__icon md-icon"></span>
            🍀 CHAPTER 3 识别细胞结构
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 Data integration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_4" >
        
          
          <label class="md-nav__link" for="__nav_4_1_4" id="__nav_4_1_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏱️ CHAPTER 4 推断细胞轨迹
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_4">
            <span class="md-nav__icon md-icon"></span>
            ⏱️ CHAPTER 4 推断细胞轨迹
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotemporal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 Pseudotemporal ordering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../rna_velocity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 RNA velocity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../lineage_tracing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 Lineage tracing
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_5" >
        
          
          <label class="md-nav__link" for="__nav_4_1_5" id="__nav_4_1_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🎁 CHAPTER 5 应对特殊情况
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_5">
            <span class="md-nav__icon md-icon"></span>
            🎁 CHAPTER 5 应对特殊情况
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../differential_gene_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 Differential gene expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compositional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 Compositional analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gsea_pathway/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 Gene set enrichment and pathway analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../perturbation_modeling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 Perturbation modeling
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_6" >
        
          
          <label class="md-nav__link" for="__nav_4_1_6" id="__nav_4_1_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⚙️ CHAPTER 6 机制建模
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_6">
            <span class="md-nav__icon md-icon"></span>
            ⚙️ CHAPTER 6 机制建模
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_regulatory_networks/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 Gene regulatory networks
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cell_cell_communication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 Cell-cell communication
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_7" >
        
          
          <label class="md-nav__link" for="__nav_4_1_7" id="__nav_4_1_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🔓 CHAPTER 7 反卷积
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_7">
            <span class="md-nav__icon md-icon"></span>
            🔓 CHAPTER 7 反卷积
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bulk_deconvolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 Bulk deconvolution
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    简明易懂的单细胞分析流程（使用 R 语言）
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            简明易懂的单细胞分析流程（使用 R 语言）
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%89%8D%E8%A8%80/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    前言
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    环境配置
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20%E6%9E%84%E5%BB%BA%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1 构建单细胞 RNA seq 分析环境
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20R%20%E7%8E%AF%E5%A2%83%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 R 环境安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Docker%20%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Docker 安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.4%20%E5%9C%A8%20Docker%20%E4%B8%AD%E5%90%AF%E5%8A%A8%20Rstudio/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.4 在 Docker 中启动 Rstudio
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.5%20%E6%9C%89%E5%85%B3%20Bioconductor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.5 有关 Bioconductor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20R%20%E7%9A%84%E5%9F%BA%E6%9C%AC%E7%94%A8%E6%B3%95/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 R 的基本用法
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20%E5%A6%82%E4%BD%95%E5%A4%84%E7%90%86%20DataFrame%EF%BC%9F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 如何处理 DataFrame？
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3%20%E4%BD%BF%E7%94%A8%20ggplot2%20%E5%8F%AF%E8%A7%86%E5%8C%96%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 使用 ggplot2 可视化数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1%20PCA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 PCA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2%20%E4%BD%BF%E7%94%A8%20clusterPlofiler%20%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 使用 clusterPlofiler 进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E5%9F%BA%E5%9B%A0%E6%9C%AC%E4%BD%93%EF%BC%88GO%EF%BC%89%E4%B8%AD%E5%90%84%20BP%E3%80%81MF%20%E5%92%8C%20CC%20%E7%9A%84%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 基因本体（GO）中各 BP、MF 和 CC 的富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E4%BD%BF%E7%94%A8%20DEseq2%20%E6%8F%90%E5%8F%96%20RNA-seq%20%E5%88%86%E6%9E%90%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%20DEG/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 使用 DEseq2 提取 RNA seq 分析数据中的 DEG
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.4%20%E4%BD%BF%E7%94%A8%E5%85%B6%E4%BB%96%E6%95%B0%E6%8D%AE%E5%BA%93%EF%BC%88KEGG%E3%80%81Reactome%20%E7%AD%89%EF%BC%89%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.4 使用其他数据库（KEGG、Reactome 等）进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.1%20scRNA-seq%20%E6%A6%82%E8%A6%81/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 scRNA seq 概要
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.2%20%E5%87%86%E5%A4%87%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 准备使用 Seurat 进行分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.3%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 使用 Seurat 进行数据预处理
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.4%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E7%9A%84%E4%B8%80%E8%88%AC%E9%A1%BA%E5%BA%8F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.4 使用 Seurat 进行数据分析的一般顺序
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.1%20%E5%A6%82%E4%BD%95%E6%9F%A5%E6%89%BE%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 如何查找单细胞 RNA seq 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.2%20%E4%B8%BA%E6%AF%8F%E4%B8%AA%E7%B0%87%E8%87%AA%E5%8A%A8%E6%B7%BB%E5%8A%A0%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E6%A0%87%E7%AD%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 为每个簇自动添加细胞注释标签
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.3%20%E4%BD%BF%E7%94%A8%20FeaturePlot%20%E6%A3%80%E6%9F%A5%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 使用 FeaturePlot 检查基因表达
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 将单细胞 RNA seq 数据集相互结合
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.5%20%E7%BA%A0%E6%AD%A3%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94%EF%BC%8C%E6%AF%94%E8%BE%83%E5%AF%B9%E7%85%A7%E7%BB%84%E5%92%8C%E5%88%BA%E6%BF%80%E7%BB%84/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.5 纠正批次效应，比较对照组和刺激组
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.6 确定每个群组的标记基因
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.1%20%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E7%BB%9F%E8%AE%A1%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E6%B0%B4%E5%B9%B3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 如何使用 Cell Ranger 统计 scRNA seq 数据中的基因表达水平
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.2%20%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E5%A4%84%E7%90%86%20NCBI%20Sequence%20Read%20Archive%20%28SRA%29%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 使用 Cell Ranger 处理 NCBI Sequence Read Archive (SRA) 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.1%20%E5%85%B3%E4%BA%8E%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 关于 scRNA seq 数据的质量控制
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.2%20%E4%BD%BF%E7%94%A8%20DoubleFinder%20%E6%A3%80%E6%B5%8B%E5%8F%8C%E9%87%8D%E7%BB%86%E8%83%9E/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.2 使用 DoubleFinder 检测双重细胞
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.3%20%E7%94%A8%20CellBender%20%E5%8E%BB%E9%99%A4%E8%83%8C%E6%99%AF%20RNA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.3 用 CellBender 去除背景 RNA
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    🧬 scRNA Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            🧬 scRNA Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    🐍 Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            🐍 Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Large scale foundation model on single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scGPT%20toward%20building%20a%20foundation%20model%20for%20single-cell%20multi-omics%20using%20generative%20AI/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scGPT toward building a foundation model for single cell multi omics using generative AI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scBERT%20as%20a%20large-scale%20pretrained%20deep%20language%20model%20for%20cell%20type%20annotation%20of%20single-cell%20RNA-seq%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scBERT as a large scale pretrained deep language model for cell type annotation of single cell RNA seq data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RNA%20velocity%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA velocity of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TFvelo_gene%20regulation%20inspired%20RNA%20velocity%20estimation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TFvelo gene regulation inspired RNA velocity estimation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Disentanglement%20of%20single-cell%20data%20with%20biolord/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Disentanglement of single cell data with biolord
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Spatial transition tensor of single cells
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Spatial transition tensor of single cells
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      摘要
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      主文
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      结果
    </span>
  </a>
  
    <nav class="md-nav" aria-label="结果">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#stt" class="md-nav__link">
    <span class="md-ellipsis">
      STT 概述
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_1" class="md-nav__link">
    <span class="md-ellipsis">
      多稳态细胞状态恢复的 STT 基准测试
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt-ics" class="md-nav__link">
    <span class="md-ellipsis">
      STT 突出了命运决策中的 iCS
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_2" class="md-nav__link">
    <span class="md-ellipsis">
      STT 识别空间吸引子和路径相似性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_3" class="md-nav__link">
    <span class="md-ellipsis">
      STT 揭示了鸡心的空间吸引子和谱系
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_4" class="md-nav__link">
    <span class="md-ellipsis">
      STT 阐明了区域特异性的空间吸引子及其稳定性
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      讨论
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      方法
    </span>
  </a>
  
    <nav class="md-nav" aria-label="方法">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      基因表达和剪接模型中的多稳态性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_7" class="md-nav__link">
    <span class="md-ellipsis">
      参数估计和吸引子归属量化的迭代方案
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_8" class="md-nav__link">
    <span class="md-ellipsis">
      初始化迭代
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_9" class="md-nav__link">
    <span class="md-ellipsis">
      动态流形和过渡路径的可视化
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_10" class="md-nav__link">
    <span class="md-ellipsis">
      数据可用性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_11" class="md-nav__link">
    <span class="md-ellipsis">
      代码可用性
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRank%20infers%20drug-responsive%20cell%20types%20from%20untreated%20scRNA-seq%20data%20using%20a%20target-perturbed%20gene%20regulatory%20network/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRank infers drug responsive cell types from untreated scRNA seq data using a target perturbed gene regulatory network
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    📔 文献
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            📔 文献
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ZNF689%20deficiency%20promotes%20intratumor%20heterogeneity%20and%20immunotherapy%20resistance%20in%20triple-negative%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ZNF689 deficiency promotes intratumor heterogeneity and immunotherapy resistance in triple negative breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Tumour%20vasculature%20at%20single-cell%20resolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tumour vasculature at single-cell resolution
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRNA-seq%20of%20gastric%20tumor%20shows%20complex%20intercellular%20interaction%20with%20an%20alternative%20T%20cell%20exhaustion%20trajectory%EF%BC%88%E5%8F%82%E8%80%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRNA seq of gastric tumor shows complex intercellular interaction with an alternative T cell exhaustion trajectory（参考）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20reveals%20new%20evolutionary%20complexity%20in%20uveal%20melanoma%EF%BC%88%E5%8F%82%E8%80%83%202%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis reveals new evolutionary complexity in uveal melanoma（参考 2）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20characterization%20of%20macrophages%20in%20uveal%20melanoma%20uncovers%20transcriptionally%20heterogeneous%20subsets%20conferring%20poor%20prognosis%20and%20aggressive%20behavior%EF%BC%88%E8%8C%83%E4%BE%8B%201%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell characterization of macrophages in uveal melanoma uncovers transcriptionally heterogeneous subsets conferring poor prognosis and aggressive behavior（范例 1）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ScRNA-seq%20of%20gastric%20cancer%20tissues%20reveals%20differences%20in%20the%20immune%20microenvironment%20of%20primary%20tumors%20and%20metastases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ScRNA seq of gastric cancer tissues reveals differences in the immune microenvironment of primary tumors and metastases
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Blimp-1%20and%20c-Maf%20regulate%20immune%20gene%20networks%20to%20protect%20against%20distinct%20pathways%20of%20pathobiont-induced%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Blimp 1 and c Maf regulate immune gene networks to protect against distinct pathways of pathobiont induced colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single%20cell%20analysis%20unveils%20B%20cell-dominated%20immune%20subtypes%20in%20HNSCC%20for%20enhanced%20prognostic%20and%20therapeutic%20stratification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis unveils B cell dominated immune subtypes in HNSCC for enhanced prognostic and therapeutic stratification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20spatial%20multi-omics%20and%20deep%20learning%20dissect%20enhancer-driven%20gene%20regulatory%20networks%20in%20liver%20zonation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell spatial multi omics and deep learning dissect enhancer driven gene regulatory networks in liver zonation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Neutrophil%20profiling%20illuminates%20anti-tumor%20antigen%20presenting%20potency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Neutrophil profiling illuminates anti tumor antigen presenting potency
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20blueprint%20for%20tumor-infiltrating%20B%20cells%20across%20human%20cancers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A blueprint for tumor infiltrating B cells across human cancers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20transcriptomic%20analyses%20reveal%20distinct%20immune%20cell%20contributions%20to%20epithelial%20barrier%20dysfunction%20in%20checkpoint%20inhibitor%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell transcriptomic analyses reveal distinct immune cell contributions to epithelial barrier dysfunction in checkpoint inhibitor colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Systematic%20dissection%20of%20tumor-normal%20single-cell%20ecosystems%20across%20a%20thousand%20tumors%20of%2030%20cancer%20types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Systematic dissection of tumor normal single cell ecosystems across a thousand tumors of 30 cancer types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../The%20aged%20tumor%20microenvironment%20limits%20T%20cell%20control%20of%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The aged tumor microenvironment limits T cell control of cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TIM-3%2B%20CD8%20T%20cells%20with%20a%20terminally%20exhausted%20phenotype%20retain%20functional%20capacity%20in%20hematological%20malignancies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TIM 3+ CD8 T cells with a terminally exhausted phenotype retain functional capacity in hematological malignancies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20lactate-SREBP2%20signaling%20axis%20drives%20tolerogenic%20dendritic%20cell%20maturation%20and%20promotes%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A lactate SREBP2 signaling axis drives tolerogenic dendritic cell maturation and promotes cancer progression
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20human%20omentum-specific%20mesothelial-like%20stromal%20population%20inhibits%20adipogenesis%20through%20IGFBP2%20secretion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A human omentum specific mesothelial like stromal population inhibits adipogenesis through IGFBP2 secretion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Interferon-stimulated%20neutrophils%20as%20a%20predictor%20of%20immunotherapy%20response/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interferon stimulated neutrophils as a predictor of immunotherapy response
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Repression%20of%20latent%20NF-%CE%BAB%20enhancers%20by%20PDX1%20regulates%20%CE%B2%20cell%20functional%20heterogeneity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Repression of latent NF κB enhancers by PDX1 regulates β cell functional heterogeneity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20systematic%20pan-cancer%20study%20on%20deep%20learning-based%20prediction%20of%20multi-omic%20biomarkers%20from%20routine%20pathology%20images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A systematic pan cancer study on deep learning based prediction of multi omic biomarkers from routine pathology images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20and%20spatial%20transcriptomics%20analysis%20of%20non-small%20cell%20lung%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell and spatial transcriptomics analysis of non small cell lung cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Seeing%20data%20as%20t-SNE%20and%20UMAP%20do/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Seeing data as t SNE and UMAP do
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mapping%20the%20cellular%20biogeography%20of%20human%20bone%20marrow%20niches%20using%20single-cell%20transcriptomics%20and%20proteomic%20imaging/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mapping the cellular biogeography of human bone marrow niches using single cell transcriptomics and proteomic imaging
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Onco-fetal%20Reprogramming%20of%20Endothelial%20Cells%20Drives%20Immunosuppressive%20Macrophages%20in%20Hepatocellular%20Carcinoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Onco fetal Reprogramming of Endothelial Cells Drives Immunosuppressive Macrophages in Hepatocellular Carcinoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20of%20anti-BCMA%20CAR%20T%20cell%20therapy%20in%20patients%20with%20central%20nervous%20system%20autoimmunity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis of anti BCMA CAR T cell therapy in patients with central nervous system autoimmunity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20sequencing%20depicts%20tumor%20architecture%20and%20empowers%20clinical%20decision%20in%20metastatic%20conjunctival%20melanoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell sequencing depicts tumor architecture and empowers clinical decision in metastatic conjunctival melanoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PR-SET7%20epigenetically%20restrains%20uterine%20interferon%20response%20and%20cell%20death%20governing%20proper%20postnatal%20stromal%20development/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PR SET7 epigenetically restrains uterine interferon response and cell death governing proper postnatal stromal development
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Clonal%20associations%20between%20lymphocyte%20subsets%20and%20functional%20states%20in%20rheumatoid%20arthritis%20synovium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clonal associations between lymphocyte subsets and functional states in rheumatoid arthritis synovium
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📐 Data Structures
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            📐 Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 1 章 python 入门
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%202%20%E7%AB%A0%20%E9%9D%A2%E5%90%91%E5%AF%B9%E8%B1%A1%E7%BC%96%E7%A8%8B/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 2 章 面向对象编程
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%203%20%E7%AB%A0%20%E7%AE%97%E6%B3%95%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 3 章 算法分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%204%20%E7%AB%A0%20%E9%80%92%E5%BD%92/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 4 章 递归
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%205%20%E7%AB%A0%20%E5%9F%BA%E4%BA%8E%E6%95%B0%E7%BB%84%E7%9A%84%E5%BA%8F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 5 章 基于数组的序列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%206%20%E7%AB%A0%20%E6%A0%88%E3%80%81%E9%98%9F%E5%88%97%E5%92%8C%E5%8F%8C%E7%AB%AF%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 6 章 栈、队列和双端队列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%207%20%E7%AB%A0%20%E9%93%BE%E8%A1%A8%28Linked%20Lists%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 7 章 链表(Linked Lists)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%208%20%E7%AB%A0%20%E6%A0%91%20%E2%98%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 8 章 树 ★
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%209%20%E7%AB%A0%20%E4%BC%98%E5%85%88%E7%BA%A7%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 9 章 优先级队列
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💡 Reviews
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            💡 Reviews
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_1" >
        
          
          <label class="md-nav__link" for="__nav_7_1" id="__nav_7_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧭 scRNA analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_1">
            <span class="md-nav__icon md-icon"></span>
            🧭 scRNA analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Best practices for single cell analysis across modalities
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Multi-omics%20integration%20in%20biomedical%20research%20%E2%80%93%20A%20metabolomics-centric%20review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multi omics integration in biomedical research – A metabolomics centric review
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Museum%20of%20spatial%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Museum of spatial transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20guide%20to%20artificial%20intelligence%20for%20cancer%20researchers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A guide to artificial intelligence for cancer researchers
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_2" >
        
          
          <label class="md-nav__link" for="__nav_7_2" id="__nav_7_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🦠 Metabolism
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_2">
            <span class="md-nav__icon md-icon"></span>
            🦠 Metabolism
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20analysis%20as%20a%20driver%20for%20discovery%2C%20diagnosis%2C%20and%20therapy/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic analysis as a driver for discovery, diagnosis, and therapy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20heterogeneity%20in%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic heterogeneity in cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Toward%20modeling%20metabolic%20state%20from%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Toward modeling metabolic state from single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20reprogramming%20and%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic reprogramming and cancer progression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_3" >
        
          
          <label class="md-nav__link" for="__nav_7_3" id="__nav_7_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Protein
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_3">
            <span class="md-nav__icon md-icon"></span>
            Protein
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Proteomic%20and%20interactomic%20insights%20into%20the%20molecular%20basis%20of%20cell%20functional%20diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Proteomic and interactomic insights into the molecular basis of cell functional diversity
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      摘要
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      主文
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      结果
    </span>
  </a>
  
    <nav class="md-nav" aria-label="结果">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#stt" class="md-nav__link">
    <span class="md-ellipsis">
      STT 概述
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_1" class="md-nav__link">
    <span class="md-ellipsis">
      多稳态细胞状态恢复的 STT 基准测试
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt-ics" class="md-nav__link">
    <span class="md-ellipsis">
      STT 突出了命运决策中的 iCS
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_2" class="md-nav__link">
    <span class="md-ellipsis">
      STT 识别空间吸引子和路径相似性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_3" class="md-nav__link">
    <span class="md-ellipsis">
      STT 揭示了鸡心的空间吸引子和谱系
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#stt_4" class="md-nav__link">
    <span class="md-ellipsis">
      STT 阐明了区域特异性的空间吸引子及其稳定性
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      讨论
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      方法
    </span>
  </a>
  
    <nav class="md-nav" aria-label="方法">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      基因表达和剪接模型中的多稳态性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_7" class="md-nav__link">
    <span class="md-ellipsis">
      参数估计和吸引子归属量化的迭代方案
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_8" class="md-nav__link">
    <span class="md-ellipsis">
      初始化迭代
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_9" class="md-nav__link">
    <span class="md-ellipsis">
      动态流形和过渡路径的可视化
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_10" class="md-nav__link">
    <span class="md-ellipsis">
      数据可用性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_11" class="md-nav__link">
    <span class="md-ellipsis">
      代码可用性
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/aletolia/aletolia.github.io.git/edit/master/docs/Spatial transition tensor of single cells.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/aletolia/aletolia.github.io.git/raw/master/docs/Spatial transition tensor of single cells.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


  <h1>Spatial transition tensor of single cells</h1>

<h3 id="_1">摘要<a class="headerlink" href="#_1" title="Permanent link">&para;</a></h3>
<p>空间转录组学和信使 RNA 剪接编码了细胞状态和转变的广泛时空信息。当前的谱系推断方法要么缺乏状态转变的空间动态，要么无法捕捉与多个细胞状态和转变路径相关的不同动态。在此，我们提出了空间转变张量（STT），这是一种通过多尺度动态模型使用信使 RNA 剪接和空间转录组学来表征空间中多稳态的方法。通过学习四维转变张量和空间约束的随机游走，STT 通过细胞之间的短时间局部张量流线和吸引子之间的长期转变路径重构了细胞状态特异性动态和空间状态转变。在多种技术下对上皮 - 间充质转变、血液发育、空间分辨的小鼠大脑和鸡心脏发育等多个转录组数据集的基准测试和应用表明，STT 能够恢复现有方法未能发现的细胞状态特异性动态及其相关基因。总体而言，STT 提供了单细胞转录组数据在多个时空尺度上的一致多尺度描述。</p>
<h3 id="_2">主文<a class="headerlink" href="#_2" title="Permanent link">&para;</a></h3>
<p>单细胞基因表达谱技术的进步为解析细胞命运决策提供了前所未有的分辨率。类似性或低维流形上的距离等指标应用于单细胞 RNA 测序（scRNA-seq）数据，以推断伪时间排序 1,2、网络抽象 3 或细胞随机游走分析 4,5 等动态特性。利用未剪接和已剪接计数，RNA 速度方法 6,7 显式地建模信使 RNA（mRNA）的动态，将细胞未来的剪接状态投射到 scRNA-seq 数据上，以揭示细胞命运决定的方向性 8，并改进轨迹推断 9,10,11、低维嵌入 12,13 和基因调控网络推断 14,15。</p>
<p>空间转录组学在单个细胞或小组细胞的点上测量额外的空间信息，允许分析空间中的异质细胞状态 16,17。为了推断空间转录组学中的时间动态，SpaceFlow18 使用邻近信息约束细胞嵌入和伪时间排序以实现空间一致性。SIRV19 开发了一种空间解析 RNA 速度方法，通过使用参考 scRNA-seq 数据来改进未剪接和已剪接 mRNA 的估计，以丰富空间转录组学基因表达矩阵。</p>
<p>尽管 RNA 速度已被广泛使用，但在重建稳健的时空动态方面仍然存在基本挑战 20。例如，当前模型无法捕捉复杂空间组织中的多谱系或多个亚稳态 21,22,23，因为由于非线性基因调控或多细胞信号传导，剪接和未剪接的转录水平可能会出现偏离。此外，mRNA 剪接的时间尺度在几分钟或几小时内 24,25，在此期间，当前的 RNA 速度模型会收敛到一个全局平衡，而细胞状态的转变可能跨越数天到数周（例如，在造血过程中 8,20,25）。尽管可以使用细胞特异性的基因表达速率来适应连续的细胞命运承诺过程 25,26，但需要额外的测量，如代谢标记 27,28,29，这在空间转录组学中很难获得。最后，当前主要的 RNA 速度方法仅关注剪接计数的速度，忽略了与基因调控密切相关的未剪接计数的速度 15，这可能提供关于某些细胞状态“吸引力”的进一步信息。</p>
<p>多尺度细胞吸引子理论 30,31,32,33,34,35 为跨不同时间尺度和分辨率建模动态提供了一种自然工具，并考虑到多稳定状态。在这样的理论中，基因表达的时间变化及其相互调控被建模为一组微分方程组成的动态系统。稳定的细胞类型对应于在基因调控的轻微扰动下动态系统的多个局部稳定固定点（即多稳定状态），其中表达的细胞状态被“困住”，而高可塑性的过渡细胞被建模为系统的“鞍点”，使细胞可以通过某些方向进行状态转换。使用这种方法，MuTrans5 在不同尺度上粗粒化 scRNA-seq 数据以识别吸引子和鞍点，允许描述细胞在吸引子周围的短时间波动，同时捕捉细胞在多个吸引子之间的长时间尺度转换，鞍点在其中。MuTrans 中的高斯核将其范围限制在平衡和遍历系统 4,5。对于非平衡系统，使用 RNA 速度作为输入，CellRank8 使用速度核构建细胞随机游走，随后进行粗粒化分析，Dynamo25 通过使用连续函数拟合离散 RNA 速度进行吸引几何和转换分析。然而，在这些方法中，线性 RNA 速度模型与数据中继承的多稳定吸引子的存在不兼容，导致转变速度和下游分析之间的不一致。此外，这些方法不能直接用于空间转录组数据。</p>
<p>在此，我们提出了一种空间转变张量（STT）方法，使用未剪接和已剪接 mRNA 计数在空间转录组数据中重建细胞吸引子，以允许量化空间吸引子之间的转换路径以及单个过渡细胞的分析。与具有一个全局平衡的线性 RNA 速度模型不同（图 1a），STT 假设在联合未剪接（U）- 剪接（S）计数空间中存在多个吸引子，细胞在吸引子盆地之间进行转换（图 1a,b）。构建跨细胞、基因、剪接状态和吸引子的四维转变张量，每个吸引子盆地都关联有吸引子特异性量（图 1b）。通过迭代细化张量估计并分解张量引导和空间约束的细胞随机游走（图 1c–e,g），STT 连接了局部基因表达和剪接动态与吸引子之间的全局状态转换之间的尺度。此外，STT 对与多稳定表达模式最相关的基因进行排序，并分类具有类似 STT 特性的途径（图 1g）。通过研究非空间和空间数据集，我们展示了 STT 在揭示细胞多稳定吸引子及其在不同时空尺度上发生的转换特性方面的独特能力。</p>
<p><img alt="" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-024-02266-x/MediaObjects/41592_2024_2266_Fig1_HTML.png?as=webp" /></p>
<blockquote>
<p>a. 比较 RNA 速度（线性和单一平衡）与 STT 张量模型（多稳态和多个吸引子）。
b. 通过平均细胞在不同吸引子中的归属，定义过渡张量和诱导的 RNA 速度。
c–f. STT 的工作流程。
c. 输入的 U 和 S 计数矩阵。
d,e. 过渡张量的动力学参数估计（d）与动力学分解和粗粒化（e）之间的迭代方案。
f. STT 的输出结果
g. 使用 STT 分析空间转录组学数据，其中基于空间细胞坐标的空间相似核与张量诱导和基因表达诱导的核相结合，以推断细胞在吸引子中的归属。在路径相似性图中，Dim.表示降维后的坐标。</p>
</blockquote>
<h3 id="_3">结果<a class="headerlink" href="#_3" title="Permanent link">&para;</a></h3>
<h4 id="stt">STT 概述<a class="headerlink" href="#stt" title="Permanent link">&para;</a></h4>
<p>STT 的输入包括单细胞基因表达矩阵的 <span class="arithmatex">\(S\)</span> 和 <span class="arithmatex">\(U\)</span> 计数（图 1c），以及细胞注释（或归属），这些作为细胞状态的初始猜测。此外，还需要每个细胞（或位置）的空间坐标，以用于空间转录组数据。通过参数估计和动力学分解之间的迭代，STT 构建了一个以吸引子为单位的速度张量，称为过渡张量，其形状为 <span class="arithmatex">\(\mathbb{R}^{N_{\mathrm{C}} \times 2 \times K \times N_{\mathrm{G}}}\)</span>，其中 <span class="arithmatex">\(N_{\mathrm{C}}\)</span> 表示细胞数量，<span class="arithmatex">\(N_{\mathrm{G}}\)</span> 表示基因数量，<span class="arithmatex">\(K\)</span> 表示吸引子数量。张量动力学的其他量，包括细胞在吸引子中的归属、过渡概率和过渡路径，随后在此构建中获得（方法部分）。</p>
<p>STT 使用以下基因表达和剪接动力学的随机模型</p>
<div class="arithmatex">\[
\left\{\begin{aligned}
&amp; \mathrm{d} U_i=\left(f_i\left(t, S_1, \ldots, S_{N_G}\right)-\beta_i U_i\right) \mathrm{d} t+\sigma_i \mathrm{~d} W_{i, t} \\
&amp; \mathrm{~d} S_i=\left(\beta_i U_i-\gamma_i S_i\right) \mathrm{d} t+\sigma_i \mathrm{~d} Z_{i, t}
\end{aligned}\right.
\]</div>
<p>其中 <span class="arithmatex">\(U_i\)</span> 和 <span class="arithmatex">\(S_i\)</span> 分别是基因 <span class="arithmatex">\(i\)</span> 的未剪接和剪接计数。非线性函数 <span class="arithmatex">\(f_i\)</span> <span class="arithmatex">\(\left(t, S_1, \ldots, S_{N_{\mathrm{G}}}\right)\)</span> 模拟了其他基因如何调控基因 <span class="arithmatex">\(i\)</span> 的生产速率。该系统可以具有多个固定点或吸引子，代表不同的细胞状态。参数 <span class="arithmatex">\(\beta_i\)</span> 表示 mRNA 剪接速率，<span class="arithmatex">\(\gamma_i\)</span> 是剪接 mRNA 降解速率。独立的维纳过程项 <span class="arithmatex">\(W_{i, t}\)</span> 和 <span class="arithmatex">\(Z_{i, t}\)</span> 代表基因表达中的噪声。这种随机性可能会在比剪接动力学更长的时间尺度上诱导多稳态吸引子之间的噪声诱导细胞状态转变。</p>
<p>当大多数细胞位于对应不同细胞状态的多个吸引子盆地内，只有少数细胞跨越鞍点进行过渡（这是对细胞分布的自然假设）时，未剪接 mRNA 的生产项可以展开并近似为其线性展开，从而引入吸引子相关的 mRNA 转录速率（图 1d 和方法部分）。这种展开允许对参数进行稳健估计，并初始化每个细胞的吸引子速度分配，我们称之为过渡张量（图 1d 和方法部分）。</p>
<p>通过构建内积速度核（图 1d，方法部分和补充说明 1），张量提供了一种与连续随机微分方程（SDE）渐近一致的细胞随机游走描述（即方程（1））。结合基因表达相似性的高斯核和细胞空间坐标（图 1g 和方法部分），构建的细胞随机游走为每个吸引子中的细胞提供了一致的速度、过渡方向和相似的基因表达。此外，构建的随机游走鼓励细胞更可能向物理空间中其他空间上相邻的细胞过渡。通过在吸引子水平上粗粒化和分解随机游走，然后获得细胞在不同吸引子中的归属函数（图 1e 和方法部分）。在张量模型构建和随机游走分解之间的每次迭代中，更新的归属函数通过引入吸引子不确定性来改进方程（1）中的参数估计（方法部分）。在迭代过程中，识别出在 U-S 空间中与吸引子特性最一致的基因（方法部分）。包括一个监控模块，具有正则化和早停策略，通过用户控制可以提高迭代的稳健性（方法部分）。最后，将描述吸引子细节的张量流线以及描绘长时间过渡的粗粒化过渡路径投影在低维动力学流形上，以显示细胞状态的转变（图 1f 和方法部分）。</p>
<h4 id="stt_1">多稳态细胞状态恢复的 STT 基准测试<a class="headerlink" href="#stt_1" title="Permanent link">&para;</a></h4>
<p>我们首先应用 STT 分析了基于模拟多稳态系统的两个合成数据集。在双稳态开关电路中，STT 中吸引子平均速度的流线比 RNA 速度和其他方法的流线展示了更清晰的两个吸引子结构（图 2a 和补充图 1）。尽管由 scVelo7 和 UniTVelo36 计算的 RNA 速度流线倾向于从吸引子位置发散，STT 流线则向吸引子收敛，从而提供了更易解释的开关景观表示（图 2a 和补充图 1）。此外，STT 计算出一个熵值来区分固定点附近的稳定细胞和跨越鞍点的过渡细胞（补充图 1）。如流线显示的过渡张量的两个组成部分所示（补充图 1），只有当未剪接和剪接的数量同时考虑时，才能揭示出吸引子盆地。尽管剪接张量与标准的 RNA 速度一致（图 2a），显示了吸引子之间的过渡，未剪接张量则自然引入了一种“吸引力”，将细胞“拉”向每个吸引子的中心，而与 cellDancer37 的流线相比，后者细胞被吸引到吸引子内的“端点”（补充图 1）。未剪接计数为 STT 提供了一个对细胞状态吸引子的“吸引”水平的测量。为了进一步验证 STT 的准确性，我们比较了 STT 未剪接或剪接张量成分与模型的真实速度之间的余弦相似性，发现 STT 在估计剪接和未剪接速度方面均排名靠前（图 2b）。此外，STT 在对数据集进行子抽样时表现出了良好的稳健性（补充图 1）。</p>
<p><img alt="" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-024-02266-x/MediaObjects/41592_2024_2266_Fig2_HTML.png?as=webp" /></p>
<blockquote>
<p>a. 比较 STT 与其他方法在开关数据集上的流线。STT 中的细胞根据吸引子进行着色，而 scVelo 和 UniTVelo 中的细胞则根据 Leiden 聚类结果进行着色。STT、scVelo 和真实结果嵌入在联合剪接和未剪接计数的 PCA 中，UniTVelo 的结果绘制在剪接计数的坐标上。
b. 计算速度与真实值在不同方法中余弦相似性的所有细胞（n = 10,010）的箱线图。中央箱体表示四分位范围，从第 25 百分位数（下边界）到第 75 百分位数（上边界），箱体内的水平线表示中位数（第 50 百分位数）。须线延伸到下四分位数和上四分位数 1.5 倍四分位范围内的值。点表示离群值。
c,d. 比较 STT 与其他方法在合成 EMT 电路数据集上的流线。
c. 细胞根据 STT 的吸引子分配进行着色，低维嵌入是基于联合剪接和未剪接计数的 UMAP。流线使用吸引子上的平均速度进行可视化。
d. 细胞根据 Leiden 聚类结果进行着色，低维嵌入是仅基于剪接计数的 UMAP。流线使用 RNA 速度进行可视化。</p>
</blockquote>
<p>接下来，我们分析了在上皮 - 间质转化（EMT）过程中模拟的基因调控电路，其中在某些参数范围内，可能共存三个吸引子，分别表示上皮状态（E）、间质状态（M）和中间细胞状态（ICS）（方法部分）。与 scVelo 计算的 RNA 速度相比（图 2），STT 平均速度（图 2c）清晰地恢复了这三个吸引子。总体而言，STT 能够重建单细胞基因表达数据集中复杂的多稳态细节。</p>
<h4 id="stt-ics">STT 突出了命运决策中的 iCS<a class="headerlink" href="#stt-ics" title="Permanent link">&para;</a></h4>
<p>接下来，我们分析了人类肺 A549 细胞系在 EMT 诱导实验中的 scRNA-seq 数据，包括从 TGFB1 处理后的前 7 天收集的一系列时间点快照。STT 识别出三个吸引子，分别为 E、ICS 和 M，与数据收集时间点的顺序一致（图 3a、b 和补充图 2）。此外，ICS 吸引子附近的细胞，主要在诱导后 8 小时或 1 天收集（图 3b），具有较高的熵值（图 3c），表明这种状态比上皮（第 0 天和 8 小时）和间质（第 3 天后）状态更具可塑性。这与在癌症中提出的中间上皮和/或间质状态的表型可塑性很好地吻合。</p>
<p><img alt="" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-024-02266-x/MediaObjects/41592_2024_2266_Fig3_HTML.png?as=webp" /></p>
<blockquote>
<p>a. EMT 的全局过渡路径分析。细胞嵌入在构建的动力学流形的过渡坐标（trans. coord.）中，数字表示过渡流的比例。细胞按 STT 吸引子着色。
b. 过渡坐标中的细胞按收集时间点着色
c. 不同吸引子中细胞归属熵的小提琴图。
d. 使用多稳态核诱导的随机游走计算的细胞吸收到不同吸引子的吸收概率。
e. 在 EMT 中与吸引子的多稳态一致的主要基因。
f. 各种过渡张量成分的流线，包括吸引子平均张量和吸引子特定张量。低维嵌入是剪接和未剪接计数的 UMAP。左图中，细胞按吸引子分配着色。右图中，细胞按其在每个吸引子中的归属着色，仅显示归属度大于 0.2 的细胞的张量。</p>
</blockquote>
<p>使用过渡向量预测在上皮 - 间质景观中连接吸引子的过渡路径，我们发现从 E 到 M 的过渡概率流量始终通过 ICS（图 3a）。换句话说，经历 EMT 的上皮细胞不会直接转换为间质状态，而是首先获得中间特征。未剪接和剪接计数通常表现出吸引子的多稳态性（图 3e 和补充图 2、3）。具有高多稳态评分的基因在各种吸引子中的未剪接和剪接计数中表现出不同的表达水平，并在 E-ICS-M 过渡过程中显示出逐渐变化。虽然像 ITGA11 这样高排名的多稳态基因在差异基因表达分析中并未显著地作为吸引子的最高评分标志基因检测到（补充图 3），但它们在促进 EMT 过渡和肿瘤进展中被发现具有重要作用。尽管剪接动力学的张量流线显示了从 E 到 M 通过 ICS 的总体方向，这也与 UniTVelo 的结果一致（补充图 2），但未剪接计数以及 STT 和 cellDancer 在联合 U-S 空间中预测的基因表达动力学都表明细胞在 EMT 期间被“吸引”到 ICS 盆地（图 3f 和补充图 2）。这也与基于张量诱导的多稳态核的 CellRank 吸收概率分析一致（图 3d）。总之，张量成分及其全局过渡路径分析突出了 ICS 作为 EMT 期间的一个独特吸引子盆地，作为中心状态。</p>
<p>此外，我们将 STT 应用于血液和胰腺发育数据集，发现其能够解决复杂的状态过渡，其多稳态张量核与 CellRank 分析一致（补充图 4 和图 5）。</p>
<h4 id="stt_2">STT 识别空间吸引子和路径相似性<a class="headerlink" href="#stt_2" title="Permanent link">&para;</a></h4>
<p>我们接下来将 STT 应用于小鼠脑发育的 HybISS 空间数据集。为了丰富未剪接和剪接计数以便更好地进行张量估计，我们使用 SIRV19 算法推算了 E10 和 E11 在 40μm 处的原始空间数据切片中的一个。与仅基于细胞相似性的聚类相比（补充图 6），STT 识别出与不同细胞状态的空间位置一致的吸引子（图 4a）和原始出版物中的脑区注释（图 4b）：同一吸引子内的细胞往往具有相似的空间坐标，并属于相同的区域。此外，发现细胞分配对空间扩散核的权重（补充图 6）、吸引子初始化（补充图 7）、多稳态基因过滤（补充图 8）和吸引子数量（补充图 9）具有稳健性。前脑和后脑吸引子的局部过渡张量（图 4c）与 UniTVelo 分析一致（补充图 6）。</p>
<p><img alt="" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-024-02266-x/MediaObjects/41592_2024_2266_Fig4_HTML.png?as=webp" /></p>
<blockquote>
<p>a,b. 数据的空间注释和 STT 检测到的吸引子，细胞按不同类别着色：吸引子（a）和区域（b）。
c. 在特定吸引子 6 和 3 中的局部过渡张量流线。细胞按其在相应吸引子中的归属着色。
d. 不同 KEGG 通路之间的过渡张量相似性。左侧显示了小鼠脑发育空间动态中相似生物通路的聚类二维嵌入，其中来自各种通路的平均张量流线展示了不同的过渡动态。具有至少三个与 STT 多稳态基因重叠的基因的通路显示在低维嵌入中。右侧显示了来自不同聚类的特定通路的流线，细胞嵌入在空间坐标中。</p>
</blockquote>
<p>为了评估张量流线的生物学意义，我们进行了通路特异性分析，以评估与细胞状态转变和通路调控相关的功能（图 4d）。我们使用了京都基因与基因组百科全书（KEGG）知识数据库，并基于多稳态基因的张量相关性计算各通路之间的相似性（方法部分）。确实，通路特异性张量展示了不同的吸引子动态。基于张量相关性的潜在嵌入和通路聚类（图 4d）揭示了发育过程中通路之间空间状态转变的功能相似性。已知在胚胎发生过程中存在交叉对话并协同作用的 TGF-beta 和 WNT 通路在潜在嵌入中来自不同的聚类，并且它们的张量流线方向相反，特别是在中脑和前脑吸引子中（图 4d）。在脑发育中两个其他重要的通路，Hippo 和甲状腺激素信号通路，也来自不同的通路张量聚类，显示出在中脑和前脑区域相反的流线（图 4d）。总体而言，STT 提供了细胞状态的空间组织和在发育过程中调控状态转变的通路之间关系的动态信息。</p>
<h4 id="stt_3">STT 揭示了鸡心的空间吸引子和谱系<a class="headerlink" href="#stt_3" title="Permanent link">&para;</a></h4>
<p>我们将 STT 应用于通过 10X Visium 技术测量的空间解析鸡心数据。我们的分析重点是数据集在第 14 天的最后时间点，此时四腔室的发育已经完成，心脏发生和明确的空间边界已经完成。</p>
<p>使用 SIRV 推测的未剪接和剪接计数，STT 识别出五个空间解析的吸引子（图 5a 和补充图 10）。其中，吸引子 2 与原始研究中的“瓣膜”区域一致，主要由成纤维细胞组成（图 5a,b,d,e）。吸引子 0 主要由来自右心室区域的细胞组成（图 5e）。吸引子 1 主要位于“心房”区域（图 5e），由红细胞组成。而其余的吸引子（3、4）分布在几个连接区域，它们都包括注释表型为心肌细胞的细胞（图 5a,b,d,e）。动力流形揭示了这些离散吸引子（图 5c）与不同的细胞谱系有关。吸引子 1 和 2 包含空间定位的成纤维细胞和红细胞谱系，都展示了张量流线中的“吸引力”（图 5f）。相比之下，吸引子 3 和 4（均含有心肌细胞）内的张量流线表明它们在空间上的短暂性，并显示出向心房区域过渡的趋势，这在未剪接成分的“吸引”中也可以观察到。这可能部分解释了心房中另一组心肌细胞的存在（图 5b,d）。总体而言，与空间区域或细胞类型注释的一致性表明 STT 有能力解析空间解析的吸引子。</p>
<p><img alt="" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-024-02266-x/MediaObjects/41592_2024_2266_Fig5_HTML.png?as=webp" /></p>
<blockquote>
<p>a,b. 分析数据的空间点，其中空间点按 STT 检测到的吸引子区域（a）或原始研究中的注释（b）进行着色。
c. 构建的数据动力学景观，其中空间点按吸引子进行着色。
d. 按原始研究中的细胞类型注释进行着色的空间点。
e. Sankey 图显示 STT 吸引子（左）与空间区域注释（右）之间的关系。链接的宽度表示同时共享连接的吸引子标签和区域注释标签的细胞数量。
f. 特定吸引子 1、2、3 和 4 中的局部过渡张量流线。细胞按其在相应吸引子中的归属进行着色。</p>
</blockquote>
<h4 id="stt_4">STT 阐明了区域特异性的空间吸引子及其稳定性<a class="headerlink" href="#stt_4" title="Permanent link">&para;</a></h4>
<p>我们接下来分析了使用 60 大小的 bin 处理的高分辨率 Stereo-seq 小鼠成年冠状脑半球数据集，该数据集揭示了具有各种生物功能的神经细胞的复杂区域。直接应用 STT 显示了多个区域特异性的空间吸引子，这些吸引子与脑区的功能注释非常一致（图 6a,b）。张量的收敛流线（图 6c）表明，基因表达动力学的多稳态性在诸如皮层子区（吸引子 4）和纹状体背侧区域（吸引子 10）等区域得到了很好的保持。张量的所有成分在丘脑区域（吸引子 8）向外流动（图 6c），表明其相对较高的可塑性。基于它们的张量动力学的通路嵌入显示，已知的相互作用通路如 cGMP-PKG 和钙信号通路共享相似的张量动力学（图 6d）。它还表明 cGMP-PKG 不同于催产素通路，其中流线表明主要差异发生在杏仁核区域（图 6d）。总体而言，结果表明 STT 能够通过吸引子分析发现空间区域并量化其稳定性，即使在成熟组织中也是如此。</p>
<p><img alt="" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-024-02266-x/MediaObjects/41592_2024_2266_Fig6_HTML.png?as=webp" /></p>
<blockquote>
<p>a,b. 细胞的空间位置按 STT 吸引子（a）和原始研究中的注释（b）进行着色。
c. 吸引子 4、8 和 10 中的局部张量流线。
d. 通路动力学的二维嵌入（上）和两个特定通路的平均张量流线（下），细胞按吸引子着色并嵌入空间坐标中。具有至少八个与 STT 多稳态基因重叠的基因的通路显示在二维嵌入中。</p>
</blockquote>
<h3 id="_4">讨论<a class="headerlink" href="#_4" title="Permanent link">&para;</a></h3>
<p>量化和建模未剪接和剪接计数之间的相对丰度，使得能够采用有效的机械方法来剖析来自 scRNA-seq 数据集的细胞状态转变。为了连接基因表达、mRNA 剪接和细胞谱系动力学之间的不同时间尺度，并研究这些状态的潜在吸引子，我们开发了 STT 用于：(1) 构建吸引子特异的过渡张量，(2) 分析概率过渡路径和过渡细胞，(3) 推断导致细胞状态多稳态的基因。这是通过在 (1) 过渡张量模型中的参数推断和 (2) 张量诱导的随机动力系统的多尺度分析之间的迭代计算过程完成的。</p>
<p>与 RNA 速度模型相比，STT 在揭示基因表达和剪接动力学的潜在吸引子以及量化它们之间的过渡方面具有独特性。通过假设多稳态性，STT 对初始状态规范或隐藏时间校正具有鲁棒性。细胞归属函数在估计过渡张量时量化了过渡细胞，自然地桥接了下游的多尺度动力学分析。</p>
<p>为了识别过渡细胞和推断过渡路径，STT 利用计算的过渡张量，而不是直接使用 RNA 速度，如 CellRank 或 Dynamo。在下游分析中，发现多稳态过渡张量与吸引子假设更兼容。STT 在张量构建和动力学解剖之间的迭代方案被发现更能确保这种自洽性。然而，由于吸引子假设不考虑振荡动力学，STT 需要改进以捕捉具有强细胞周期效应的数据集的非平衡特征。</p>
<p>基于速度核的细胞随机游走从过渡张量中推导出来，是连接 STT 中张量推断和动力学分解模块的关键，允许在不同尺度上更好地连接动力学。理论分析表明，不同的速度核选择会导致各种形式的常微分方程或随机微分方程的连续极限。在 STT 中，使用内积核构建细胞随机游走，证明其与基因表达的随机化学朗之万模型一致，而余弦核则正确恢复了速度场的方向性，用于可视化吸引子内的局部流线。除了微分方程模型之外，将来可能有趣的是在 RNA 速度的化学主方程框架中制定 STT。</p>
<p>作为一种基于机械模型的方法，STT 在多个方面可能会有所改进。与其在方程 (1) 中使用吸引子特异的非线性基因表达速率函数的零阶近似，不如考虑最近提出的用于基因调控网络推断的高阶基因相互作用。可以将包括单细胞表观基因组学或蛋白质组学数据在内的多模态信息纳入多稳态动力系统，以增强过渡张量计算。在多稳态模型中自动检测根状态和目标状态始终是一个挑战，先前的知识或了解与细胞分化能力相关的属性可能会有所帮助。</p>
<p>总体而言，STT 提供了一种统一的方法，通过跨越不同时间尺度和组织区域的过程，从单细胞数据集中提取时空信息。我们的方法允许对组织时空结构进行多尺度描述，连接基因表达和剪接的微观动力学，以及新兴吸引子之间的细胞状态转变的宏观动力学。</p>
<h3 id="_5">方法<a class="headerlink" href="#_5" title="Permanent link">&para;</a></h3>
<h4 id="_6">基因表达和剪接模型中的多稳态性<a class="headerlink" href="#_6" title="Permanent link">&para;</a></h4>
<p>我们使用一个简单的动力学模型，并在每个稳态附近采用不同的参数来近似基因 <span class="arithmatex">\(i\)</span> 的 mRNA 剪接动力学：</p>
<div class="arithmatex">\[
\left\{\begin{array}{l}
\frac{\mathrm{d} U_i}{\mathrm{~d} t}=\alpha_{c, i}-\beta_i U_i \\
\frac{\mathrm{d} S_i}{\mathrm{~d} t}=\beta_i U_i-\gamma_i S_i
\end{array}\right.
\]</div>
<p>这里，<span class="arithmatex">\(\alpha_{c, i}\)</span> 是吸引子 <span class="arithmatex">\(c\)</span> 中状态依赖的未剪接 mRNA 转录速率，<span class="arithmatex">\(\beta_i\)</span> 是 mRNA 剪接速率，<span class="arithmatex">\(\gamma_i\)</span> 是 mRNA 降解速率。假设系统接近稳态，我们有 <span class="arithmatex">\(\epsilon_i=\alpha_{c, i}-\beta_i U_i, \eta_i=\beta_i U_i-\gamma_i S_i\)</span>，其中 <span class="arithmatex">\(\epsilon_i\)</span> 和 <span class="arithmatex">\(\eta_i\)</span> 是独立且同分布的均值为零的高斯变量。由于表达的不可变性</p>
<div class="arithmatex">\[
\min _{\alpha_c, \beta} \sum_{c=1}^K \sum_{k=1}^{N_{\mathrm{C}}}\left(\alpha_c-\beta U_k\right)^2 1_{k \in \Omega_c}+\sum_{k=1}^{N_{\mathrm{C}}}\left(\beta U_k-S_k\right)^2 .
\]</div>
<p>由于不同基因的参数是独立估计的，为简化符号，这里我们省略了基因下标 <span class="arithmatex">\(i\)</span>，并引入下标 <span class="arithmatex">\(k\)</span> 表示细胞索引。吸引子指示函数 <span class="arithmatex">\(1_{k \in \Omega_c}\)</span> 使用用户提供的细胞标签或标准 Leiden 或 Louvain 聚类算法输出进行初始化，并在迭代中使用归属函数进行更新（如下所述）。估计得出的解为：</p>
<div class="arithmatex">\[
\alpha_c^{(*)}=m_c \beta^{(*)}, \beta^{(*)}=\frac{\sum_{k=1}^{N_{\mathrm{C}}} U_k S_k}{\sum_{k=1}^{N_{\mathrm{C}}}\left(U_k^2+\sum_{c=1}^K\left(U_k-m_c\right)^2 1_{k \in \Omega_c}\right)}
\]</div>
<p>其中 <span class="arithmatex">\(m_c=\frac{\sum_{k=1}^{N_{\mathrm{C}}} U_k 1_{k \in \Omega_c}}{N_c}\)</span>。与标准 RNA 速度模型中的稳态参数估计相比，多稳态模型中的剪接速率参数 <span class="arithmatex">\(\beta\)</span> 不仅是吸引子类型特异的，还取决于未剪接和剪接计数。</p>
<p>对于每个细胞 <span class="arithmatex">\(k\)</span>，其计数为 <span class="arithmatex">\(\left(U_k, S_k\right)\)</span>，其相对于每个吸引子 <span class="arithmatex">\(c\)</span> 的速度定义为 <span class="arithmatex">\(v_{k, u, c}=\alpha_c^{(*)}-\beta^{(*)} U_k, v_{k, s, c}=\beta^{(*)} U_k-S_k\)</span>，其中下标 <span class="arithmatex">\(u\)</span> 和 <span class="arithmatex">\(s\)</span> 分别对应未剪接和剪接计数。该估计对于每个基因重复，因此得到一个四维过渡张量 <span class="arithmatex">\(v_{k, l, c, g} \in \mathbb{R}^{N_{\mathrm{C}} \times 2 \times K \times N_{\mathrm{G}}}\)</span>。</p>
<p>基于张量和空间约束的过渡动力学
接下来，STT 基于计算的张量、基因表达相似性和空间坐标（如果有）构建个体细胞之间的马尔可夫链过渡概率（图 1g）。过渡概率由三个部分构成：<span class="arithmatex">\(P=w_1 P^{\mathrm{v}}+w_2 P^{\mathrm{c}}+\left(1-w_1-w_2\right) P^{\mathrm{s}}\)</span>，其中 <span class="arithmatex">\(P^{\mathrm{v}}, P^{\mathrm{c}}\)</span> 和 <span class="arithmatex">\(P^{\mathrm{s}}\)</span> 分别是由速度、相似性和空间核引导的过渡概率。这里 <span class="arithmatex">\(w_1\)</span> 和 <span class="arithmatex">\(w_2\)</span> 是算法的超参数，用于平衡张量动力学、基因表达相似性和空间邻近性的不同模态的影响。它们对输出的影响已在补充图 6 中测试。</p>
<p>为了构建 <span class="arithmatex">\(P^{\mathrm{v}}\)</span>，我们首先通过沿吸引子维度取平均值将吸引子特异张量转换为吸引子无关的速度 <span class="arithmatex">\(V\)</span>：</p>
<div class="arithmatex">\[
V_{k, u, g}=\sum_c \rho_{k, c} v_{k, u, c, g}, V_{k, s, g}=\sum_c \rho_{k, c} v_{k, s, c, g}
\]</div>
<p>这里 <span class="arithmatex">\(\rho_{k, c}\)</span> 表示细胞 <span class="arithmatex">\(k\)</span> 在吸引子 <span class="arithmatex">\(c\)</span> 中的归属函数。位于吸引子盆地固定点附近的稳定细胞 <span class="arithmatex">\(j\)</span> 使得 <span class="arithmatex">\(\rho_{j, d}=1\)</span>，而位于鞍点附近的过渡细胞 <span class="arithmatex">\(k\)</span> 在 <span class="arithmatex">\(\rho_{k}\)</span> 中有多个正分量，指向细胞可以过渡到的吸引子。</p>
<p>这里 <span class="arithmatex">\(\rho_{k, c}\)</span> 表示细胞 <span class="arithmatex">\(k\)</span> 在吸引子 <span class="arithmatex">\(c\)</span> 中的归属函数。位于吸引子盆地固定点附近的稳定细胞 <span class="arithmatex">\(j\)</span> 使得 <span class="arithmatex">\(\rho_{j, d}=1\)</span>，而位于鞍点附近的过渡细胞 <span class="arithmatex">\(k\)</span> 在 <span class="arithmatex">\(\rho_{k}\)</span> 中有多个正分量，指向细胞可以过渡到的吸引子。</p>
<p>计算了张量后，我们使用内积核 <span class="arithmatex">\({}^{49}\)</span>（补充说明 1）构建由速度引导的过渡概率 <span class="arithmatex">\(P^{\mathrm{v}}\)</span>。从细胞 <span class="arithmatex">\(k\)</span> 到 <span class="arithmatex">\(l\)</span> 的过渡倾向权重为 <span class="arithmatex">\(w_{k l}=\exp \left(V_{k, u}^T \Delta U_{k l}+V_{k, s}^T \Delta S_{k l}\right)\)</span>，其中 <span class="arithmatex">\(\Delta U_{k l}=U_l-U_k\)</span> 和 <span class="arithmatex">\(\Delta S_{k l}=S_l-S_k\)</span>。由这种核引导的随机游走与方程（1）的 SDE 模型一致（参考文献 <span class="arithmatex">\({}^{49}\)</span> 和补充说明 1）。由细胞相似性引导的过渡概率 <span class="arithmatex">\(P^{\mathrm{c}}\)</span> 是从基于基因表达计数的扩散图的高斯核构建的 <span class="arithmatex">\({}^{2}\)</span>。最后，空间约束过渡概率 <span class="arithmatex">\(P^{\mathrm{s}}\)</span> 是从空间位置坐标的高斯核构建的，使得具有相似空间位置的细胞更有可能在彼此之间进行过渡。因此，这些细胞更有可能被分配到相同的吸引子盆地。</p>
<p>为了计算吸引子中的归属函数，我们使用 GPPCA <span class="arithmatex">\({}^{57}\)</span> 算法分解构建的随机游走过渡概率矩阵 <span class="arithmatex">\(P\)</span>，并对非平衡马尔可夫链进行粗粒化，获得 <span class="arithmatex">\(\rho_{k, c}\)</span>。该算法允许对细胞随机游走的非平衡过渡概率矩阵进行因式分解和“粗粒化”，这对于大多数由速度引导的动力学来说是真实的，从而获得数据中的吸引子以及细胞在每个吸引子中的相关位置（即归属）。使用 GPPCA 算法同时获得吸引子层面的粗粒化（cg）过渡概率矩阵 <span class="arithmatex">\(P_{\mathrm{cg}}^{K \times K}\)</span>（<span class="arithmatex">\(K\)</span> 是吸引子的总数）。给定细胞的归属函数，其过渡熵可以定义为 <span class="arithmatex">\(\varepsilon_i=-\sum_{c=1}^K \rho_{i, c} \ln \rho_{i, c}\)</span>。较大的熵值表明细胞在吸引子之间进行过渡的倾向较高。</p>
<h4 id="_7">参数估计和吸引子归属量化的迭代方案<a class="headerlink" href="#_7" title="Permanent link">&para;</a></h4>
<p>在获得归属函数后，更新张量模型的参数以纳入细胞在吸引子中位置的不确定性。我们定义损失函数</p>
<div class="arithmatex">\[
\mathcal{J}\left(\alpha_c, \beta, \rho_{k, c}\right)=\sum_{c=1}^K \sum_{k=1}^{N_{\mathrm{C}}}\left(\alpha_c-\beta U_k\right)^2 \rho_{k, c}+\sum_{k=1}^{N_{\mathrm{C}}}\left(\beta U_k-S_k\right)^2+\lambda \sum_{c=1}^{N_{\mathrm{C}}} \alpha_c^2+\lambda \beta^2,
\]</div>
<p>其中 <span class="arithmatex">\(\lambda\)</span> 表示动力学参数正则化项的强度。直观上，吸引子 <span class="arithmatex">\(c\)</span> 中的“稳定细胞”在回归损失函数中具有较大的权重值，因为对稳态的置信水平较高。我们解析地求解优化器</p>
<div class="arithmatex">\[
\alpha_c^{(*)}=m_c \beta^{(*)}, \beta^{(*)}=\frac{\sum_{k=1}^{N_C} U_k S_k}{\sum_{k=1}^{N_C}\left(U_k^2+\sum_{c=1}^K\left(U_k-m_c\right)^2 \rho_{k, c}\right)+\lambda},
\]</div>
<p>其中 <span class="arithmatex">\(m_c=\frac{\sum_{k=1}^{N_{\mathrm{C}}} U_k \rho_{k, c}}{\sum_{k=1}^{N_{\mathrm{C}}} \rho_{k, c}+\lambda}\)</span>
反过来，使用新优化的参数更新的张量会导致更新的归属函数。在 STT 中，我们采用一个迭代方案来联合更新张量参数和吸引子归属，</p>
<div class="arithmatex">\[
\begin{aligned}
&amp; \alpha_c^{n+1}, \beta^{n+1}=\operatorname{argmin}_{\alpha_c, \beta} \mathcal{J}\left(\alpha_c, \beta, \rho_{k, c}^n\right), \\
&amp; \rho_{k, c}^{n+1}=\text { DynamicalAnalysis }\left(\alpha_c^n, \beta^n\right)
\end{aligned}
\]</div>
<p>其中上标 <span class="arithmatex">\(n\)</span> 表示迭代次数，DynamicalAnalysis 表示更新归属函数的过程。一旦归属函数在一定阈值内不再改进，或者迭代次数超过允许的最大迭代次数，该方案便停止。</p>
<p>为了准确剖析多稳态动力学，我们在每次迭代中基于基因对模型拟合优度进行基因筛选，包括表现出多稳态性的基因。拟合优度指标或基因多稳态评分定义为 <span class="arithmatex">\(1-\frac{J\left(\alpha_c, \beta, \rho_{k, c}\right)}{N_{\mathrm{C}}(\operatorname{Var}(U)+\operatorname{Var}(S))}\)</span>，其中 <span class="arithmatex">\(N_{\mathrm{C}}\)</span> 表示细胞数量。仅使用多稳态评分超过某个阈值的筛选基因的张量来计算速度核，从而更新归属函数。STT 的超参数及其在分析数据集中选择的值见补充表格 1 和 2。</p>
<p>为了稳健控制迭代方案，我们引入了一个监控模块，该模块输出训练集（默认情况下占所有样本大小的 80%）和测试集（占所有样本的 20%）中基因的多稳态评分。训练集用于拟合动力学参数（<span class="arithmatex">\(\left.\alpha_c, \beta\right)\)</span>），并在测试集中计算基因的多稳态评分。监控模块输出多稳态评分和通过阈值的基因数量。根据输出，用户可以交互式地选择（1）修改用于筛选多稳态基因的阈值，（2）调整张量诱导核相对于基因表达相似性或空间核的权重，以提高过渡矩阵的质量，或者（3）决定是否停止迭代，从而促进自适应精度。监控模块的界面示于补充图 1f。我们还在补充表格 3 和补充图 10 中展示了 STT 算法的效率和可扩展性。</p>
<h4 id="_8">初始化迭代<a class="headerlink" href="#_8" title="Permanent link">&para;</a></h4>
<p>为了开始迭代，STT 请求输入现有的聚类结果，通过独热编码创建吸引子归属。推荐使用数据集的先前生物学注释或空间区域分割结果作为输入。当没有此类信息时，用户可以采用聚类算法，如 Leiden 或 Louvain，根据表达计数（仅剪接或剪接和未剪接联合）或细胞的空间位置对细胞进行聚类。STT 对初始化的鲁棒性进行了研究（补充图 Z）。无论何时用户倾向于使用其他聚类方法和/或更系统的分析，STT 都提供了一个选项，可以将用户生成的聚类输出作为 STT 初始化的输入。</p>
<h4 id="_9">动态流形和过渡路径的可视化<a class="headerlink" href="#_9" title="Permanent link">&para;</a></h4>
<p>为了可视化细胞的低维嵌入，STT 使用每个细胞 <span class="arithmatex">\(k\)</span> 的联合状态 <span class="arithmatex">\(x_k=\left(U_k, S_k\right) \in \mathbb{R}^{2 N_{\mathrm{G}}}\)</span> 作为降维算法如主成分分析（PCA）或统一流形近似与投影（UMAP）的输入。为了可视化动态流形，我们将细胞在二维平面中的位置定义为 <span class="arithmatex">\(y_k=\sum_{c=1}^K \rho_{k, c} \mu_c\)</span>，其中 <span class="arithmatex">\(\mu_c\)</span> 是每个吸引子的 PCA 或 UMAP 嵌入的中心，<span class="arithmatex">\(K\)</span> 是吸引子数量。然后，使用期望最大化算法为所有 <span class="arithmatex">\(y_k\)</span> 构建高斯混合密度估计 <span class="arithmatex">\(\mathcal{P}(y)\)</span>，其中 <span class="arithmatex">\(K\)</span> 个成分的初始权重是前一节中推导的吸引子层面、粗粒化随机游走过渡概率矩阵 <span class="arithmatex">\(P\)</span> 的平稳分布。动态流形的表面计算为 <span class="arithmatex">\(\phi(y)=-\ln \mathcal{P}(y)\)</span>。在二维平面中使用线性（PCA）或非线性（UMAP）投影方法与余弦核计算速度 <span class="arithmatex">\(V_k=\left(V_{k, u}, V_{k, s}\right) \in \mathbb{R}^{2 N_{\mathrm{G}}}\)</span> 的流线。给定初始和最终状态，使用过渡路径理论 <span class="arithmatex">\(\frac{58}{}\)</span> 和 PyEMMA 软件包 <span class="arithmatex">\({}^{59}\)</span> 计算过渡路径及其在总过渡概率流中的比例。</p>
<h4 id="_10">数据可用性<a class="headerlink" href="#_10" title="Permanent link">&para;</a></h4>
<p>本文使用的所有数据集均可公开获取。数据集的详细预处理过程在补充说明中描述。合成电路的模拟数据集可在以下网址获取：https://github.com/cliffzhou92/STT/tree/release/data。人类肺 A549 细胞系的 EMT 数据集可在 GSE147405 获取。胰腺数据集（最初在 GSE132188 可用）和成人骨髓数据集（最初在 https://data.humancellatlas.org/explore/projects/091cf39b-01bc-42e5-9437-f419a66c8a45 可用）可以从 scvelo==0.2.4 软件包的内置数据集中下载（https://scvelo.readthedocs.io/en/stable/api.html）。小鼠脑和鸡心的空间数据集以及用于插补的 scRNA-seq 数据集可以从 https://doi.org/10.5281/zenodo.6798658 (ref. 62) 下载。具有未剪接和剪接计数的 Stereo-seq 小鼠脑数据集从 Spateo 软件包下载（https://github.com/aristoteleo/spateo-tutorials, https://www.dropbox.com/s/c5tu4drxda01m0u/mousebrain_bin60.h5ad?dl=0）。KEGG 数据库最初在 Enrichr 网页上可用（https://maayanlab.cloud/Enrichr/#librariesdownloaded），使用 gseapy==1.0.4 下载。用于分析的处理数据集也存储在 https://disk.pku.edu.cn/link/AAD1681DAD531D47699D459BB46C4651D8。</p>
<h4 id="_11">代码可用性<a class="headerlink" href="#_11" title="Permanent link">&para;</a></h4>
<p>STT 作为 Python 包实现，可在以下网址获取：https://github.com/cliffzhou92/STT/tree/release。用于模拟的源代码和重现本文中所有分析的笔记本文件也可在以下网址获取：https://github.com/cliffzhou92/STT/tree/release/example_notebooks。</p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/" class="md-footer__link md-footer__link--prev" aria-label="Previous: ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer
              </div>
            </div>
          </a>
        
        
          
          <a href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/" class="md-footer__link md-footer__link--next" aria-label="Next: Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.code.select", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.instant.prefetch", "navigation.instant.progress", "navigation.path", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "header.autohide"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.525ec568.min.js"></script>
      
        <script src="../mathjax-config.js"></script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../extra.js"></script>
      
        <script src="../js/toggle_sidebar.js"></script>
      
    
  </body>
</html>