
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://localhost:8000/5.5%20%E7%BA%A0%E6%AD%A3%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94%EF%BC%8C%E6%AF%94%E8%BE%83%E5%AF%B9%E7%85%A7%E7%BB%84%E5%92%8C%E5%88%BA%E6%BF%80%E7%BB%84/">
      
      
        <link rel="prev" href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/">
      
      
        <link rel="next" href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/">
      
      
      <link rel="icon" href="../img/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>5.5 纠正批次效应，比较对照组和刺激组 - One Last Kiss</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lxgw-wenkai-webfont@1.1.0/style.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-M0R7V8QGQ7"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-M0R7V8QGQ7",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-M0R7V8QGQ7",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#_1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="One Last Kiss" class="md-header__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            One Last Kiss
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              5.5 纠正批次效应，比较对照组和刺激组
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
    
  
  🏠 Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Mygo/" class="md-tabs__link">
          
  
    
  
  ⛏️ CV and Multimodels

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Session%201/" class="md-tabs__link">
          
  
    
  
  🚀 Re:0 Road to CPath & Visualization

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-tabs__link">
          
  
    
  
  💤 Re:0 Road to scRNA-seq Analysis

        </a>
      </li>
    
  

    
  

    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-tabs__link">
          
  
    
  
  🧬 scRNA Analysis

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-tabs__link">
          
  
    
  
  📐 Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-tabs__link">
          
  
    
  
  💡 Reviews

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="One Last Kiss" class="md-nav__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    One Last Kiss
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href=".." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    🏠 Home
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            🏠 Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⛏️ CV and Multimodels
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ⛏️ CV and Multimodels
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mygo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MyGO: Discrete Modality Information as Fine-Grained Tokens for Multi-modal Knowledge Graph Completion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GMM/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GMM
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../DINO/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    DINO
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mamba/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mamba
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ViT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ViT
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCoV2%2C3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo V2
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%80%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (一)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%20%28%E4%BA%8C%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (二)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%89%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (三)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E5%9B%9B%29%20%28TNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (四)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E7%9F%A5%E8%AF%86%E8%92%B8%E9%A6%8F%20%E9%A2%84%E8%AE%AD%E7%BB%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (知识蒸馏 预训练）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28PyramidTNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer(Pyramid TNT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%82%85%E9%87%8C%E5%8F%B6%E7%BA%A7%E6%95%B0%E5%92%8C%E5%82%85%E9%87%8C%E5%8F%B6%E5%8F%98%E6%8D%A2%E4%B8%93%E9%A2%98/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    傅里叶级数和傅里叶变换专题
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Python%20%E5%AD%98%E5%82%A8%E4%B8%8E%E8%AF%BB%E5%8F%96%20HDF5%20%E6%96%87%E4%BB%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Python 存储与读取 HDF5 文件
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Segment%20Anything%20in%20Medical%20Images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Segment Anything in Medical Images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Mamba%20Efficient%20Visual%20Representation%20Learning%20with%20Bidirectional%20State%20Space%20Model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Mamba Efficient Visual Representation Learning with Bidirectional State Space Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_18" >
        
          
          <label class="md-nav__link" for="__nav_2_18" id="__nav_2_18_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚑 医学影像相关
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_18_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_18">
            <span class="md-nav__icon md-icon"></span>
            🚑 医学影像相关
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Why%20is%20the%20winner%20the%20best/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Why is the winner the best
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PORPOISE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PORPOISE
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HIPT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Scaling Vision Transformers to Gigapixel Images via Hierarchical Self-Supervised Learning(HIPT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Data-efficient%20and%20weakly%20supervised%20computational%20pathology%20on%20whole-slide%20images%28CLAM%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data efficient and weakly supervised computational pathology on whole slide images(CLAM)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Towards%20a%20general-purpose%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Towards a general purpose foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20whole-slide%20foundation%20model%20for%20digital%20pathology%20from%20real-world%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A whole slide foundation model for digital pathology from real world data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20visual-language%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A visual language foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Recurrent memory with optimal polynomial projections.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Modeling%20Dense%20Multimodal%20Interactions%20Between%20Biological%20Pathways%20and%20Histology%20for%20Survival%20Prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Prediction%20of%20recurrence%20risk%20in%20endometrial%20cancer%20with%20multimodal%20deep%20learning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction of recurrence risk in endometrial cancer with multimodal deep learning
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚀 Re:0 Road to CPath & Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            🚀 Re:0 Road to CPath & Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%201/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    理解全幅切片图像
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%202/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    在计算病理学中用于癌症诊断的弱监督深度学习
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%203/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    基于注意力的多示例学习可解释性
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../How%20to%20visualize%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to visualize Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AttentionViz%20A%20Global%20View%20of%20Transformer%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AttentionViz A Global View of Transformer Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Transformer%20Interpretability%20Beyond%20Attention%20Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transformer Interpretability Beyond Attention Visualization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualization Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Attention%20Generation%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Attention Generation Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🪛 PyTorch
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            🪛 PyTorch
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Autograd%20mechanics%20%E2%80%94%20PyTorch%202_3%20documentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Autograd mechanics — PyTorch 2 3 documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyTorch Autograd Explained   In depth Tutorial
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    💤 Re:0 Road to scRNA-seq Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            💤 Re:0 Road to scRNA-seq Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏳ CHAPTER 1 简介
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ⏳ CHAPTER 1 简介
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1.The building block of life
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20Raw%20Data%20Processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 Raw data processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Analysis%20frameworks%20and%20tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Analysis frameworks and tools
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🌏 CHAPTER 2 数据准备与可视化
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            🌏 CHAPTER 2 数据准备与可视化
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20normalization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 Normalization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3_feature_selection/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 Feature selection
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.4_dimensionality_reduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.4 Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_3" >
        
          
          <label class="md-nav__link" for="__nav_4_1_3" id="__nav_4_1_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🍀 CHAPTER 3 识别细胞结构
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_3">
            <span class="md-nav__icon md-icon"></span>
            🍀 CHAPTER 3 识别细胞结构
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 Data integration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_4" >
        
          
          <label class="md-nav__link" for="__nav_4_1_4" id="__nav_4_1_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏱️ CHAPTER 4 推断细胞轨迹
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_4">
            <span class="md-nav__icon md-icon"></span>
            ⏱️ CHAPTER 4 推断细胞轨迹
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotemporal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 Pseudotemporal ordering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../rna_velocity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 RNA velocity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../lineage_tracing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 Lineage tracing
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_5" >
        
          
          <label class="md-nav__link" for="__nav_4_1_5" id="__nav_4_1_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🎁 CHAPTER 5 应对特殊情况
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_5">
            <span class="md-nav__icon md-icon"></span>
            🎁 CHAPTER 5 应对特殊情况
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../differential_gene_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 Differential gene expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compositional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 Compositional analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gsea_pathway/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 Gene set enrichment and pathway analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../perturbation_modeling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 Perturbation modeling
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_6" >
        
          
          <label class="md-nav__link" for="__nav_4_1_6" id="__nav_4_1_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⚙️ CHAPTER 6 机制建模
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_6">
            <span class="md-nav__icon md-icon"></span>
            ⚙️ CHAPTER 6 机制建模
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_regulatory_networks/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 Gene regulatory networks
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cell_cell_communication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 Cell-cell communication
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_7" >
        
          
          <label class="md-nav__link" for="__nav_4_1_7" id="__nav_4_1_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🔓 CHAPTER 7 反卷积
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_7">
            <span class="md-nav__icon md-icon"></span>
            🔓 CHAPTER 7 反卷积
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bulk_deconvolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 Bulk deconvolution
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    简明易懂的单细胞分析流程（使用 R 语言）
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            简明易懂的单细胞分析流程（使用 R 语言）
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%89%8D%E8%A8%80/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    前言
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    环境配置
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20%E6%9E%84%E5%BB%BA%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1 构建单细胞 RNA seq 分析环境
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20R%20%E7%8E%AF%E5%A2%83%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 R 环境安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Docker%20%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Docker 安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.4%20%E5%9C%A8%20Docker%20%E4%B8%AD%E5%90%AF%E5%8A%A8%20Rstudio/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.4 在 Docker 中启动 Rstudio
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.5%20%E6%9C%89%E5%85%B3%20Bioconductor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.5 有关 Bioconductor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20R%20%E7%9A%84%E5%9F%BA%E6%9C%AC%E7%94%A8%E6%B3%95/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 R 的基本用法
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20%E5%A6%82%E4%BD%95%E5%A4%84%E7%90%86%20DataFrame%EF%BC%9F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 如何处理 DataFrame？
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3%20%E4%BD%BF%E7%94%A8%20ggplot2%20%E5%8F%AF%E8%A7%86%E5%8C%96%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 使用 ggplot2 可视化数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1%20PCA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 PCA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2%20%E4%BD%BF%E7%94%A8%20clusterPlofiler%20%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 使用 clusterPlofiler 进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E5%9F%BA%E5%9B%A0%E6%9C%AC%E4%BD%93%EF%BC%88GO%EF%BC%89%E4%B8%AD%E5%90%84%20BP%E3%80%81MF%20%E5%92%8C%20CC%20%E7%9A%84%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 基因本体（GO）中各 BP、MF 和 CC 的富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E4%BD%BF%E7%94%A8%20DEseq2%20%E6%8F%90%E5%8F%96%20RNA-seq%20%E5%88%86%E6%9E%90%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%20DEG/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 使用 DEseq2 提取 RNA seq 分析数据中的 DEG
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.4%20%E4%BD%BF%E7%94%A8%E5%85%B6%E4%BB%96%E6%95%B0%E6%8D%AE%E5%BA%93%EF%BC%88KEGG%E3%80%81Reactome%20%E7%AD%89%EF%BC%89%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.4 使用其他数据库（KEGG、Reactome 等）进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.1%20scRNA-seq%20%E6%A6%82%E8%A6%81/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 scRNA seq 概要
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.2%20%E5%87%86%E5%A4%87%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 准备使用 Seurat 进行分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.3%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 使用 Seurat 进行数据预处理
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.4%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E7%9A%84%E4%B8%80%E8%88%AC%E9%A1%BA%E5%BA%8F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.4 使用 Seurat 进行数据分析的一般顺序
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.1%20%E5%A6%82%E4%BD%95%E6%9F%A5%E6%89%BE%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 如何查找单细胞 RNA seq 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.2%20%E4%B8%BA%E6%AF%8F%E4%B8%AA%E7%B0%87%E8%87%AA%E5%8A%A8%E6%B7%BB%E5%8A%A0%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E6%A0%87%E7%AD%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 为每个簇自动添加细胞注释标签
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.3%20%E4%BD%BF%E7%94%A8%20FeaturePlot%20%E6%A3%80%E6%9F%A5%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 使用 FeaturePlot 检查基因表达
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 将单细胞 RNA seq 数据集相互结合
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    5.5 纠正批次效应，比较对照组和刺激组
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    5.5 纠正批次效应，比较对照组和刺激组
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      关于批次效应校正
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      数据集之间的批次效应校正
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      确定条件间的表达变化基因
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.6 确定每个群组的标记基因
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.1%20%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E7%BB%9F%E8%AE%A1%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E6%B0%B4%E5%B9%B3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 如何使用 Cell Ranger 统计 scRNA seq 数据中的基因表达水平
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.2%20%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E5%A4%84%E7%90%86%20NCBI%20Sequence%20Read%20Archive%20%28SRA%29%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 使用 Cell Ranger 处理 NCBI Sequence Read Archive (SRA) 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.1%20%E5%85%B3%E4%BA%8E%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 关于 scRNA seq 数据的质量控制
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.2%20%E4%BD%BF%E7%94%A8%20DoubleFinder%20%E6%A3%80%E6%B5%8B%E5%8F%8C%E9%87%8D%E7%BB%86%E8%83%9E/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.2 使用 DoubleFinder 检测双重细胞
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.3%20%E7%94%A8%20CellBender%20%E5%8E%BB%E9%99%A4%E8%83%8C%E6%99%AF%20RNA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.3 用 CellBender 去除背景 RNA
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧬 scRNA Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            🧬 scRNA Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🐍 Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            🐍 Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Large scale foundation model on single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scGPT%20toward%20building%20a%20foundation%20model%20for%20single-cell%20multi-omics%20using%20generative%20AI/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scGPT toward building a foundation model for single cell multi omics using generative AI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scBERT%20as%20a%20large-scale%20pretrained%20deep%20language%20model%20for%20cell%20type%20annotation%20of%20single-cell%20RNA-seq%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scBERT as a large scale pretrained deep language model for cell type annotation of single cell RNA seq data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RNA%20velocity%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA velocity of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TFvelo_gene%20regulation%20inspired%20RNA%20velocity%20estimation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TFvelo gene regulation inspired RNA velocity estimation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Disentanglement%20of%20single-cell%20data%20with%20biolord/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Disentanglement of single cell data with biolord
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Spatial%20transition%20tensor%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Spatial transition tensor of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRank%20infers%20drug-responsive%20cell%20types%20from%20untreated%20scRNA-seq%20data%20using%20a%20target-perturbed%20gene%20regulatory%20network/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRank infers drug responsive cell types from untreated scRNA seq data using a target perturbed gene regulatory network
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📔 文献
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            📔 文献
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ZNF689%20deficiency%20promotes%20intratumor%20heterogeneity%20and%20immunotherapy%20resistance%20in%20triple-negative%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ZNF689 deficiency promotes intratumor heterogeneity and immunotherapy resistance in triple negative breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Tumour%20vasculature%20at%20single-cell%20resolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tumour vasculature at single-cell resolution
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRNA-seq%20of%20gastric%20tumor%20shows%20complex%20intercellular%20interaction%20with%20an%20alternative%20T%20cell%20exhaustion%20trajectory%EF%BC%88%E5%8F%82%E8%80%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRNA seq of gastric tumor shows complex intercellular interaction with an alternative T cell exhaustion trajectory（参考）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20reveals%20new%20evolutionary%20complexity%20in%20uveal%20melanoma%EF%BC%88%E5%8F%82%E8%80%83%202%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis reveals new evolutionary complexity in uveal melanoma（参考 2）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20characterization%20of%20macrophages%20in%20uveal%20melanoma%20uncovers%20transcriptionally%20heterogeneous%20subsets%20conferring%20poor%20prognosis%20and%20aggressive%20behavior%EF%BC%88%E8%8C%83%E4%BE%8B%201%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell characterization of macrophages in uveal melanoma uncovers transcriptionally heterogeneous subsets conferring poor prognosis and aggressive behavior（范例 1）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ScRNA-seq%20of%20gastric%20cancer%20tissues%20reveals%20differences%20in%20the%20immune%20microenvironment%20of%20primary%20tumors%20and%20metastases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ScRNA seq of gastric cancer tissues reveals differences in the immune microenvironment of primary tumors and metastases
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Blimp-1%20and%20c-Maf%20regulate%20immune%20gene%20networks%20to%20protect%20against%20distinct%20pathways%20of%20pathobiont-induced%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Blimp 1 and c Maf regulate immune gene networks to protect against distinct pathways of pathobiont induced colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single%20cell%20analysis%20unveils%20B%20cell-dominated%20immune%20subtypes%20in%20HNSCC%20for%20enhanced%20prognostic%20and%20therapeutic%20stratification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis unveils B cell dominated immune subtypes in HNSCC for enhanced prognostic and therapeutic stratification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20spatial%20multi-omics%20and%20deep%20learning%20dissect%20enhancer-driven%20gene%20regulatory%20networks%20in%20liver%20zonation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell spatial multi omics and deep learning dissect enhancer driven gene regulatory networks in liver zonation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Neutrophil%20profiling%20illuminates%20anti-tumor%20antigen%20presenting%20potency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Neutrophil profiling illuminates anti tumor antigen presenting potency
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20blueprint%20for%20tumor-infiltrating%20B%20cells%20across%20human%20cancers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A blueprint for tumor infiltrating B cells across human cancers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20transcriptomic%20analyses%20reveal%20distinct%20immune%20cell%20contributions%20to%20epithelial%20barrier%20dysfunction%20in%20checkpoint%20inhibitor%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell transcriptomic analyses reveal distinct immune cell contributions to epithelial barrier dysfunction in checkpoint inhibitor colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Systematic%20dissection%20of%20tumor-normal%20single-cell%20ecosystems%20across%20a%20thousand%20tumors%20of%2030%20cancer%20types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Systematic dissection of tumor normal single cell ecosystems across a thousand tumors of 30 cancer types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../The%20aged%20tumor%20microenvironment%20limits%20T%20cell%20control%20of%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The aged tumor microenvironment limits T cell control of cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TIM-3%2B%20CD8%20T%20cells%20with%20a%20terminally%20exhausted%20phenotype%20retain%20functional%20capacity%20in%20hematological%20malignancies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TIM 3+ CD8 T cells with a terminally exhausted phenotype retain functional capacity in hematological malignancies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20lactate-SREBP2%20signaling%20axis%20drives%20tolerogenic%20dendritic%20cell%20maturation%20and%20promotes%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A lactate SREBP2 signaling axis drives tolerogenic dendritic cell maturation and promotes cancer progression
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20human%20omentum-specific%20mesothelial-like%20stromal%20population%20inhibits%20adipogenesis%20through%20IGFBP2%20secretion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A human omentum specific mesothelial like stromal population inhibits adipogenesis through IGFBP2 secretion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Interferon-stimulated%20neutrophils%20as%20a%20predictor%20of%20immunotherapy%20response/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interferon stimulated neutrophils as a predictor of immunotherapy response
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Repression%20of%20latent%20NF-%CE%BAB%20enhancers%20by%20PDX1%20regulates%20%CE%B2%20cell%20functional%20heterogeneity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Repression of latent NF κB enhancers by PDX1 regulates β cell functional heterogeneity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20systematic%20pan-cancer%20study%20on%20deep%20learning-based%20prediction%20of%20multi-omic%20biomarkers%20from%20routine%20pathology%20images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A systematic pan cancer study on deep learning based prediction of multi omic biomarkers from routine pathology images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20and%20spatial%20transcriptomics%20analysis%20of%20non-small%20cell%20lung%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell and spatial transcriptomics analysis of non small cell lung cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Seeing%20data%20as%20t-SNE%20and%20UMAP%20do/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Seeing data as t SNE and UMAP do
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mapping%20the%20cellular%20biogeography%20of%20human%20bone%20marrow%20niches%20using%20single-cell%20transcriptomics%20and%20proteomic%20imaging/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mapping the cellular biogeography of human bone marrow niches using single cell transcriptomics and proteomic imaging
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Onco-fetal%20Reprogramming%20of%20Endothelial%20Cells%20Drives%20Immunosuppressive%20Macrophages%20in%20Hepatocellular%20Carcinoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Onco fetal Reprogramming of Endothelial Cells Drives Immunosuppressive Macrophages in Hepatocellular Carcinoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20of%20anti-BCMA%20CAR%20T%20cell%20therapy%20in%20patients%20with%20central%20nervous%20system%20autoimmunity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis of anti BCMA CAR T cell therapy in patients with central nervous system autoimmunity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20sequencing%20depicts%20tumor%20architecture%20and%20empowers%20clinical%20decision%20in%20metastatic%20conjunctival%20melanoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell sequencing depicts tumor architecture and empowers clinical decision in metastatic conjunctival melanoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PR-SET7%20epigenetically%20restrains%20uterine%20interferon%20response%20and%20cell%20death%20governing%20proper%20postnatal%20stromal%20development/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PR SET7 epigenetically restrains uterine interferon response and cell death governing proper postnatal stromal development
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Clonal%20associations%20between%20lymphocyte%20subsets%20and%20functional%20states%20in%20rheumatoid%20arthritis%20synovium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clonal associations between lymphocyte subsets and functional states in rheumatoid arthritis synovium
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📐 Data Structures
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            📐 Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 1 章 python 入门
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%202%20%E7%AB%A0%20%E9%9D%A2%E5%90%91%E5%AF%B9%E8%B1%A1%E7%BC%96%E7%A8%8B/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 2 章 面向对象编程
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%203%20%E7%AB%A0%20%E7%AE%97%E6%B3%95%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 3 章 算法分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%204%20%E7%AB%A0%20%E9%80%92%E5%BD%92/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 4 章 递归
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%205%20%E7%AB%A0%20%E5%9F%BA%E4%BA%8E%E6%95%B0%E7%BB%84%E7%9A%84%E5%BA%8F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 5 章 基于数组的序列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%206%20%E7%AB%A0%20%E6%A0%88%E3%80%81%E9%98%9F%E5%88%97%E5%92%8C%E5%8F%8C%E7%AB%AF%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 6 章 栈、队列和双端队列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%207%20%E7%AB%A0%20%E9%93%BE%E8%A1%A8%28Linked%20Lists%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 7 章 链表(Linked Lists)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%208%20%E7%AB%A0%20%E6%A0%91%20%E2%98%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 8 章 树 ★
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%209%20%E7%AB%A0%20%E4%BC%98%E5%85%88%E7%BA%A7%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 9 章 优先级队列
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💡 Reviews
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            💡 Reviews
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_1" >
        
          
          <label class="md-nav__link" for="__nav_7_1" id="__nav_7_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧭 scRNA analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_1">
            <span class="md-nav__icon md-icon"></span>
            🧭 scRNA analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Best practices for single cell analysis across modalities
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Multi-omics%20integration%20in%20biomedical%20research%20%E2%80%93%20A%20metabolomics-centric%20review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multi omics integration in biomedical research – A metabolomics centric review
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Museum%20of%20spatial%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Museum of spatial transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20guide%20to%20artificial%20intelligence%20for%20cancer%20researchers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A guide to artificial intelligence for cancer researchers
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_2" >
        
          
          <label class="md-nav__link" for="__nav_7_2" id="__nav_7_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🦠 Metabolism
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_2">
            <span class="md-nav__icon md-icon"></span>
            🦠 Metabolism
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20analysis%20as%20a%20driver%20for%20discovery%2C%20diagnosis%2C%20and%20therapy/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic analysis as a driver for discovery, diagnosis, and therapy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20heterogeneity%20in%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic heterogeneity in cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Toward%20modeling%20metabolic%20state%20from%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Toward modeling metabolic state from single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20reprogramming%20and%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic reprogramming and cancer progression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_3" >
        
          
          <label class="md-nav__link" for="__nav_7_3" id="__nav_7_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Protein
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_3">
            <span class="md-nav__icon md-icon"></span>
            Protein
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Proteomic%20and%20interactomic%20insights%20into%20the%20molecular%20basis%20of%20cell%20functional%20diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Proteomic and interactomic insights into the molecular basis of cell functional diversity
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      关于批次效应校正
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      数据集之间的批次效应校正
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      确定条件间的表达变化基因
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/aletolia/aletolia.github.io.git/edit/master/docs/5.5 纠正批次效应，比较对照组和刺激组.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/aletolia/aletolia.github.io.git/raw/master/docs/5.5 纠正批次效应，比较对照组和刺激组.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


  <h1>5.5 纠正批次效应，比较对照组和刺激组</h1>

<p>本节中，我们将基于 Seurat 手册中的这篇文章编写代码。</p>
<p>单个样本的 scRNA-seq 数据分析可以在一定范围内进行可视化，但要比较多个样本，尤其是像对照组和刺激组之间的条件比较，就需要将这些样本进行整合后再进行比较。掌握数据整合的方法，可以清晰地了解不同条件下的细胞响应变化和复杂细胞类型之间的差异。</p>
<p>Seurat 提供了将两个以上的单细胞数据集整合并进行分析的方法。
例如，通过如下图所示的数据整合，我们可以看到对照组（CTRL）和刺激组（STIM）的 UMAP 几乎完全一致。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407091214545.png" /></p>
<p>有了这个整合样本，就可以确定与对照组相比，每个基因中表达上调的基因（下图以 IFI6 为例）。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407091214464.png" /></p>
<p>引用自 https://satijalab.org/seurat/articles/integration_introduction 并进行部分改编</p>
<h2 id="_1">关于批次效应校正<a class="headerlink" href="#_1" title="Permanent link">&para;</a></h2>
<p>当整合不同条件或实验的数据集时，技术差异会导致批次效应。使用锚定点（anchors），可以识别并校正这些批次效应。锚定点是一种框架的一部分，用于在不同的单细胞 RNA 测序（scRNA-seq）数据集中找到生物学上一致（或非常相似）的细胞对。锚定点有助于评估不同数据集或条件之间基因表达的相对变化（例如，刺激前后的变化），了解细胞群如何转变，以及哪些细胞相互关联等生物学见解。锚定点使用在数据集中一致变化的基因（特征）。</p>
<h2 id="_2">数据集之间的批次效应校正<a class="headerlink" href="#_2" title="Permanent link">&para;</a></h2>
<p>接下来介绍如何实现 scRNA-seq 数据整合。流程如下：创建 Seurat 对象→数据集之间的批次效应校正→解析。我们一步一步来看。</p>
<p>首先，创建 Seurat 对象。我们将使用 SeuratData 包来安装名为“ifnb”的数据集。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="nf">library</span><span class="p">(</span><span class="n">SeuratData</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="nf">library</span><span class="p">(</span><span class="n">patchwork</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span>
<span class="linenos" data-linenos="5 "></span><span class="c1"># 安装数据集</span>
<span class="linenos" data-linenos="6 "></span><span class="nf">InstallData</span><span class="p">(</span><span class="s">&quot;ifnb&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="7 "></span>
<span class="linenos" data-linenos="8 "></span><span class="c1"># 加载数据集</span>
<span class="linenos" data-linenos="9 "></span><span class="nf">LoadData</span><span class="p">(</span><span class="s">&quot;ifnb&quot;</span><span class="p">)</span>
</code></pre></div>
<p>在 ifnb 数据集中，包含了 IMMUNE_CTRL 和 IMMUNE_STIM 两个实验运行（Run）。orig.ident 列用于指示每个细胞的原始样本或运行。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407091215965.png" /></p>
<p>这段代码通过对每个刺激状态的数据进行标准化和识别变异基因，进行前处理，以便在后续分析（例如聚类、基因表达比较等）中获得更准确的结果。此外，这段代码还为多个数据集的有效整合做准备。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 将数据集拆分为两个Seurat对象的列表（stim和CTRL）</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">ifnb.list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">SplitObject</span><span class="p">(</span><span class="n">ifnb</span><span class="p">,</span><span class="w"> </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 3 "></span>
<span class="linenos" data-linenos=" 4 "></span><span class="c1"># 独立地对每个数据集进行标准化和识别变异基因</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">ifnb.list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="n">X</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifnb.list</span><span class="p">,</span><span class="w"> </span><span class="n">FUN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">    </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">NormalizeData</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">    </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">selection.method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;vst&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">nfeatures</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2000</span><span class="p">)</span>
<span class="linenos" data-linenos=" 8 "></span><span class="p">})</span>
<span class="linenos" data-linenos=" 9 "></span>
<span class="linenos" data-linenos="10 "></span><span class="c1"># 选择跨数据集重复可变的特征以进行整合</span>
<span class="linenos" data-linenos="11 "></span><span class="n">features</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">SelectIntegrationFeatures</span><span class="p">(</span><span class="n">object.list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifnb.list</span><span class="p">)</span>
</code></pre></div>
<p>详细的代码解释如下：</p>
<ol>
<li><code>SplitObject(ifnb, split.by = "stim")</code>：根据 Seurat 对象 ifnb 中 meta.data 的 "stim" 列（表示刺激状态的元数据），将 ifnb 分割为两个不同的 Seurat 对象。这样可以根据刺激状态单独分析数据。</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">ifnb.list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">SplitObject</span><span class="p">(</span><span class="n">ifnb</span><span class="p">,</span><span class="w"> </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407091216774.png" /></p>
<p>使用 lapply 函数对每个 Seurat 对象（每种刺激状态的数据）执行以下操作：</p>
<ol>
<li><strong>NormalizeData(x)</strong>: 对数据进行归一化。这是通过根据每个细胞的 RNA 总量对计数进行缩放来实现的。</li>
<li><strong>FindVariableFeatures(...)</strong>: 查找变量基因（在每个细胞群中表现出不同表达的基因）。此步骤对于后续的聚类和其他分析中识别重要基因至关重要。</li>
<li><strong>SelectIntegrationFeatures(object.list = ifnb.list)</strong>: 在整合两个或更多数据集之前，选择那些在数据集中表现出一致变动的基因（特征）。此步骤在使用 <code>FindIntegrationAnchors()</code> 和 <code>IntegrateData()</code> 等函数整合数据集时非常重要。</li>
</ol>
<p>顺便说一下，有些人可能会好奇，为什么一开始要将 seurat 对象拆分。对象拆分有以下几个原因：</p>
<ol>
<li>
<p><strong>理解每个数据集的特性</strong>: 对每种状态（此情况下为刺激状态或非刺激状态）进行细胞反应可能不同。为了理解这些差异，需要分别解析每个状态。通过这种独立解析，可以识别每种状态特有的变量基因或细胞簇。</p>
</li>
<li>
<p><strong>校正批次效应</strong>: 通过对每种状态进行独立解析，可以在后续整合阶段更有效地校正批次效应（技术变动）。特别是当每种状态是在不同实验条件下获得时，这一点尤为重要。</p>
</li>
<li>
<p><strong>高级比较分析</strong>: 将每种状态作为独立的对象进行分析，可以实现更高级的比较分析，例如基因表达的差异和细胞簇的差异。</p>
</li>
<li>
<p><strong>提高整合精度</strong>: Seurat 的整合方法使用每个数据集（此情况下为每种状态）中选择的变量基因来找到“锚点”（表示相应细胞状态的点），然后基于这些锚点整合数据集。这个过程需要每个数据集独立预处理的信息。</p>
</li>
</ol>
<p>以上这些原因使得在很多情况下，先将数据分成独立的对象分别进行预处理和基本分析，然后再进行整合是一个合理的方法。</p>
<p>接下来，使用锚点对两个数据集进行批次效应校正并进行整合。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># Find integration anchors</span>
<span class="linenos" data-linenos="2 "></span><span class="n">immune.anchors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindIntegrationAnchors</span><span class="p">(</span>
<span class="linenos" data-linenos="3 "></span><span class="w">  </span><span class="n">object.list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifnb.list</span><span class="p">,</span><span class="w"> </span><span class="n">anchor.features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">features</span>
<span class="linenos" data-linenos="4 "></span><span class="p">)</span>
<span class="linenos" data-linenos="5 "></span>
<span class="linenos" data-linenos="6 "></span><span class="c1"># This command creates an &#39;integrated&#39; data assay</span>
<span class="linenos" data-linenos="7 "></span><span class="n">immune.combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">IntegrateData</span><span class="p">(</span><span class="n">anchorset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">immune.anchors</span><span class="p">)</span>
</code></pre></div>
<ul>
<li><strong>FindIntegrationAnchors 函数</strong>: 这个函数根据指定的特征量（变量基因）在多个 scRNA-seq 数据集之间找到称为“锚点”的匹配细胞对。</li>
<li><code>object.list = ifnb.list</code>: 这个参数传递要整合的 Seurat 对象列表（这里是 ifnb.list）。</li>
<li>
<p><code>anchor.features = features</code>: 用于整合的特征量（基因）列表。通常是那些在多个数据集中变动较大的基因。</p>
</li>
<li>
<p><strong>IntegrateData 函数</strong>: 这个函数使用前一步找到的锚点将多个数据集整合为一个“整合的”数据集。</p>
</li>
<li><code>anchorset = immune.anchors</code>: 这个参数包含前一步找到的锚点。</li>
</ul>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092106493.png" /></p>
<p>对于 ifnb 数据集，两个条件存在于同一个 Seurat 对象 (IMMUNE_CTRL 和 IMMUNE_STIM) 中，但这并未进行数据整合。进行整合后的 immune.combined 对象中，新生成了一个 "integrated" 的分析层，包含了整合后的数据。</p>
<p>在 ifnb 对象阶段，仅存在 "RNA" 分析层，这可以解释为两个数据集的条件未整合，只是作为一个 Seurat 对象保存。</p>
<p>接下来，将整合后的数据集带入下游分析。</p>
<p>首先，通过以下代码将 DefaultAssay 设置为 integrated。这段代码将 immune.combined 对象的默认分析层设置为 "integrated"，这样接下来的分析将以整合数据为目标：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;integrated&quot;</span>
</code></pre></div>
<p>这意味着在这个 Seurat 对象 (immune.combined) 中，默认的分析层已经被指定为 "integrated"，因此后续的所有分析都将针对这个整合后的数据集进行。</p>
<p>这样设置 DefaultAssay 为 integrated 后，可以确保后续的分析步骤（如聚类、降维等）都基于整合后的数据进行。这一步非常重要，因为它明确了接下来分析的基准数据。</p>
<p>接下来进行下游处理。由于这是在 Seurat 中进行的常规预处理，因此省略详细步骤，但使用 magrittr 库和 %&gt;% 运算符进行流水线处理。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">library</span><span class="p">(</span><span class="n">magrittr</span><span class="p">)</span><span class="w">  </span><span class="c1"># 加载 magrittr 包以使用 %&gt;% 运算符</span>
<span class="linenos" data-linenos="2 "></span>
<span class="linenos" data-linenos="3 "></span><span class="c1"># 现在可以使用管道运算符</span>
<span class="linenos" data-linenos="4 "></span><span class="n">immune.combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">immune.combined</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="linenos" data-linenos="5 "></span><span class="w">  </span><span class="nf">ScaleData</span><span class="p">(</span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="linenos" data-linenos="6 "></span><span class="w">  </span><span class="nf">RunPCA</span><span class="p">(</span><span class="n">npcs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">30</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="linenos" data-linenos="7 "></span><span class="w">  </span><span class="nf">RunUMAP</span><span class="p">(</span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;pca&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="linenos" data-linenos="8 "></span><span class="w">  </span><span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;pca&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="linenos" data-linenos="9 "></span><span class="w">  </span><span class="nf">FindClusters</span><span class="p">(</span><span class="n">resolution</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span>
</code></pre></div>
<p>接下来，让我们可视化 UMAP。为了比较有无整合处理的效果，分别输出 immune.combined（整合后的数据集）和 ifnb（未整合的数据集）。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># 可视化</span>
<span class="linenos" data-linenos="2 "></span><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">repel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span><span class="n">p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p2</span>
</code></pre></div>
<p>这段代码将可视化使用 immune.combined（整合后的数据集）的 UMAP。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092107333.png" /></p>
<p>如果使用 ifnb 进行可视化（未整合的数据集），可以使用以下代码：</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092108539.png" /></p>
<p>如你所见，在 ifnb 数据集中，左侧的 UMAP 图中可以确认 CTRL 和 STIM 完全没有重叠。因此，在左侧的聚类结果中，可以看到同样的细胞群（例如 0 和 3 可能属于同一细胞群）被检测为不同的聚类，这使得分析结果不准确。这说明了整合处理的重要性。</p>
<p>顺便说一下，指定 <code>split.by = "stim"</code> 可以将 CTRL 和 STIM 的图表合并。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">)</span>
</code></pre></div>
<p>这样，您就可以在一张图中比较 CTRL 和 STIM 的 UMAP 分布，更清楚地看到它们的差异和相似之处。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092108445.png" /></p>
<p>在进入下一部分之前，我们先为每个聚类分配细胞类型。（这次将省略具体的细胞类型和标志基因的确定过程。）</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">immune.combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RenameIdents</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span>
<span class="linenos" data-linenos=" 2 "></span><span class="w">  </span><span class="n">`0`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD14 Mono&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`1`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD4 Naive T&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`2`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD4 Memory T&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 3 "></span><span class="w">  </span><span class="n">`3`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD16 Mono&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`4`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;B&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`5`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD8 T&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`6`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;NK&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`7`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;T activated&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 4 "></span><span class="w">  </span><span class="n">`8`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;DC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`9`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;B Activated&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`10`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Mk&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`11`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;pDC&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`12`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Eryth&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">  </span><span class="n">`13`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Mono/Mk Doublets&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">`14`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;HSPC&quot;</span>
<span class="linenos" data-linenos=" 6 "></span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span>
<span class="linenos" data-linenos="10 "></span><span class="w">  </span><span class="n">immune.combined</span><span class="p">,</span>
<span class="linenos" data-linenos="11 "></span><span class="w">  </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="12 "></span><span class="w">  </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_annotations&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="13 "></span><span class="w">  </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span>
<span class="linenos" data-linenos="14 "></span><span class="w">  </span><span class="n">repel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span>
<span class="linenos" data-linenos="15 "></span><span class="p">)</span>
<span class="linenos" data-linenos="16 "></span><span class="n">p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p2</span>
</code></pre></div>
<p>以上代码将为每个聚类分配对应的细胞类型，并可视化结果。第一个图 (<code>p1</code>) 显示了根据刺激状态分组的 UMAP 图，第二个图 (<code>p2</code>) 显示了根据细胞类型标注的 UMAP 图。通过这种方式，可以直观地查看不同细胞类型在 UMAP 中的分布情况。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092109299.png" /></p>
<h2 id="_3">确定条件间的表达变化基因<a class="headerlink" href="#_3" title="Permanent link">&para;</a></h2>
<p>那么，让我们实际比较一下对照组和刺激组。</p>
<p>以下代码用于可视化特定标记基因的表达水平，基于不同的聚类或类别。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 设置细胞类型的顺序</span>
<span class="linenos" data-linenos=" 2 "></span><span class="nf">Idents</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span>
<span class="linenos" data-linenos=" 3 "></span><span class="w">  </span><span class="nf">Idents</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">),</span>
<span class="linenos" data-linenos=" 4 "></span><span class="w">  </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">    </span><span class="s">&quot;HSPC&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Mono/Mk Doublets&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;pDC&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Eryth&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Mk&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;DC&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD14 Mono&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">    </span><span class="s">&quot;CD16 Mono&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;B Activated&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;B&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8 T&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;NK&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;T activated&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">    </span><span class="s">&quot;CD4 Naive T&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4 Memory T&quot;</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">  </span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span>
<span class="linenos" data-linenos="11 "></span><span class="c1"># 定义要可视化的标记基因列表</span>
<span class="linenos" data-linenos="12 "></span><span class="n">markers.to.plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span>
<span class="linenos" data-linenos="13 "></span><span class="w">  </span><span class="s">&quot;CD3D&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CREM&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;HSPH1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;SELL&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;GIMAP5&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CACYBP&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;GNLY&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;NKG7&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="14 "></span><span class="w">  </span><span class="s">&quot;CCL5&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8A&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;MS4A1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD79A&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;MIR155HG&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;NME1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;FCGR3A&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;VMO1&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="15 "></span><span class="w">  </span><span class="s">&quot;CCL2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;S100A9&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;HLA-DQA1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;GPR183&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;PPBP&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;GNG11&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;HBA2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;HBB&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="16 "></span><span class="w">  </span><span class="s">&quot;TSPAN13&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;IL3RA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;IGJ&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;PRSS57&quot;</span>
<span class="linenos" data-linenos="17 "></span><span class="p">)</span>
<span class="linenos" data-linenos="18 "></span>
<span class="linenos" data-linenos="19 "></span><span class="c1"># 使用 DotPlot 可视化标记基因的表达</span>
<span class="linenos" data-linenos="20 "></span><span class="nf">DotPlot</span><span class="p">(</span>
<span class="linenos" data-linenos="21 "></span><span class="w">  </span><span class="n">immune.combined</span><span class="p">,</span>
<span class="linenos" data-linenos="22 "></span><span class="w">  </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">markers.to.plot</span><span class="p">,</span>
<span class="linenos" data-linenos="23 "></span><span class="w">  </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;blue&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">),</span>
<span class="linenos" data-linenos="24 "></span><span class="w">  </span><span class="n">dot.scale</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span>
<span class="linenos" data-linenos="25 "></span><span class="w">  </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span>
<span class="linenos" data-linenos="26 "></span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">RotatedAxis</span><span class="p">()</span>
</code></pre></div>
<ul>
<li><code>Idents</code> 函数用于从 Seurat 对象中获取当前的聚类 ID 或其他身份信息。通过重新排序 <code>immune.combined</code> 的身份，可以控制后续绘图中的显示顺序。</li>
<li><code>markers.to.plot &lt;- c("CD3D", "CREM", ...</code> 定义了要绘制的标记基因列表。</li>
<li><code>DotPlot</code> 函数用于绘制点图，显示标记基因在不同条件下的表达水平。<code>split.by = "stim"</code> 参数将对照组（CTRL）和刺激组（STIM）的比较显示在同一图中。</li>
</ul>
<p>成功执行上述代码后，输出应为一个点图，其中行代表每个细胞类型在 CTRL 和 STIM 条件下的比较，列代表基因特征（markers）。这样，您可以直观地查看各基因在不同条件下的表达变化。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092111555.png" /></p>
<p>接下来，我们将确定条件间表达不同的基因。</p>
<p>这段代码旨在视觉比较刺激细胞和对照细胞的基因表达差异。具体来说，它比较了 Naive T 细胞和 CD14 单核细胞群体的表达，并在散点图中强调显示因干扰素刺激而反应的特定基因。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 加载可视化库</span>
<span class="linenos" data-linenos=" 2 "></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="linenos" data-linenos=" 3 "></span><span class="nf">library</span><span class="p">(</span><span class="n">cowplot</span><span class="p">)</span>
<span class="linenos" data-linenos=" 4 "></span><span class="nf">theme_set</span><span class="p">(</span><span class="nf">theme_cowplot</span><span class="p">())</span>
<span class="linenos" data-linenos=" 5 "></span>
<span class="linenos" data-linenos=" 6 "></span><span class="c1"># 提取和处理 &quot;CD4 Naive T&quot; 细胞的数据</span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">t.cells</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span><span class="w"> </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD4 Naive T&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 8 "></span><span class="nf">Idents</span><span class="p">(</span><span class="n">t.cells</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;stim&quot;</span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">avg.t.cells</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span>
<span class="linenos" data-linenos="10 "></span><span class="w">  </span><span class="nf">log1p</span><span class="p">(</span><span class="nf">AverageExpression</span><span class="p">(</span><span class="n">t.cells</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="o">$</span><span class="n">RNA</span><span class="p">)</span>
<span class="linenos" data-linenos="11 "></span><span class="p">)</span>
<span class="linenos" data-linenos="12 "></span><span class="n">avg.t.cells</span><span class="o">$</span><span class="n">gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">avg.t.cells</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span>
<span class="linenos" data-linenos="14 "></span><span class="c1"># 提取和处理 &quot;CD14 Mono&quot; 细胞的数据</span>
<span class="linenos" data-linenos="15 "></span><span class="n">cd14.mono</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">,</span><span class="w"> </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD14 Mono&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="16 "></span><span class="nf">Idents</span><span class="p">(</span><span class="n">cd14.mono</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;stim&quot;</span>
<span class="linenos" data-linenos="17 "></span><span class="n">avg.cd14.mono</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span>
<span class="linenos" data-linenos="18 "></span><span class="w">  </span><span class="nf">log1p</span><span class="p">(</span><span class="nf">AverageExpression</span><span class="p">(</span><span class="n">cd14.mono</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="o">$</span><span class="n">RNA</span><span class="p">)</span>
<span class="linenos" data-linenos="19 "></span><span class="p">)</span>
<span class="linenos" data-linenos="20 "></span><span class="n">avg.cd14.mono</span><span class="o">$</span><span class="n">gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">avg.cd14.mono</span><span class="p">)</span>
<span class="linenos" data-linenos="21 "></span>
<span class="linenos" data-linenos="22 "></span><span class="c1"># 指定在图中要突出的基因</span>
<span class="linenos" data-linenos="23 "></span><span class="n">genes.to.label</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span>
<span class="linenos" data-linenos="24 "></span><span class="w">  </span><span class="s">&quot;ISG15&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;LY6E&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;IFI6&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ISG20&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;MX1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;IFIT2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;IFIT1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CXCL10&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CCL8&quot;</span>
<span class="linenos" data-linenos="25 "></span><span class="p">)</span>
<span class="linenos" data-linenos="26 "></span>
<span class="linenos" data-linenos="27 "></span><span class="c1"># 创建散点图并标注指定的基因</span>
<span class="linenos" data-linenos="28 "></span><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">avg.t.cells</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">CTRL</span><span class="p">,</span><span class="w"> </span><span class="n">STIM</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="29 "></span><span class="w">  </span><span class="nf">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="30 "></span><span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;CD4 Naive T Cells&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="31 "></span><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">LabelPoints</span><span class="p">(</span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p1</span><span class="p">,</span><span class="w"> </span><span class="n">points</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genes.to.label</span><span class="p">,</span><span class="w"> </span><span class="n">repel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="linenos" data-linenos="32 "></span><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">avg.cd14.mono</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">CTRL</span><span class="p">,</span><span class="w"> </span><span class="n">STIM</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="33 "></span><span class="w">  </span><span class="nf">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="34 "></span><span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;CD14 Monocytes&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="35 "></span><span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">LabelPoints</span><span class="p">(</span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p2</span><span class="p">,</span><span class="w"> </span><span class="n">points</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genes.to.label</span><span class="p">,</span><span class="w"> </span><span class="n">repel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="linenos" data-linenos="36 "></span>
<span class="linenos" data-linenos="37 "></span><span class="c1"># 并排显示图表</span>
<span class="linenos" data-linenos="38 "></span><span class="n">p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p2</span>
</code></pre></div>
<p>代码的处理流程如下：</p>
<ol>
<li>获取刺激的 Naive T 细胞和对照的 CD14 单核细胞群体的平均基因表达值。</li>
<li>使用 ggplot2 和 cowplot 包绘制这两种细胞类型的基因表达。</li>
<li>突出显示并标注特定基因（如 ISG15、LY6E 等）。</li>
<li>最后，分别显示两个图表（Naive T 细胞和 CD14 单核细胞）并排显示。</li>
</ol>
<p>成功执行上述代码后，您将会看到以下的散点图，其中显示了 Naive T 细胞和 CD14 单核细胞在 CTRL 和 STIM 条件下的基因表达比较。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092112549.png" /></p>
<p>以下的代码针对单细胞 RNA-seq 数据集 immune.combined，根据刺激状态进行细胞类型识别，确定差异表达基因，并在 Feature plot 中可视化特定基因的表达水平。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 创建一个结合细胞类型和刺激状态的新标识符</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">immune.combined</span><span class="o">$</span><span class="n">celltype.stim</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span>
<span class="linenos" data-linenos=" 3 "></span><span class="w">  </span><span class="nf">Idents</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">),</span>
<span class="linenos" data-linenos=" 4 "></span><span class="w">  </span><span class="n">immune.combined</span><span class="o">$</span><span class="n">stim</span><span class="p">,</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">  </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;_&quot;</span>
<span class="linenos" data-linenos=" 6 "></span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span>
<span class="linenos" data-linenos=" 8 "></span><span class="c1"># 备份当前的细胞类型标识符，并更新为新的组合标识符</span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">immune.combined</span><span class="o">$</span><span class="n">celltype</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">Idents</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span><span class="nf">Idents</span><span class="p">(</span><span class="n">immune.combined</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;celltype.stim&quot;</span>
<span class="linenos" data-linenos="11 "></span>
<span class="linenos" data-linenos="12 "></span><span class="c1"># 识别刺激B细胞和对照B细胞之间差异表达的基因</span>
<span class="linenos" data-linenos="13 "></span><span class="n">b.interferon.response</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindMarkers</span><span class="p">(</span>
<span class="linenos" data-linenos="14 "></span><span class="w">  </span><span class="n">immune.combined</span><span class="p">,</span>
<span class="linenos" data-linenos="15 "></span><span class="w">  </span><span class="n">ident.1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;B_STIM&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="16 "></span><span class="w">  </span><span class="n">ident.2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;B_CTRL&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="17 "></span><span class="w">  </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span>
<span class="linenos" data-linenos="18 "></span><span class="p">)</span>
<span class="linenos" data-linenos="19 "></span>
<span class="linenos" data-linenos="20 "></span><span class="c1"># 可视化选定基因的表达，按刺激状态分割</span>
<span class="linenos" data-linenos="21 "></span><span class="nf">FeaturePlot</span><span class="p">(</span>
<span class="linenos" data-linenos="22 "></span><span class="w">  </span><span class="n">immune.combined</span><span class="p">,</span>
<span class="linenos" data-linenos="23 "></span><span class="w">  </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD3D&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;GNLY&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;IFI6&quot;</span><span class="p">),</span>
<span class="linenos" data-linenos="24 "></span><span class="w">  </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos="25 "></span><span class="w">  </span><span class="n">max.cutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span>
<span class="linenos" data-linenos="26 "></span><span class="w">  </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;grey&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="27 "></span><span class="p">)</span>
</code></pre></div>
<p>解释：</p>
<ul>
<li><code>immune.combined$celltype.stim</code>：创建一个结合细胞类型和刺激状态的新标识符。</li>
<li>备份当前的细胞类型标识符，并使用新的组合标识符进行更新。</li>
<li>使用 <code>FindMarkers</code> 函数识别刺激 B 细胞和对照 B 细胞之间差异表达的基因。</li>
<li>使用 <code>FeaturePlot</code> 可视化选定基因（CD3D、GNLY、IFI6）的表达，并按刺激状态分割显示。</li>
</ul>
<p>成功执行上述代码后，您将看到以下的 Feature plot。图中可以清晰地看到在 CTRL 和 STIM 条件下，哪些聚类的基因表达发生了变化（例如，IFI6 基因）。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092113372.png" /></p>
<p>最后，让我们通过小提琴图可视化各细胞类型的基因表达。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 为指定基因生成小提琴图，根据刺激状态分组并按细胞类型进行分组</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">plots</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">VlnPlot</span><span class="p">(</span>
<span class="linenos" data-linenos=" 3 "></span><span class="w">  </span><span class="n">immune.combined</span><span class="p">,</span>
<span class="linenos" data-linenos=" 4 "></span><span class="w">  </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;LYZ&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ISG15&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CXCL10&quot;</span><span class="p">),</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">  </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">  </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;celltype&quot;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">  </span><span class="n">pt.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">  </span><span class="n">combine</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span>
<span class="linenos" data-linenos=" 9 "></span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span>
<span class="linenos" data-linenos="11 "></span><span class="c1"># 以单列布局显示生成的小提琴图</span>
<span class="linenos" data-linenos="12 "></span><span class="nf">wrap_plots</span><span class="p">(</span><span class="n">plots</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plots</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
</code></pre></div>
<p>解释：</p>
<ul>
<li><code>VlnPlot</code> 函数生成指定基因（LYZ、ISG15、CXCL10）的小提琴图，这些图根据刺激状态分割，并按细胞类型分组。</li>
<li><code>wrap_plots</code> 函数将生成的小提琴图在单列布局中显示。</li>
</ul>
<p>成功执行上述代码后，您将看到如图所示的小提琴图。您可以清楚地看到在刺激后的样本中，ISG15和CXCL10的表达水平在各细胞类型中显著上升。</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202407092113470.png" /></p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 5.4 将单细胞 RNA seq 数据集相互结合">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                5.4 将单细胞 RNA seq 数据集相互结合
              </div>
            </div>
          </a>
        
        
          
          <a href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/" class="md-footer__link md-footer__link--next" aria-label="Next: 5.6 确定每个群组的标记基因">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                5.6 确定每个群组的标记基因
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.code.select", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.instant.prefetch", "navigation.instant.progress", "navigation.path", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "header.autohide"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.525ec568.min.js"></script>
      
        <script src="../mathjax-config.js"></script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../extra.js"></script>
      
        <script src="../js/toggle_sidebar.js"></script>
      
    
  </body>
</html>