
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://localhost:8000/1.3%20Analysis%20frameworks%20and%20tools/">
      
      
        <link rel="prev" href="../1.2%20Raw%20Data%20Processing/">
      
      
        <link rel="next" href="../2.1%20quality_control/">
      
      
      <link rel="icon" href="../img/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>1.3 Analysis frameworks and tools - One Last Kiss</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lxgw-wenkai-webfont@1.1.0/style.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-M0R7V8QGQ7"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-M0R7V8QGQ7",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-M0R7V8QGQ7",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#141" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="One Last Kiss" class="md-header__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            One Last Kiss
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1.3 Analysis frameworks and tools
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
    
  
  🏠 Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Mygo/" class="md-tabs__link">
          
  
    
  
  ⛏️ CV and Multimodels

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Session%201/" class="md-tabs__link">
          
  
    
  
  🚀 Re:0 Road to CPath & Visualization

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-tabs__link">
          
  
    
  
  💤 Re:0 Road to scRNA-seq Analysis

        </a>
      </li>
    
  

    
  

    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-tabs__link">
          
  
    
  
  🧬 scRNA Analysis

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-tabs__link">
          
  
    
  
  📐 Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-tabs__link">
          
  
    
  
  💡 Reviews

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="One Last Kiss" class="md-nav__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    One Last Kiss
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href=".." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    🏠 Home
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            🏠 Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⛏️ CV and Multimodels
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ⛏️ CV and Multimodels
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mygo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MyGO: Discrete Modality Information as Fine-Grained Tokens for Multi-modal Knowledge Graph Completion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GMM/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GMM
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../DINO/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    DINO
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mamba/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mamba
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ViT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ViT
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCoV2%2C3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo V2
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%80%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (一)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%20%28%E4%BA%8C%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (二)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%89%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (三)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E5%9B%9B%29%20%28TNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (四)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E7%9F%A5%E8%AF%86%E8%92%B8%E9%A6%8F%20%E9%A2%84%E8%AE%AD%E7%BB%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (知识蒸馏 预训练）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28PyramidTNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer(Pyramid TNT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%82%85%E9%87%8C%E5%8F%B6%E7%BA%A7%E6%95%B0%E5%92%8C%E5%82%85%E9%87%8C%E5%8F%B6%E5%8F%98%E6%8D%A2%E4%B8%93%E9%A2%98/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    傅里叶级数和傅里叶变换专题
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Python%20%E5%AD%98%E5%82%A8%E4%B8%8E%E8%AF%BB%E5%8F%96%20HDF5%20%E6%96%87%E4%BB%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Python 存储与读取 HDF5 文件
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Segment%20Anything%20in%20Medical%20Images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Segment Anything in Medical Images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Mamba%20Efficient%20Visual%20Representation%20Learning%20with%20Bidirectional%20State%20Space%20Model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Mamba Efficient Visual Representation Learning with Bidirectional State Space Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_18" >
        
          
          <label class="md-nav__link" for="__nav_2_18" id="__nav_2_18_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚑 医学影像相关
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_18_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_18">
            <span class="md-nav__icon md-icon"></span>
            🚑 医学影像相关
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Why%20is%20the%20winner%20the%20best/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Why is the winner the best
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PORPOISE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PORPOISE
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HIPT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Scaling Vision Transformers to Gigapixel Images via Hierarchical Self-Supervised Learning(HIPT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Data-efficient%20and%20weakly%20supervised%20computational%20pathology%20on%20whole-slide%20images%28CLAM%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data efficient and weakly supervised computational pathology on whole slide images(CLAM)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Towards%20a%20general-purpose%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Towards a general purpose foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20whole-slide%20foundation%20model%20for%20digital%20pathology%20from%20real-world%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A whole slide foundation model for digital pathology from real world data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20visual-language%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A visual language foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Recurrent memory with optimal polynomial projections.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Modeling%20Dense%20Multimodal%20Interactions%20Between%20Biological%20Pathways%20and%20Histology%20for%20Survival%20Prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Prediction%20of%20recurrence%20risk%20in%20endometrial%20cancer%20with%20multimodal%20deep%20learning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction of recurrence risk in endometrial cancer with multimodal deep learning
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚀 Re:0 Road to CPath & Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            🚀 Re:0 Road to CPath & Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%201/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    理解全幅切片图像
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%202/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    在计算病理学中用于癌症诊断的弱监督深度学习
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%203/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    基于注意力的多示例学习可解释性
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../How%20to%20visualize%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to visualize Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AttentionViz%20A%20Global%20View%20of%20Transformer%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AttentionViz A Global View of Transformer Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Transformer%20Interpretability%20Beyond%20Attention%20Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transformer Interpretability Beyond Attention Visualization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualization Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Attention%20Generation%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Attention Generation Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🪛 PyTorch
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            🪛 PyTorch
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Autograd%20mechanics%20%E2%80%94%20PyTorch%202_3%20documentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Autograd mechanics — PyTorch 2 3 documentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyTorch Autograd Explained   In depth Tutorial
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    💤 Re:0 Road to scRNA-seq Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            💤 Re:0 Road to scRNA-seq Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏳ CHAPTER 1 简介
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ⏳ CHAPTER 1 简介
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1.The building block of life
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20Raw%20Data%20Processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 Raw data processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    1.3 Analysis frameworks and tools
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    1.3 Analysis frameworks and tools
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#141" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.1. 单细胞分析框架和联盟
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#142-anndata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2. 使用 AnnData 存储单模态数据
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.2. 使用 AnnData 存储单模态数据">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1421" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1422-anndata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.2. 初始化 AnnData 对象
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1423" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.3. 添加对齐的元数据
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.2.3. 添加对齐的元数据">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#14231" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.3.1. 观测值或变量级别
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14232" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.3.2. 使用元数据进行子集化
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1424" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.4. 观测值/变量级别矩阵
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1425" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.5. 非结构化元数据
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1426" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.6. 层
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1427-dataframe" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.7. 转换为 DataFrame
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1428-anndata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.8. 读取和写入 AnnData 对象
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1429" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.9. 高效数据访问
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.2.9. 高效数据访问">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#14291" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.9.1. 视图和副本
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14292" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.9.2. 大数据的部分读取
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#143-scanpy" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3. 使用 scanpy 进行单模态数据分析
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.3. 使用 scanpy 进行单模态数据分析">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1431" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1432-scanpy-api" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3.2. Scanpy API 设计
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1433-scanpy" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3.3. Scanpy 示例 #
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#144-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4. 进阶：使用 MuData 存储多模态数据
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.4. 进阶：使用 MuData 存储多模态数据">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1441" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1442-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.2. 初始化 MuData 对象
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1443-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.3. MuData 属性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1444" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.4. 变量映射
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1445-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.5. MuData 视图
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1446" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.6. 共同的观测值
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1447-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.7. MuData 对象的读取和写入
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1448" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.8. 多模态方法
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#145-muon" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5. 进阶：使用 muon 进行多模态数据分析
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.5. 进阶：使用 muon 进行多模态数据分析">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1451" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1452-api" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5.2. API 概览
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1453-muon-api" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5.3. muon API 演示
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🌏 CHAPTER 2 数据准备与可视化
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            🌏 CHAPTER 2 数据准备与可视化
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20normalization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 Normalization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3_feature_selection/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 Feature selection
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.4_dimensionality_reduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.4 Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_3" >
        
          
          <label class="md-nav__link" for="__nav_4_1_3" id="__nav_4_1_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🍀 CHAPTER 3 识别细胞结构
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_3">
            <span class="md-nav__icon md-icon"></span>
            🍀 CHAPTER 3 识别细胞结构
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 Data integration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_4" >
        
          
          <label class="md-nav__link" for="__nav_4_1_4" id="__nav_4_1_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏱️ CHAPTER 4 推断细胞轨迹
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_4">
            <span class="md-nav__icon md-icon"></span>
            ⏱️ CHAPTER 4 推断细胞轨迹
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotemporal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 Pseudotemporal ordering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../rna_velocity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 RNA velocity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../lineage_tracing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 Lineage tracing
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_5" >
        
          
          <label class="md-nav__link" for="__nav_4_1_5" id="__nav_4_1_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🎁 CHAPTER 5 应对特殊情况
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_5">
            <span class="md-nav__icon md-icon"></span>
            🎁 CHAPTER 5 应对特殊情况
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../differential_gene_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 Differential gene expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compositional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 Compositional analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gsea_pathway/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 Gene set enrichment and pathway analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../perturbation_modeling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 Perturbation modeling
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_6" >
        
          
          <label class="md-nav__link" for="__nav_4_1_6" id="__nav_4_1_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⚙️ CHAPTER 6 机制建模
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_6">
            <span class="md-nav__icon md-icon"></span>
            ⚙️ CHAPTER 6 机制建模
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_regulatory_networks/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 Gene regulatory networks
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cell_cell_communication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 Cell-cell communication
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_7" >
        
          
          <label class="md-nav__link" for="__nav_4_1_7" id="__nav_4_1_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🔓 CHAPTER 7 反卷积
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_7">
            <span class="md-nav__icon md-icon"></span>
            🔓 CHAPTER 7 反卷积
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bulk_deconvolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 Bulk deconvolution
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    简明易懂的单细胞分析流程（使用 R 语言）
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            简明易懂的单细胞分析流程（使用 R 语言）
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%89%8D%E8%A8%80/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    前言
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    环境配置
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20%E6%9E%84%E5%BB%BA%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1 构建单细胞 RNA seq 分析环境
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20R%20%E7%8E%AF%E5%A2%83%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 R 环境安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Docker%20%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Docker 安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.4%20%E5%9C%A8%20Docker%20%E4%B8%AD%E5%90%AF%E5%8A%A8%20Rstudio/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.4 在 Docker 中启动 Rstudio
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.5%20%E6%9C%89%E5%85%B3%20Bioconductor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.5 有关 Bioconductor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20R%20%E7%9A%84%E5%9F%BA%E6%9C%AC%E7%94%A8%E6%B3%95/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 R 的基本用法
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20%E5%A6%82%E4%BD%95%E5%A4%84%E7%90%86%20DataFrame%EF%BC%9F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 如何处理 DataFrame？
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3%20%E4%BD%BF%E7%94%A8%20ggplot2%20%E5%8F%AF%E8%A7%86%E5%8C%96%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 使用 ggplot2 可视化数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1%20PCA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 PCA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2%20%E4%BD%BF%E7%94%A8%20clusterPlofiler%20%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 使用 clusterPlofiler 进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E5%9F%BA%E5%9B%A0%E6%9C%AC%E4%BD%93%EF%BC%88GO%EF%BC%89%E4%B8%AD%E5%90%84%20BP%E3%80%81MF%20%E5%92%8C%20CC%20%E7%9A%84%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 基因本体（GO）中各 BP、MF 和 CC 的富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E4%BD%BF%E7%94%A8%20DEseq2%20%E6%8F%90%E5%8F%96%20RNA-seq%20%E5%88%86%E6%9E%90%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%20DEG/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 使用 DEseq2 提取 RNA seq 分析数据中的 DEG
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.4%20%E4%BD%BF%E7%94%A8%E5%85%B6%E4%BB%96%E6%95%B0%E6%8D%AE%E5%BA%93%EF%BC%88KEGG%E3%80%81Reactome%20%E7%AD%89%EF%BC%89%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.4 使用其他数据库（KEGG、Reactome 等）进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.1%20scRNA-seq%20%E6%A6%82%E8%A6%81/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 scRNA seq 概要
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.2%20%E5%87%86%E5%A4%87%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 准备使用 Seurat 进行分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.3%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 使用 Seurat 进行数据预处理
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.4%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E7%9A%84%E4%B8%80%E8%88%AC%E9%A1%BA%E5%BA%8F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.4 使用 Seurat 进行数据分析的一般顺序
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.1%20%E5%A6%82%E4%BD%95%E6%9F%A5%E6%89%BE%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 如何查找单细胞 RNA seq 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.2%20%E4%B8%BA%E6%AF%8F%E4%B8%AA%E7%B0%87%E8%87%AA%E5%8A%A8%E6%B7%BB%E5%8A%A0%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E6%A0%87%E7%AD%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 为每个簇自动添加细胞注释标签
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.3%20%E4%BD%BF%E7%94%A8%20FeaturePlot%20%E6%A3%80%E6%9F%A5%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 使用 FeaturePlot 检查基因表达
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 将单细胞 RNA seq 数据集相互结合
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.5%20%E7%BA%A0%E6%AD%A3%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94%EF%BC%8C%E6%AF%94%E8%BE%83%E5%AF%B9%E7%85%A7%E7%BB%84%E5%92%8C%E5%88%BA%E6%BF%80%E7%BB%84/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.5 纠正批次效应，比较对照组和刺激组
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.6 确定每个群组的标记基因
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.1%20%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E7%BB%9F%E8%AE%A1%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E6%B0%B4%E5%B9%B3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 如何使用 Cell Ranger 统计 scRNA seq 数据中的基因表达水平
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.2%20%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E5%A4%84%E7%90%86%20NCBI%20Sequence%20Read%20Archive%20%28SRA%29%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 使用 Cell Ranger 处理 NCBI Sequence Read Archive (SRA) 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.1%20%E5%85%B3%E4%BA%8E%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 关于 scRNA seq 数据的质量控制
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.2%20%E4%BD%BF%E7%94%A8%20DoubleFinder%20%E6%A3%80%E6%B5%8B%E5%8F%8C%E9%87%8D%E7%BB%86%E8%83%9E/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.2 使用 DoubleFinder 检测双重细胞
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.3%20%E7%94%A8%20CellBender%20%E5%8E%BB%E9%99%A4%E8%83%8C%E6%99%AF%20RNA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.3 用 CellBender 去除背景 RNA
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧬 scRNA Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            🧬 scRNA Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🐍 Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            🐍 Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Large scale foundation model on single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scGPT%20toward%20building%20a%20foundation%20model%20for%20single-cell%20multi-omics%20using%20generative%20AI/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scGPT toward building a foundation model for single cell multi omics using generative AI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scBERT%20as%20a%20large-scale%20pretrained%20deep%20language%20model%20for%20cell%20type%20annotation%20of%20single-cell%20RNA-seq%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scBERT as a large scale pretrained deep language model for cell type annotation of single cell RNA seq data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RNA%20velocity%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA velocity of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TFvelo_gene%20regulation%20inspired%20RNA%20velocity%20estimation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TFvelo gene regulation inspired RNA velocity estimation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Disentanglement%20of%20single-cell%20data%20with%20biolord/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Disentanglement of single cell data with biolord
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Spatial%20transition%20tensor%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Spatial transition tensor of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRank%20infers%20drug-responsive%20cell%20types%20from%20untreated%20scRNA-seq%20data%20using%20a%20target-perturbed%20gene%20regulatory%20network/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRank infers drug responsive cell types from untreated scRNA seq data using a target perturbed gene regulatory network
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📔 文献
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            📔 文献
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ZNF689%20deficiency%20promotes%20intratumor%20heterogeneity%20and%20immunotherapy%20resistance%20in%20triple-negative%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ZNF689 deficiency promotes intratumor heterogeneity and immunotherapy resistance in triple negative breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Tumour%20vasculature%20at%20single-cell%20resolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tumour vasculature at single-cell resolution
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRNA-seq%20of%20gastric%20tumor%20shows%20complex%20intercellular%20interaction%20with%20an%20alternative%20T%20cell%20exhaustion%20trajectory%EF%BC%88%E5%8F%82%E8%80%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRNA seq of gastric tumor shows complex intercellular interaction with an alternative T cell exhaustion trajectory（参考）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20reveals%20new%20evolutionary%20complexity%20in%20uveal%20melanoma%EF%BC%88%E5%8F%82%E8%80%83%202%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis reveals new evolutionary complexity in uveal melanoma（参考 2）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20characterization%20of%20macrophages%20in%20uveal%20melanoma%20uncovers%20transcriptionally%20heterogeneous%20subsets%20conferring%20poor%20prognosis%20and%20aggressive%20behavior%EF%BC%88%E8%8C%83%E4%BE%8B%201%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell characterization of macrophages in uveal melanoma uncovers transcriptionally heterogeneous subsets conferring poor prognosis and aggressive behavior（范例 1）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ScRNA-seq%20of%20gastric%20cancer%20tissues%20reveals%20differences%20in%20the%20immune%20microenvironment%20of%20primary%20tumors%20and%20metastases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ScRNA seq of gastric cancer tissues reveals differences in the immune microenvironment of primary tumors and metastases
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Blimp-1%20and%20c-Maf%20regulate%20immune%20gene%20networks%20to%20protect%20against%20distinct%20pathways%20of%20pathobiont-induced%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Blimp 1 and c Maf regulate immune gene networks to protect against distinct pathways of pathobiont induced colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single%20cell%20analysis%20unveils%20B%20cell-dominated%20immune%20subtypes%20in%20HNSCC%20for%20enhanced%20prognostic%20and%20therapeutic%20stratification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis unveils B cell dominated immune subtypes in HNSCC for enhanced prognostic and therapeutic stratification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20spatial%20multi-omics%20and%20deep%20learning%20dissect%20enhancer-driven%20gene%20regulatory%20networks%20in%20liver%20zonation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell spatial multi omics and deep learning dissect enhancer driven gene regulatory networks in liver zonation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Neutrophil%20profiling%20illuminates%20anti-tumor%20antigen%20presenting%20potency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Neutrophil profiling illuminates anti tumor antigen presenting potency
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20blueprint%20for%20tumor-infiltrating%20B%20cells%20across%20human%20cancers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A blueprint for tumor infiltrating B cells across human cancers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20transcriptomic%20analyses%20reveal%20distinct%20immune%20cell%20contributions%20to%20epithelial%20barrier%20dysfunction%20in%20checkpoint%20inhibitor%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell transcriptomic analyses reveal distinct immune cell contributions to epithelial barrier dysfunction in checkpoint inhibitor colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Systematic%20dissection%20of%20tumor-normal%20single-cell%20ecosystems%20across%20a%20thousand%20tumors%20of%2030%20cancer%20types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Systematic dissection of tumor normal single cell ecosystems across a thousand tumors of 30 cancer types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../The%20aged%20tumor%20microenvironment%20limits%20T%20cell%20control%20of%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The aged tumor microenvironment limits T cell control of cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TIM-3%2B%20CD8%20T%20cells%20with%20a%20terminally%20exhausted%20phenotype%20retain%20functional%20capacity%20in%20hematological%20malignancies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TIM 3+ CD8 T cells with a terminally exhausted phenotype retain functional capacity in hematological malignancies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20lactate-SREBP2%20signaling%20axis%20drives%20tolerogenic%20dendritic%20cell%20maturation%20and%20promotes%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A lactate SREBP2 signaling axis drives tolerogenic dendritic cell maturation and promotes cancer progression
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20human%20omentum-specific%20mesothelial-like%20stromal%20population%20inhibits%20adipogenesis%20through%20IGFBP2%20secretion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A human omentum specific mesothelial like stromal population inhibits adipogenesis through IGFBP2 secretion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Interferon-stimulated%20neutrophils%20as%20a%20predictor%20of%20immunotherapy%20response/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interferon stimulated neutrophils as a predictor of immunotherapy response
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Repression%20of%20latent%20NF-%CE%BAB%20enhancers%20by%20PDX1%20regulates%20%CE%B2%20cell%20functional%20heterogeneity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Repression of latent NF κB enhancers by PDX1 regulates β cell functional heterogeneity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20systematic%20pan-cancer%20study%20on%20deep%20learning-based%20prediction%20of%20multi-omic%20biomarkers%20from%20routine%20pathology%20images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A systematic pan cancer study on deep learning based prediction of multi omic biomarkers from routine pathology images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20and%20spatial%20transcriptomics%20analysis%20of%20non-small%20cell%20lung%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell and spatial transcriptomics analysis of non small cell lung cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Seeing%20data%20as%20t-SNE%20and%20UMAP%20do/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Seeing data as t SNE and UMAP do
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mapping%20the%20cellular%20biogeography%20of%20human%20bone%20marrow%20niches%20using%20single-cell%20transcriptomics%20and%20proteomic%20imaging/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mapping the cellular biogeography of human bone marrow niches using single cell transcriptomics and proteomic imaging
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Onco-fetal%20Reprogramming%20of%20Endothelial%20Cells%20Drives%20Immunosuppressive%20Macrophages%20in%20Hepatocellular%20Carcinoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Onco fetal Reprogramming of Endothelial Cells Drives Immunosuppressive Macrophages in Hepatocellular Carcinoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20of%20anti-BCMA%20CAR%20T%20cell%20therapy%20in%20patients%20with%20central%20nervous%20system%20autoimmunity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis of anti BCMA CAR T cell therapy in patients with central nervous system autoimmunity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20sequencing%20depicts%20tumor%20architecture%20and%20empowers%20clinical%20decision%20in%20metastatic%20conjunctival%20melanoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell sequencing depicts tumor architecture and empowers clinical decision in metastatic conjunctival melanoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PR-SET7%20epigenetically%20restrains%20uterine%20interferon%20response%20and%20cell%20death%20governing%20proper%20postnatal%20stromal%20development/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PR SET7 epigenetically restrains uterine interferon response and cell death governing proper postnatal stromal development
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Clonal%20associations%20between%20lymphocyte%20subsets%20and%20functional%20states%20in%20rheumatoid%20arthritis%20synovium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clonal associations between lymphocyte subsets and functional states in rheumatoid arthritis synovium
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📐 Data Structures
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            📐 Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 1 章 python 入门
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%202%20%E7%AB%A0%20%E9%9D%A2%E5%90%91%E5%AF%B9%E8%B1%A1%E7%BC%96%E7%A8%8B/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 2 章 面向对象编程
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%203%20%E7%AB%A0%20%E7%AE%97%E6%B3%95%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 3 章 算法分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%204%20%E7%AB%A0%20%E9%80%92%E5%BD%92/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 4 章 递归
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%205%20%E7%AB%A0%20%E5%9F%BA%E4%BA%8E%E6%95%B0%E7%BB%84%E7%9A%84%E5%BA%8F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 5 章 基于数组的序列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%206%20%E7%AB%A0%20%E6%A0%88%E3%80%81%E9%98%9F%E5%88%97%E5%92%8C%E5%8F%8C%E7%AB%AF%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 6 章 栈、队列和双端队列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%207%20%E7%AB%A0%20%E9%93%BE%E8%A1%A8%28Linked%20Lists%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 7 章 链表(Linked Lists)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%208%20%E7%AB%A0%20%E6%A0%91%20%E2%98%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 8 章 树 ★
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%209%20%E7%AB%A0%20%E4%BC%98%E5%85%88%E7%BA%A7%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 9 章 优先级队列
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💡 Reviews
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            💡 Reviews
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_1" >
        
          
          <label class="md-nav__link" for="__nav_7_1" id="__nav_7_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧭 scRNA analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_1">
            <span class="md-nav__icon md-icon"></span>
            🧭 scRNA analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Best practices for single cell analysis across modalities
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Multi-omics%20integration%20in%20biomedical%20research%20%E2%80%93%20A%20metabolomics-centric%20review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multi omics integration in biomedical research – A metabolomics centric review
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Museum%20of%20spatial%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Museum of spatial transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20guide%20to%20artificial%20intelligence%20for%20cancer%20researchers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A guide to artificial intelligence for cancer researchers
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_2" >
        
          
          <label class="md-nav__link" for="__nav_7_2" id="__nav_7_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🦠 Metabolism
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_2">
            <span class="md-nav__icon md-icon"></span>
            🦠 Metabolism
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20analysis%20as%20a%20driver%20for%20discovery%2C%20diagnosis%2C%20and%20therapy/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic analysis as a driver for discovery, diagnosis, and therapy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20heterogeneity%20in%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic heterogeneity in cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Toward%20modeling%20metabolic%20state%20from%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Toward modeling metabolic state from single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20reprogramming%20and%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic reprogramming and cancer progression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_3" >
        
          
          <label class="md-nav__link" for="__nav_7_3" id="__nav_7_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Protein
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_3">
            <span class="md-nav__icon md-icon"></span>
            Protein
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Proteomic%20and%20interactomic%20insights%20into%20the%20molecular%20basis%20of%20cell%20functional%20diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Proteomic and interactomic insights into the molecular basis of cell functional diversity
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#141" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.1. 单细胞分析框架和联盟
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#142-anndata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2. 使用 AnnData 存储单模态数据
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.2. 使用 AnnData 存储单模态数据">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1421" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1422-anndata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.2. 初始化 AnnData 对象
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1423" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.3. 添加对齐的元数据
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.2.3. 添加对齐的元数据">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#14231" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.3.1. 观测值或变量级别
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14232" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.3.2. 使用元数据进行子集化
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1424" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.4. 观测值/变量级别矩阵
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1425" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.5. 非结构化元数据
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1426" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.6. 层
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1427-dataframe" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.7. 转换为 DataFrame
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1428-anndata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.8. 读取和写入 AnnData 对象
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1429" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.9. 高效数据访问
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.2.9. 高效数据访问">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#14291" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.9.1. 视图和副本
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14292" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.2.9.2. 大数据的部分读取
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#143-scanpy" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3. 使用 scanpy 进行单模态数据分析
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.3. 使用 scanpy 进行单模态数据分析">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1431" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1432-scanpy-api" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3.2. Scanpy API 设计
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1433-scanpy" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3.3. Scanpy 示例 #
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#144-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4. 进阶：使用 MuData 存储多模态数据
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.4. 进阶：使用 MuData 存储多模态数据">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1441" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1442-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.2. 初始化 MuData 对象
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1443-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.3. MuData 属性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1444" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.4. 变量映射
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1445-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.5. MuData 视图
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1446" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.6. 共同的观测值
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1447-mudata" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.7. MuData 对象的读取和写入
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1448" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.4.8. 多模态方法
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#145-muon" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5. 进阶：使用 muon 进行多模态数据分析
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4.5. 进阶：使用 muon 进行多模态数据分析">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1451" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5.1. 安装
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1452-api" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5.2. API 概览
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1453-muon-api" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5.3. muon API 演示
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/aletolia/aletolia.github.io.git/edit/master/docs/1.3 Analysis frameworks and tools.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/aletolia/aletolia.github.io.git/raw/master/docs/1.3 Analysis frameworks and tools.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


  <h1>1.3 Analysis frameworks and tools</h1>

<h2 id="141">1.4.1. 单细胞分析框架和联盟<a class="headerlink" href="#141" title="Permanent link">&para;</a></h2>
<p>在获得计数矩阵后，如前所述，探索性数据分析阶段开始。由于数据的规模和复杂性，需要专门的工具。虽然在单细胞分析的早期，人们使用自定义脚本分析数据，但现在已有专门的框架来应对这一目的。最受欢迎的三个选项是基于 R 的 Bioconductor[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id138" title="Wolfgang Huber, Vincent J. Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S. Carvalho, Hector Corrada Bravo, Sean Davis, Laurent Gatto, Thomas Girke, Raphael Gottardo, Florian Hahne, Kasper D. Hansen, Rafael A. Irizarry, Michael Lawrence, Michael I. Love, James MacDonald, Valerie Obenchain, Andrzej K. Oleś, Hervé Pagès, Alejandro Reyes, Paul Shannon, Gordon K. Smyth, Dan Tenenbaum, Levi Waldron, and Martin Morgan. Orchestrating high-throughput genomic analysis with bioconductor. Nature Methods, 12(2):115-121, Feb 2015. URL: https://doi.org/10.1038/nmeth.3252, doi:10.1038/nmeth.3252.">Huber <em>et al.</em>, 2015</a>] 和 Seurat[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id112" title="Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. Cell, 2021. URL: https://doi.org/10.1016/j.cell.2021.04.048, doi:10.1016/j.cell.2021.04.048.">Hao <em>et al.</em>, 2021</a>] 生态系统，以及基于 Python 的 scverse[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id113" title="scverse. Scverse. https://scverse.org, 2022. Accessed: 2022-04-21.">scverse, 2022</a>] 生态系统。这些框架不仅在使用的编程语言上不同，在底层数据结构和可用的专业分析工具上也有所区别。</p>
<p>Bioconductor 是一个项目，致力于开发、支持和分享专注于严格和可重复的多种生物检测数据分析的免费开源软件，包括单细胞分析。统一的开发者和用户体验以及丰富的文档和用户友好的小插曲是 Bioconductor 的最大优势。Seurat 是一个专门为单细胞数据分析设计的 R 包，提供了包括多模式和空间数据在内的所有分析步骤工具。Seurat 以其写得很好的小插曲和庞大的用户群而闻名。然而，这两个 R 选项在处理极大规模数据集（超过五十万个细胞）时可能会遇到可扩展性问题，这促使了基于 Python 的社区开发 scverse 生态系统。scverse 是一个致力于生命科学基础工具的组织和生态系统，最初专注于单细胞分析。可扩展性、可扩展性以及与现有 Python 数据和机器学习工具的强大互操作性是 scverse 生态系统的一些优势。</p>
<p>所有这三个生态系统都参与了许多努力，以实现涉及框架的互操作性。这将在“互操作性”章节中讨论。本书始终专注于针对相应问题的最佳工具，因此将针对具体问题使用上述生态系统的混合方法。然而，所有分析的基础将是 scverse 生态系统，原因有二：</p>
<ol>
<li>虽然我们将在本书中经常切换生态系统甚至编程语言，但一致使用数据结构和工具有助于读者专注于概念而非实现细节。</li>
<li><a href="https://bioconductor.org/books/release/OSCA/">一本关于 Bioconductor 生态系统的优秀书籍</a> 已经存在。我们鼓励只想学习 Bioconductor 单细胞分析的用户阅读它。</li>
</ol>
<p>在接下来的部分中，将更详细地介绍 scverse 生态系统，并重点解释最重要的数据结构。这一介绍无法涵盖数据结构和框架的所有方面。介绍所有可用分析功能也超出了范围。因此，我们在必要时会参考相关框架的教程和文档。</p>
<h2 id="142-anndata">1.4.2. 使用 AnnData 存储单模态数据<a class="headerlink" href="#142-anndata" title="Permanent link">&para;</a></h2>
<p>如前所述，在对基因组数据进行比对和基因注释后，通常将其汇总为特征矩阵。这个矩阵的形状为 <code>number_observations x number_variables</code>，对于 scRNA-seq 来说，观测值是细胞条形码，变量是注释基因。在分析过程中，这个矩阵的观测值和变量会被注释上计算得出的测量值（例如质量控制指标或潜在空间嵌入）以及先验知识（例如来源供体或替代基因标识符）。在 scverse 生态系统中，使用 AnnData[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id114" title="Isaac Virshup, Sergei Rybakov, Fabian J. Theis, Philipp Angerer, and F. Alexander Wolf. Anndata: annotated data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007, arXiv:https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007.full.pdf, doi:10.1101/2021.12.16.473007.">Virshup <em>et al.</em>, 2021</a>] 将数据矩阵与这些注释关联起来。为了实现快速和内存高效的转换，AnnData 还支持稀疏矩阵和部分读取。</p>
<p>虽然 AnnData 与 R 生态系统中的数据结构（例如 <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">Bioconductor 的 SummarizedExperiment</a> 或 <a href="https://mojaveazure.github.io/seurat-object/">Seurat 的对象</a>）大体相似，但 R 包使用的是转置的特征矩阵。</p>
<p>在其核心，AnnData 对象在 <code>X</code> 中存储一个稀疏或密集矩阵（在 scRNA-Seq 的情况下为计数矩阵）。这个矩阵的维度为 <code>obs_names x var_names</code>，其中 obs（=观测值）对应于细胞的条形码，var（=变量）对应于基因标识符。这个矩阵 <code>X</code> 被 Pandas 数据框 <code>obs</code> 和 <code>var</code> 包围，分别保存细胞和基因的注释。此外，AnnData 还保存了具有相应维度的观测值（<code>obsm</code>）或变量（<code>varm</code>）的完整计算矩阵。将细胞与细胞或基因与基因关联的图状结构通常保存在 <code>obsp</code> 和 <code>varp</code> 中。任何不适合其他插槽的非结构化数据都作为非结构化数据保存在 <code>uns</code> 中。还可以在 <code>layers</code> 中存储更多的 <code>X</code> 值。例如，未标准化的原始计数数据可以存储在 <code>counts</code> 层中，而标准化的数据存储在未命名的默认层中。</p>
<p>AnnData 主要设计用于单模态（例如仅 scRNA-Seq）数据。然而，本章后面会介绍的 AnnData 扩展，如 MuData，允许高效存储和访问多模态数据。</p>
<p><img alt="" src="https://www.sc-best-practices.org/_images/anndata.jpg" /></p>
<p>Fig. 4.1 AnnData overview. Image obtained from [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id114" title="Isaac Virshup, Sergei Rybakov, Fabian J. Theis, Philipp Angerer, and F. Alexander Wolf. Anndata: annotated data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007, arXiv:https://www.biorxiv.org/content/early/2021/12/19/2021.12.16.473007.full.pdf, doi:10.1101/2021.12.16.473007.">Virshup <em>et al.</em>, 2021</a>].<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#anndata-fig" title="Link to this image">#</a></p>
<h3 id="1421">1.4.2.1. 安装<a class="headerlink" href="#1421" title="Permanent link">&para;</a></h3>
<p>AnnData 可以通过 PyPI 和 Conda 安装，使用以下任意一种方法即可完成安装：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>pip<span class="w"> </span>install<span class="w"> </span>anndata
<span class="linenos" data-linenos="2 "></span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>anndata
</code></pre></div>
<h3 id="1422-anndata">1.4.2.2. 初始化 AnnData 对象<a class="headerlink" href="#1422-anndata" title="Permanent link">&para;</a></h3>
<p>本节内容参考了 AnnData 的“入门”教程：<a href="https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html">https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html</a></p>
<p>让我们创建一个包含稀疏计数信息的简单 AnnData 对象，这个对象可以表示基因表达计数。首先，我们导入所需的包。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="linenos" data-linenos="2 "></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="linenos" data-linenos="3 "></span><span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="linenos" data-linenos="4 "></span><span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">csr_matrix</span>
</code></pre></div>
<p>作为下一步，我们用随机泊松分布的数据初始化一个 AnnData 对象。一个不成文的规定是将分析的主要 AnnData 对象命名为 <code>adata</code>。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">counts</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="n">adata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
</code></pre></div>
<p>获得的 AnnData 对象有 100 个观测值和 2000 个变量。这相当于 100 个细胞和 2000 个基因。我们传入的初始数据可以作为一个稀疏矩阵，通过 <code>adata.X</code> 访问。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>&lt;100x2000 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
<span class="linenos" data-linenos="2 "></span>    with 126413 stored elements in Compressed Sparse Row format&gt;
</code></pre></div>
<p>现在，我们分别使用 <code>.obs_names</code> 和 <code>.var_names</code> 为 <code>obs</code> 和 <code>var</code> 轴提供索引。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Cell_</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)]</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Gene_</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)]</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Index([&#39;Cell_0&#39;, &#39;Cell_1&#39;, &#39;Cell_2&#39;, &#39;Cell_3&#39;, &#39;Cell_4&#39;, &#39;Cell_5&#39;, &#39;Cell_6&#39;,
<span class="linenos" data-linenos="2 "></span>       &#39;Cell_7&#39;, &#39;Cell_8&#39;, &#39;Cell_9&#39;],
<span class="linenos" data-linenos="3 "></span>      dtype=&#39;object&#39;)
</code></pre></div>
<h3 id="1423">1.4.2.3. 添加对齐的元数据<a class="headerlink" href="#1423" title="Permanent link">&para;</a></h3>
<h4 id="14231">1.4.2.3.1. 观测值或变量级别<a class="headerlink" href="#14231" title="Permanent link">&para;</a></h4>
<p>我们的 AnnData 对象的核心部分已经就位。下一步，我们在观测值和变量级别添加元数据。请记住，我们在 AnnData 对象的 <code>.obs</code> 和 <code>.var</code> 插槽中分别存储细胞和基因的注释。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">ct</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;Monocyte&quot;</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,))</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span>  <span class="c1"># Categoricals are preferred for efficiency</span>
<span class="linenos" data-linenos="3 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</code></pre></div>
<table><thead><tr><th></th><th>cell_type</th></tr></thead><tbody><tr><th>Cell_0</th><td>B</td></tr><tr><th>Cell_1</th><td>T</td></tr><tr><th>Cell_2</th><td>B</td></tr><tr><th>Cell_3</th><td>T</td></tr><tr><th>Cell_4</th><td>Monocyte</td></tr><tr><th>...</th><td>...</td></tr><tr><th>Cell_95</th><td>T</td></tr><tr><th>Cell_96</th><td>T</td></tr><tr><th>Cell_97</th><td>B</td></tr><tr><th>Cell_98</th><td>T</td></tr><tr><th>Cell_99</th><td>Monocyte</td></tr></tbody></table>

<p>100 rows × 1 columns</p>
<p>如果我们现在再次检查 AnnData 对象的表示，会注意到它也更新了 <code>obs</code> 中的 <code>cell_type</code> 信息。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;cell_type&#39;
</code></pre></div>
<h4 id="14232">1.4.2.3.2. 使用元数据进行子集化<a class="headerlink" href="#14232" title="Permanent link">&para;</a></h4>
<p>我们还可以使用随机生成的细胞类型对 AnnData 对象进行子集化。AnnData 对象的切片和掩码操作与 Pandas DataFrame 或 R 矩阵的数据访问类似。更多详细信息见下文。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">bdata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span> <span class="o">==</span> <span class="s2">&quot;B&quot;</span><span class="p">]</span>
<span class="linenos" data-linenos="2 "></span><span class="n">bdata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>View of AnnData object with n_obs × n_vars = 26 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;cell_type&#39;
</code></pre></div>
<h3 id="1424">1.4.2.4. 观测值/变量级别矩阵<a class="headerlink" href="#1424" title="Permanent link">&para;</a></h3>
<p>我们可能还会有多维度的元数据，比如数据的 UMAP 嵌入。对于这种类型的元数据，AnnData 有 <code>.obsm/.varm</code> 属性。我们使用键来标识插入的不同矩阵。<code>.obsm/.varm</code> 的限制是，<code>.obsm</code> 矩阵的长度必须等于观测值的数量，即 <code>.n_obs</code>，而 <code>.varm</code> 矩阵的长度必须等于变量的数量，即 <code>.n_vars</code>。它们各自可以有不同的维度。</p>
<p>让我们从一个随机生成的矩阵开始，这个矩阵可以解释为我们希望存储的数据的 UMAP 嵌入，以及一些随机生成的基因级别元数据。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="s2">&quot;gene_stuff&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="linenos" data-linenos="3 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AxisArrays with keys: X_umap
</code></pre></div>
<p>再次，AnnData 对象的表示已更新。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;cell_type&#39;
<span class="linenos" data-linenos="3 "></span>    obsm: &#39;X_umap&#39;
<span class="linenos" data-linenos="4 "></span>    varm: &#39;gene_stuff&#39;
</code></pre></div>
<p>关于 <code>.obsm/.varm</code> 的几点说明：</p>
<ol>
<li>“类似数组”的元数据可以来自 Pandas DataFrame、scipy 稀疏矩阵或 numpy 密集数组。</li>
<li>使用 scanpy 时，它们的值（列）不易绘制，相反，来自 <code>.obs</code> 的项可以轻松绘制在例如 UMAP 图上。</li>
</ol>
<h3 id="1425">1.4.2.5. 非结构化元数据<a class="headerlink" href="#1425" title="Permanent link">&para;</a></h3>
<p>如上所述，AnnData 有 <code>.uns</code>，允许存储任何非结构化元数据。这可以是任何内容，例如在数据分析中有用的某些常规信息的列表或字典。尽量只将无法有效存储在其他插槽中的数据存储在此插槽中。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;random&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>OverloadedDict, wrapping:
<span class="linenos" data-linenos="2 "></span>    OrderedDict([(&#39;random&#39;, [1, 2, 3])])
<span class="linenos" data-linenos="3 "></span>With overloaded keys:
<span class="linenos" data-linenos="4 "></span>    [&#39;neighbors&#39;].
</code></pre></div>
<h3 id="1426">1.4.2.6. 层<a class="headerlink" href="#1426" title="Permanent link">&para;</a></h3>
<p>最后，我们可能会有原始核心数据的不同形式，比如一个是标准化的，一个是未标准化的。这些可以存储在 AnnData 的不同层中。例如，让我们对原始数据进行对数转换并将其存储在一个层中。.</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log_transformed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;cell_type&#39;
<span class="linenos" data-linenos="3 "></span>    uns: &#39;random&#39;
<span class="linenos" data-linenos="4 "></span>    obsm: &#39;X_umap&#39;
<span class="linenos" data-linenos="5 "></span>    varm: &#39;gene_stuff&#39;
<span class="linenos" data-linenos="6 "></span>    layers: &#39;log_transformed&#39;
</code></pre></div>
<p>我们的原始矩阵 <code>X</code> 未被修改，仍然可以访问。我们可以通过比较原始 <code>X</code> 和新的层来验证这一点。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">!=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log_transformed&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">nnz</span> <span class="o">==</span> <span class="mi">0</span>
</code></pre></div>
<h3 id="1427-dataframe">1.4.2.7. 转换为 DataFrame<a class="headerlink" href="#1427-dataframe" title="Permanent link">&para;</a></h3>
<p>可以从某一层中获取一个 Pandas DataFrame。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s2">&quot;log_transformed&quot;</span><span class="p">)</span>
</code></pre></div>
<table><thead><tr><th></th><th>Gene_0</th><th>Gene_1</th><th>Gene_2</th><th>Gene_3</th><th>Gene_4</th><th>Gene_5</th><th>Gene_6</th><th>Gene_7</th><th>Gene_8</th><th>Gene_9</th><th>...</th><th>Gene_1990</th><th>Gene_1991</th><th>Gene_1992</th><th>Gene_1993</th><th>Gene_1994</th><th>Gene_1995</th><th>Gene_1996</th><th>Gene_1997</th><th>Gene_1998</th><th>Gene_1999</th></tr></thead><tbody><tr><th>Cell_0</th><td>0.693147</td><td>0.000000</td><td>1.386294</td><td>1.386294</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>1.098612</td><td>0.000000</td><td>1.098612</td><td>...</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>1.098612</td><td>0.000000</td><td>0.000000</td><td>1.098612</td><td>0.693147</td><td>0.000000</td></tr><tr><th>Cell_1</th><td>1.098612</td><td>1.098612</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>1.386294</td><td>0.000000</td><td>...</td><td>1.098612</td><td>1.386294</td><td>1.098612</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>1.386294</td><td>0.000000</td><td>0.000000</td><td>0.693147</td></tr><tr><th>Cell_2</th><td>0.693147</td><td>0.693147</td><td>1.098612</td><td>0.693147</td><td>1.386294</td><td>1.386294</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>1.098612</td><td>...</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>1.386294</td><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.693147</td></tr><tr><th>Cell_3</th><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>1.386294</td><td>1.098612</td><td>0.693147</td><td>0.693147</td><td>...</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>1.386294</td><td>0.693147</td></tr><tr><th>Cell_4</th><td>0.000000</td><td>0.693147</td><td>0.000000</td><td>1.609438</td><td>1.098612</td><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.000000</td><td>...</td><td>1.098612</td><td>1.098612</td><td>0.000000</td><td>1.386294</td><td>1.791759</td><td>1.098612</td><td>1.609438</td><td>0.693147</td><td>0.000000</td><td>0.693147</td></tr><tr><th>...</th><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr><tr><th>Cell_95</th><td>0.693147</td><td>1.098612</td><td>0.693147</td><td>1.386294</td><td>1.098612</td><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>...</td><td>0.000000</td><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>1.386294</td><td>0.693147</td></tr><tr><th>Cell_96</th><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>1.098612</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>1.098612</td><td>0.000000</td><td>0.000000</td><td>...</td><td>1.386294</td><td>0.693147</td><td>1.098612</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>1.098612</td><td>1.386294</td><td>0.693147</td><td>0.000000</td></tr><tr><th>Cell_97</th><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>1.098612</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>...</td><td>0.000000</td><td>0.693147</td><td>1.386294</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>1.609438</td><td>0.693147</td></tr><tr><th>Cell_98</th><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>1.098612</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>1.098612</td><td>0.000000</td><td>0.000000</td><td>...</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>0.693147</td><td>1.386294</td><td>0.693147</td><td>1.098612</td><td>1.386294</td></tr><tr><th>Cell_99</th><td>1.098612</td><td>1.791759</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>0.000000</td><td>1.098612</td><td>1.609438</td><td>0.693147</td><td>0.693147</td><td>...</td><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>0.693147</td><td>0.693147</td><td>0.693147</td><td>0.693147</td><td>0.000000</td><td>0.000000</td><td>1.386294</td></tr></tbody></table>

<p>100 rows × 2000 columns</p>
<h3 id="1428-anndata">1.4.2.8. 读取和写入 AnnData 对象<a class="headerlink" href="#1428-anndata" title="Permanent link">&para;</a></h3>
<p>AnnData 对象可以保存到磁盘上的分层数组存储，如 <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">HDF5</a> 或 <a href="https://zarr.readthedocs.io/en/stable/index.html">Zarr</a>，以便在磁盘和内存中实现类似的结构。AnnData 有自己基于 HDF5 的持久文件格式：<code>h5ad</code>。如果具有少量类别的字符串列尚未转换为分类变量，AnnData 将自动将它们转换为分类变量。现在我们将把 AnnData 对象保存为 <code>h5ad</code> 格式。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;my_results.h5ad&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>X                        Group
<span class="linenos" data-linenos="2 "></span>layers                   Group
<span class="linenos" data-linenos="3 "></span>obs                      Group
<span class="linenos" data-linenos="4 "></span>obsm                     Group
<span class="linenos" data-linenos="5 "></span>uns                      Group
<span class="linenos" data-linenos="6 "></span>var                      Group
<span class="linenos" data-linenos="7 "></span>varm                     Group
</code></pre></div>
<p>… and read it back in.</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata_new</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;my_results.h5ad&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata_new</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;cell_type&#39;
<span class="linenos" data-linenos="3 "></span>    uns: &#39;random&#39;
<span class="linenos" data-linenos="4 "></span>    obsm: &#39;X_umap&#39;
<span class="linenos" data-linenos="5 "></span>    varm: &#39;gene_stuff&#39;
<span class="linenos" data-linenos="6 "></span>    layers: &#39;log_transformed&#39;
</code></pre></div>
<h3 id="1429">1.4.2.9. 高效数据访问<a class="headerlink" href="#1429" title="Permanent link">&para;</a></h3>
<h4 id="14291">1.4.2.9.1. 视图和副本<a class="headerlink" href="#14291" title="Permanent link">&para;</a></h4>
<p>我们来看另一个元数据使用案例。假设观测值来自多个年度研究中的仪器，这些研究使用来自不同地点的不同受试者的样本来表征 10 个读数。我们通常会以某种格式获取这些信息，然后将其存储在 DataFrame 中。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">obs_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="linenos" data-linenos=" 2 "></span>    <span class="p">{</span>
<span class="linenos" data-linenos=" 3 "></span>        <span class="s2">&quot;time_yr&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span>
<span class="linenos" data-linenos=" 4 "></span>        <span class="s2">&quot;subject_id&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
<span class="linenos" data-linenos=" 5 "></span>            <span class="p">[</span><span class="s2">&quot;subject 1&quot;</span><span class="p">,</span> <span class="s2">&quot;subject 2&quot;</span><span class="p">,</span> <span class="s2">&quot;subject 4&quot;</span><span class="p">,</span> <span class="s2">&quot;subject 8&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span>
<span class="linenos" data-linenos=" 6 "></span>        <span class="p">),</span>
<span class="linenos" data-linenos=" 7 "></span>        <span class="s2">&quot;instrument_type&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;type a&quot;</span><span class="p">,</span> <span class="s2">&quot;type b&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span>
<span class="linenos" data-linenos=" 8 "></span>        <span class="s2">&quot;site&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;site x&quot;</span><span class="p">,</span> <span class="s2">&quot;site y&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span>
<span class="linenos" data-linenos=" 9 "></span>    <span class="p">},</span>
<span class="linenos" data-linenos="10 "></span>    <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>  <span class="c1"># these are the same IDs of observations as above!</span>
<span class="linenos" data-linenos="11 "></span><span class="p">)</span>
</code></pre></div>
<p>这就是我们如何将读数数据与元数据结合起来。当然，以下调用中第一个参数 <em>X</em> 也可以只是一个 DataFrame。这将导致一个跟踪所有内容的单一数据容器。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">obs_meta</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</code></pre></div>
<p>对联合数据矩阵进行子集化可以帮助我们关注变量或观测值的子集，或定义机器学习模型的训练 - 测试拆分。</p>
<p><strong>注意</strong> 类似于 numpy 数组，AnnData 对象可以包含实际数据或引用另一个 <code>AnnData</code> 对象。在后一种情况下，它们被称为“视图”。子集化 AnnData 对象总是返回视图，这有两个优点：</p>
<ul>
<li>不会分配新的内存</li>
<li>可以修改底层的 AnnData 对象</li>
</ul>
<p>您可以通过对视图调用 <code>.copy()</code> 来获取实际的 AnnData 对象。通常，这是不必要的，因为对视图元素的任何修改（在视图的属性上调用 <code>.[]</code>）内部都会调用 <code>.copy()</code> 并使视图成为包含实际数据的 AnnData 对象。请参见下面的示例。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</code></pre></div>
<p><strong>注意</strong> 对 AnnData 的索引将假定传递给 <code>[]</code> 的整数参数的行为类似于 pandas 中的 <code>.iloc</code>，而字符串参数的行为类似于 <code>.loc</code>。AnnData 总是假定字符串索引。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata_view</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene_3&quot;</span><span class="p">]]</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata_view</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>View of AnnData object with n_obs × n_vars = 5 × 2
<span class="linenos" data-linenos="2 "></span>    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</code></pre></div>
<p>这是一个视图！这可以通过再次检查 AnnData 对象来验证。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 100 × 2000
<span class="linenos" data-linenos="2 "></span>    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</code></pre></div>
<p>AnnData 对象的维度没有改变，它仍然包含相同的数据。如果我们想要一个在内存中保存数据的 AnnData 对象，我们需要调用 <code>.copy()</code>。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata_subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene_3&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</code></pre></div>
<p>对于视图，我们也可以设置某一列的前 3 个元素。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Gene_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Gene_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Gene_1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>[[0.0], [2.0], [1.0]]
<span class="linenos" data-linenos="2 "></span>[[0.0], [0.0], [0.0]]
</code></pre></div>
<p>如果您尝试访问 AnnData 视图的某些部分，其内容将自动复制，并生成一个存储数据的对象。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata_subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">3</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;Gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene_2&quot;</span><span class="p">]]</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata_subset</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>View of AnnData object with n_obs × n_vars = 3 × 2
<span class="linenos" data-linenos="2 "></span>    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata_subset</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;foo&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Trying to set attribute `.obs` of view, copying.
</code></pre></div>
<p>现在 <code>adata_subset</code> 存储了实际数据，不再只是 <code>adata</code> 的引用。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 3 × 2
<span class="linenos" data-linenos="2 "></span>    obs: &#39;time_yr&#39;, &#39;subject_id&#39;, &#39;instrument_type&#39;, &#39;site&#39;, &#39;foo&#39;
</code></pre></div>
<p>显然，您可以使用 pandas 的所有功能来使用序列或布尔索引进行切片。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">time_yr</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">])]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<table><thead><tr><th></th><th>time_yr</th><th>subject_id</th><th>instrument_type</th><th>site</th></tr></thead><tbody><tr><th>Cell_0</th><td>4</td><td>subject 8</td><td>type b</td><td>site x</td></tr><tr><th>Cell_2</th><td>2</td><td>subject 2</td><td>type b</td><td>site y</td></tr><tr><th>Cell_3</th><td>4</td><td>subject 8</td><td>type a</td><td>site y</td></tr><tr><th>Cell_5</th><td>2</td><td>subject 2</td><td>type a</td><td>site y</td></tr><tr><th>Cell_10</th><td>2</td><td>subject 8</td><td>type b</td><td>site x</td></tr></tbody></table>

<h4 id="14292">1.4.2.9.2. 大数据的部分读取<a class="headerlink" href="#14292" title="Permanent link">&para;</a></h4>
<p>如果单个 <code>h5ad</code> 文件非常大，可以使用支持模式（backed mode）或当前实验性的 <a href="https://anndata.readthedocs.io/en/latest/generated/anndata.experimental.read_elem.html">read_elem</a> API 将其部分读取到内存中。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;my_results.h5ad&quot;</span><span class="p">,</span> <span class="n">backed</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
</code></pre></div>
<p>如果这样做，您需要记住 AnnData 对象与用于读取的文件保持打开连接。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">PosixPath</span><span class="p">(</span><span class="s1">&#39;my_results.h5ad&#39;</span><span class="p">)</span>
</code></pre></div>
<p>由于我们在只读模式下使用它，因此不会损坏任何内容。要继续本教程，我们仍然需要明确关闭它。</p>
<h2 id="143-scanpy">1.4.3. 使用 scanpy 进行单模态数据分析<a class="headerlink" href="#143-scanpy" title="Permanent link">&para;</a></h2>
<p>现在我们已经了解了单模态单细胞分析的基本数据结构，接下来的问题是我们如何实际分析存储的数据。在 scverse 生态系统中，存在用于特定组学数据分析的多种工具。例如，scanpy[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id115" title="F. Alexander Wolf, Philipp Angerer, and Fabian J. Theis. Scanpy: large-scale single-cell gene expression data analysis. Genome Biology, 19(1):15, Feb 2018. URL: https://doi.org/10.1186/s13059-017-1382-0, doi:10.1186/s13059-017-1382-0.">Wolf <em>et al.</em>, 2018</a>] 提供了面向一般 RNA-Seq 分析的工具，squidpy[<a href="https://www.sc-best-practices.org/spatial/spatially_variable_genes.html#id105" title="Giovanni Palla, Hannah Spitzer, Michal Klein, David Fischer, Anna Christina Schaar, Louis Benedikt Kuemmerle, Sergei Rybakov, Ignacio L. Ibarra, Olle Holmberg, Isaac Virshup, Mohammad Lotfollahi, Sabrina Richter, and Fabian J. Theis. Squidpy: a scalable framework for spatial omics analysis. Nature Methods, 19(2):171-178, Feb 2022. URL: https://doi.org/10.1038/s41592-021-01358-2, doi:10.1038/s41592-021-01358-2.">Palla <em>et al.</em>, 2022</a>] 专注于空间转录组学，scirpy[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id117" title="Gregor Sturm, Tamas Szabo, Georgios Fotakis, Marlene Haider, Dietmar Rieder, Zlatko Trajanoski, and Francesca Finotello. Scirpy: a Scanpy extension for analyzing single-cell T-cell receptor-sequencing data. Bioinformatics, 36(18):4817-4818, 07 2020. URL: https://doi.org/10.1093/bioinformatics/btaa611, arXiv:https://academic.oup.com/bioinformatics/article-pdf/36/18/4817/34560298/btaa611.pdf, doi:10.1093/bioinformatics/btaa611.">Sturm <em>et al.</em>, 2020</a>] 提供了用于 T 细胞受体（TCR）和 B 细胞受体（BCR）数据分析的工具。尽管存在多种用于各种数据模态的 scverse 扩展，但它们通常在某种程度上使用 scanpy 的预处理和可视化功能。</p>
<p>更具体地说，scanpy 是一个基于 AnnData 构建的 Python 包，用于促进单细胞基因表达数据的分析。通过 scanpy，可以访问多种方法进行预处理、嵌入、可视化、聚类、差异基因表达测试、拟时间和轨迹推断以及基因调控网络模拟。基于 Python 数据科学和机器学习库的高效实现使 scanpy 能够扩展到数百万个细胞。</p>
<p>一般来说，最佳实践的单细胞数据分析是一个互动过程。许多决策和分析步骤依赖于前一步的结果以及实验合作伙伴的输入。scflow<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id120" title="Combiz Khozoie, Nurun Fancy, Mahdi M. Marjaneh, Alan E. Murphy, Paul M. Matthews, and Nathan Skene. Scflow: a scalable and reproducible analysis pipeline for single-cell termrna sequencing data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1, arXiv:https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf, doi:10.1101/2021.08.16.456499.">Khozoie <em>et al.</em>, 2021</a>] 等管道完全自动化了一些下游分析步骤，这类管道直到最近才开始出现。这些管道必须做出假设和简化，可能不会导致最可靠的分析。因此，scanpy 设计用于与例如 Jupyter Notebooks<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id121" title="Project Jupyter. Jupyter. https://jupyter.org/, 2022. Accessed: 2022-04-21.">Jupyter, 2022</a>] 等进行交互分析。</p>
<p><img alt="" src="https://www.sc-best-practices.org/_images/scanpy.jpg" /></p>
<p>Fig. 4.2 scanpy overview. Image obtained from [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id115" title="F. Alexander Wolf, Philipp Angerer, and Fabian J. Theis. Scanpy: large-scale single-cell gene expression data analysis. Genome Biology, 19(1):15, Feb 2018. URL: https://doi.org/10.1186/s13059-017-1382-0, doi:10.1186/s13059-017-1382-0.">Wolf <em>et al.</em>, 2018</a>].<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#scanpy-fig" title="Link to this image">#</a></p>
<h3 id="1431">1.4.3.1. 安装<a class="headerlink" href="#1431" title="Permanent link">&para;</a></h3>
<p>scanpy 可以通过 PyPI 和 Conda 安装，使用以下任意一种方法即可完成安装：</p>
<ol>
<li>使用 PyPI 安装。</li>
<li>使用 Conda 安装。</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>pip<span class="w"> </span>install<span class="w"> </span>scanpy
<span class="linenos" data-linenos="2 "></span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>scanpy
</code></pre></div>
<h3 id="1432-scanpy-api">1.4.3.2. Scanpy API 设计<a class="headerlink" href="#1432-scanpy-api" title="Permanent link">&para;</a></h3>
<p>scanpy 框架的设计方式是将属于同一步骤的功能分组在相应的模块中。例如，所有预处理功能都在 <code>scanpy.preprocessing</code> 模块中，所有不是预处理的数据矩阵转换功能都在 <code>scanpy.tools</code> 中，所有可视化功能都在 <code>scanpy.plot</code> 中。这些模块通常在导入 scanpy 后访问，如 <code>import scanpy as sc</code>，相应的缩写为 <code>sc.pp</code> 代表预处理，<code>sc.tl</code> 代表工具，<code>sc.pl</code> 代表绘图。所有读取或写入数据的模块都是直接访问的。此外，一个用于各种数据集的模块可以作为 <code>sc.datasets</code> 访问。所有函数及其相应的参数和示例图都在 scanpy API 文档中记录 [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id122" title="scverse scanpy. Scanpy api. https://scanpy.readthedocs.io/en/stable/api.html#, 2022. Accessed: 2022-04-21.">scverse scanpy, 2022</a>].</p>
<p>请注意，本教程仅涵盖了 scanpy 功能和选项的一小部分。强烈建议读者查阅 <a href="https://scanpy.readthedocs.io/">scanpy 的文档</a> 以获取更多详细信息。</p>
<p><img alt="" src="https://www.sc-best-practices.org/_images/scanpy_api.png" /></p>
<p>Fig. 4.3 scanpy API overview. The API is divided into datasets, preprocessing (pp), tools (tl) and corresponding plotting (pl) functions.<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#scanpy-api" title="Link to this image">#</a></p>
<h3 id="1433-scanpy">1.4.3.3. Scanpy 示例 <a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#scanpy-example" title="链接到此标题">#</a><a class="headerlink" href="#1433-scanpy" title="Permanent link">&para;</a></h3>
<p>在接下来的内容中，我们将简要演示使用 scanpy 进行分析的工作流程。我们不会进行完整的分析，因为具体的分析步骤在相应的章节中介绍。</p>
<p>所选数据集是来自健康供体的 2700 个外周血单个核细胞的数据集，这些细胞在 Illumina NextSeq 500 上进行测序。</p>
<p>第一步，我们导入 scanpy 并为接下来的快速 scanpy 演示定义默认设置。我们使用 scanpy 的设置对象来为所有 scanpy 的绘图设置 Matplotlib 绘图默认值，最后打印 scanpy 的头部信息。这个头部信息包含当前环境中所有相关 Python 包的版本，包括 Scanpy 和 AnnData。这个输出在报告错误给 scverse 团队时特别有用，也有助于结果的可重复性。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="linenos" data-linenos="2 "></span>
<span class="linenos" data-linenos="3 "></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span><span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>scanpy==1.9.1 anndata==0.7.8 umap==0.5.2 numpy==1.20.3 scipy==1.7.2 pandas==1.4.3 scikit-learn==1.0.1 statsmodels==0.13.1 python-igraph==0.9.10 pynndescent==0.5.5
</code></pre></div>
<p>接下来，我们使用 scanpy 获取 3k PBMC 数据集作为 AnnData 对象。该数据集会自动下载并保存到路径中，这个路径可以通过 scanpy 的设置对象指定。默认情况下，该路径是文件夹 <code>data</code>。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k</span><span class="p">()</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>100%|██████████| 5.58M/5.58M [00:01&lt;00:00, 3.25MB/s]
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 2700 × 32738
<span class="linenos" data-linenos="2 "></span>    var: &#39;gene_ids&#39;
</code></pre></div>
<p>返回的 AnnData 对象包含 2700 个细胞和 32738 个基因。<code>var</code> 插槽中进一步包含基因 ID。</p>
<table><thead><tr><th></th><th>gene_ids</th></tr><tr><th>index</th><th></th></tr></thead><tbody><tr><th>MIR1302-10</th><td>ENSG00000243485</td></tr><tr><th>FAM138A</th><td>ENSG00000237613</td></tr><tr><th>OR4F5</th><td>ENSG00000186092</td></tr><tr><th>RP11-34P13.7</th><td>ENSG00000238009</td></tr><tr><th>RP11-34P13.8</th><td>ENSG00000239945</td></tr><tr><th>...</th><td>...</td></tr><tr><th>AC145205.1</th><td>ENSG00000215635</td></tr><tr><th>BAGE5</th><td>ENSG00000268590</td></tr><tr><th>CU459201.1</th><td>ENSG00000251180</td></tr><tr><th>AC002321.2</th><td>ENSG00000215616</td></tr><tr><th>AC002321.1</th><td>ENSG00000215611</td></tr></tbody></table>

<p>32738 rows × 1 columns</p>
<p>如上所述，所有 scanpy 的分析功能都可以通过 <code>sc.[pp, tl, pl]</code> 访问。为了首先概览我们的数据，我们使用 scanpy 显示在每个细胞中产生最高计数分数的基因，跨所有细胞。我们简单地调用 <code>sc.pl.highest_expr_genes</code> 函数，传递 AnnData 对象（在几乎所有情况下都是任何 scanpy 函数的第一个参数），并指定我们希望显示表达量最高的 20 个基因。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/c0fb293bb4fa88a2d4d6fb9899ed3a3529134f9475e96ba5e1f1f08dbbfd88b9.png" /></p>
<p>显然，MALAT1 是表达量最高的基因，这在聚合酶捕获的 scRNA-Seq 数据中经常检测到，与实验方案无关。研究表明，这个基因与细胞健康呈负相关，尤其是死亡/濒死细胞的 MALAT1 表达量更高。</p>
<p>现在，我们使用 scanpy 的预处理模块过滤出检测到的基因少于 200 的细胞，以及在少于 3 个细胞中发现的基因，以达到粗略的质量阈值。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div>
<p>在单细胞 RNA-Seq 分析中，一个常见的步骤是使用例如主成分分析（PCA）进行降维，以揭示主要的变异轴。这也有助于数据去噪。scanpy 提供了 PCA 作为 <code>preprocessing</code> 或 <code>tools</code> 函数，这两者是等价的。在这里，我们使用 <code>tools</code> 中的版本，没有特别的原因。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
</code></pre></div>
<p>相应的绘图函数允许我们将基因传递给 color 参数。相应的值会自动从 AnnData 对象中提取。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;CST3&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/6a951bc7f5c43678cb8758b0bdd977f1299d3fb9c3071ff7a42a14a9804f406e.png" /></p>
<p>任何高级嵌入和下游计算的一个基本步骤是使用数据矩阵的 PCA 表示计算邻域图。这一步会自动用于需要它的其他工具，例如 UMAP 的计算。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
</code></pre></div>
<p>现在，我们使用计算出的邻域图通过 UMAP 对细胞进行嵌入，UMAP 是 scanpy 实现的众多高级降维算法之一。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CST3&quot;</span><span class="p">,</span> <span class="s2">&quot;NKG7&quot;</span><span class="p">,</span> <span class="s2">&quot;PPBP&quot;</span><span class="p">])</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/90f4f980c81308de5a7a796b38f77c3083980eedeff91e0c3647136003104860.png" /></p>
<p>scanpy 的文档还提供了 <a href="https://scanpy.readthedocs.io/en/stable/tutorials.html">教程</a>，我们推荐给所有需要复习 scanpy 或新接触 scanpy 的读者。视频教程可在 <a href="https://www.youtube.com/channel/UCpsvsIAW3R5OdftJKKuLNMA">scverse YouTube 频道</a> 上观看。</p>
<h2 id="144-mudata">1.4.4. 进阶：使用 MuData 存储多模态数据<a class="headerlink" href="#144-mudata" title="Permanent link">&para;</a></h2>
<p>AnnData 主要设计用于存储和处理单模态数据。然而，多模态检测如 CITE-Seq 通过同时测量 RNA 和表面蛋白生成多模态数据。这种数据需要更高级的存储方式，这就是 MuData 的作用所在。MuData 构建在 AnnData 之上，用于存储和处理多模态数据。muon[<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id125" title="Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. Genome Biology, 23(1):42, Feb 2022. URL: https://doi.org/10.1186/s13059-021-02577-8, doi:10.1186/s13059-021-02577-8.">Bredikhin <em>et al.</em>, 2022</a>]，一个“scanpy 等价物”和 scverse 的核心包，可以用于分析多模态组学数据。以下部分基于 MuData 快速入门教程 [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id126" title="scverse mudata. Mudata quickstart. https://mudata.readthedocs.io/en/latest/notebooks/quickstart_mudata.html, 2022. Accessed: 2022-04-21.">scverse mudata, 2022</a>] 和多模态数据对象教程 [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id127" title="scverse mudata. Mudata quickstart. https://mudata.readthedocs.io/en/latest/io/mudata.html, 2022. Accessed: 2022-04-21.">scverse mudata, 2022</a>].</p>
<p><img alt="" src="https://www.sc-best-practices.org/_images/mudata.png" /></p>
<p>Fig. 4.4 MuData 概览。图片来源于 [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id125" title="Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. Genome Biology, 23(1):42, Feb 2022. URL: https://doi.org/10.1186/s13059-021-02577-8, doi:10.1186/s13059-021-02577-8.">Bredikhin <em>et al.</em>, 2022</a>].<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#mudata-fig" title="链接到此图片">#</a></p>
<h3 id="1441">1.4.4.1. 安装<a class="headerlink" href="#1441" title="Permanent link">&para;</a></h3>
<p>MuData 可以通过 PyPI 和 Conda 安装，使用以下任意一种方法即可完成安装：</p>
<ol>
<li>使用 PyPI 安装。</li>
<li>使用 Conda 安装。</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>pip<span class="w"> </span>install<span class="w"> </span>mudata
<span class="linenos" data-linenos="2 "></span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>mudata
</code></pre></div>
<p>MuData 的主要思想是，MuData 对象包含对单模态数据的单个 AnnData 对象的引用，但 MuData 对象本身也存储多模态注释。因此，可以直接访问 AnnData 对象以执行单模态数据转换，并将结果存储在相应的 AnnData 注释中，也可以聚合模态进行联合计算，并将结果存储在全局 MuData 对象中。从技术上讲，这通过 MuData 对象的 <code>.mod</code>（=模态）属性中的包含 AnnData 对象的字典来实现，每个模态对应一个 AnnData 对象。就像 AnnData 对象本身一样，它们还包含诸如 <code>.obs</code>（观测值注释，如样本或细胞）和 <code>.obsm</code>（多维注释，如嵌入）等属性。</p>
<h3 id="1442-mudata">1.4.4.2. 初始化 MuData 对象<a class="headerlink" href="#1442-mudata" title="Permanent link">&para;</a></h3>
<p>我们将从 mudata 包中导入 MuData 开始。</p>
<p>要创建一个示例 MuData 对象，我们需要模拟数据。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">n</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">10</span>
<span class="linenos" data-linenos="2 "></span>
<span class="linenos" data-linenos="3 "></span><span class="n">z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">k</span><span class="p">),</span> <span class="n">scale</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
<span class="linenos" data-linenos="4 "></span><span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
<span class="linenos" data-linenos="5 "></span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">z</span><span class="p">,</span> <span class="n">w</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
<span class="linenos" data-linenos="6 "></span><span class="n">y</span><span class="o">.</span><span class="n">shape</span>
</code></pre></div>
<p>要创建一个 MuData 对象，我们首先需要单模态的 AnnData 对象。因此，我们创建两个 AnnData 对象，<em>它们具有相同观测值的数据</em>，但 _ 变量不同 _。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;obs_</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
<span class="linenos" data-linenos="3 "></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;var_</span><span class="si">{</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">d</span><span class="p">)]</span>
<span class="linenos" data-linenos="4 "></span><span class="n">adata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 1000 × 100
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">d2</span> <span class="o">=</span> <span class="mi">50</span>
<span class="linenos" data-linenos="2 "></span><span class="n">w2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">d2</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
<span class="linenos" data-linenos="3 "></span><span class="n">y2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">z</span><span class="p">,</span> <span class="n">w2</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span>
<span class="linenos" data-linenos="5 "></span><span class="n">adata2</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">y2</span><span class="p">)</span>
<span class="linenos" data-linenos="6 "></span><span class="n">adata2</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;obs_</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
<span class="linenos" data-linenos="7 "></span><span class="n">adata2</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;var2_</span><span class="si">{</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">d2</span><span class="p">)]</span>
<span class="linenos" data-linenos="8 "></span><span class="n">adata2</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 1000 × 50
</code></pre></div>
<p>这两个 AnnData 对象（两个“模态”）然后可以被包装成一个 MuData 对象。在这里，我们将模态一命名为 <code>A</code>，模态二命名为 <code>B</code>。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">md</span><span class="o">.</span><span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="n">adata2</span><span class="p">})</span>
<span class="linenos" data-linenos="2 "></span><span class="n">mdata</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>MuData object with n_obs × n_vars = 1000 × 150
<span class="linenos" data-linenos="2 "></span>  2 modalities
<span class="linenos" data-linenos="3 "></span>    A:  1000 x 100
<span class="linenos" data-linenos="4 "></span>    B:  1000 x 50
</code></pre></div>
<p>观测值和变量在 MuData 对象中是全局的，这意味着不同模态中具有相同名称（<code>.obs_names</code>）的观测值被视为相同的观测值。这也意味着变量名称（<code>.var_names</code>）应该是唯一的。这在上面的对象描述中得到了反映：<code>mdata</code> 有 1000 个观测值和 150 = 100 + 50 个变量。</p>
<h3 id="1443-mudata">1.4.4.3. MuData 属性<a class="headerlink" href="#1443-mudata" title="Permanent link">&para;</a></h3>
<p>MuData 对象包含如 AnnData 对象的注释（如 <code>.obs</code> 或 <code>.var</code>），但通过 <code>.mod</code> 扩展了这种行为，<code>.mod</code> 用于访问单个模态。</p>
<p>模态存储在一个集合中，可通过 MuData 对象的 <code>.mod</code> 属性访问，模态名称为键，AnnData 对象为值。</p>
<p>可以通过 <code>.mod</code> 属性或通过 MuData 对象本身作为简写来访问单个模态。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
<span class="linenos" data-linenos="2 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData object with n_obs × n_vars = 1000 × 100
<span class="linenos" data-linenos="2 "></span>AnnData object with n_obs × n_vars = 1000 × 100
</code></pre></div>
<p>样本（细胞）注释可以通过 <code>.obs</code> 属性访问，默认情况下包括来自各个模态的 <code>.obs</code> 数据框的列的副本。同样，<code>.var</code> 包含变量（特征）的注释。从各个模态复制的观测值列包含模态名称作为前缀，例如 <code>rna:n_genes</code>。变量列也是如此。然而，如果多个模态的 <code>.var</code> 中有相同名称的列，例如 <code>n_cells</code>，这些列将在模态之间合并，不添加前缀。当这些插槽在模态的 AnnData 对象中发生变化时，例如添加新列或过滤出样本（细胞），必须使用 <code>.update()</code> 方法获取更改（见下文）。</p>
<p>样本（细胞）的多维注释可以在 <code>.obsm</code> 属性中访问。例如，这可以是基于所有模态共同学习的 UMAP 坐标。</p>
<p>MuData 对象的形状由两个数字表示，分别计算为各个模态形状的总和——一个用于观测值的数量，一个用于变量的数量。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">)</span>
</code></pre></div>
<p>默认情况下，变量始终被计为属于单个模态，而具有相同名称的观测值被计为相同的观测值，这些观测值在多个模态中测量了变量。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span> <span class="k">for</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>[(1000, 100), (1000, 50)]
</code></pre></div>
<p>如果模态的形状发生变化，需要运行 <code>MuData.update()</code> 来将相应的更新带入 MuData 对象。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">adata2</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;var_ad2_&quot;</span> <span class="o">+</span> <span class="n">e</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">adata2</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Outdated variables names: ...,&quot;</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]))</span>
<span class="linenos" data-linenos="2 "></span><span class="n">mdata</span><span class="o">.</span><span class="n">update</span><span class="p">()</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Updated variables names: ...,&quot;</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]))</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Outdated variables names: ..., var2_48, var2_49, var2_50
<span class="linenos" data-linenos="2 "></span>Updated variables names: ..., var_ad2_48, var_ad2_49, var_ad2_50
</code></pre></div>
<p>重要的是，单个模态作为对原始对象的引用存储。因此，如果原始 AnnData 发生变化，MuData 对象中也会反映出这些变化。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># Add some unstructured data to the original object</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;misc&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;adata&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># Access modality A via the .mod attribute</span>
<span class="linenos" data-linenos="2 "></span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;misc&quot;</span><span class="p">]</span>
</code></pre></div>
<h3 id="1444">1.4.4.4. 变量映射<a class="headerlink" href="#1444" title="Permanent link">&para;</a></h3>
<p>在构建 MuData 对象时，会创建观测值与单个模态之间的全局二进制映射，以及变量与模态之间的映射。由于在 <code>mdata</code> 中各个模态之间的所有观测值都是相同的，因此观测值映射中的所有值都设置为 <code>True</code>。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="n">n</span>
</code></pre></div>
<p>然而，对于变量而言，这些是长度为 150 的向量。模态 <code>A</code> 有 100 个 <code>True</code> 值，接着是 50 个 <code>False</code> 值。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span>array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 2 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 3 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 4 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 5 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 6 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 7 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 8 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos=" 9 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos="10 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos="11 "></span>        True,  True,  True,  True,  True,  True,  True,  True,  True,
<span class="linenos" data-linenos="12 "></span>        True, False, False, False, False, False, False, False, False,
<span class="linenos" data-linenos="13 "></span>       False, False, False, False, False, False, False, False, False,
<span class="linenos" data-linenos="14 "></span>       False, False, False, False, False, False, False, False, False,
<span class="linenos" data-linenos="15 "></span>       False, False, False, False, False, False, False, False, False,
<span class="linenos" data-linenos="16 "></span>       False, False, False, False, False, False, False, False, False,
<span class="linenos" data-linenos="17 "></span>       False, False, False, False, False, False])
</code></pre></div>
<h3 id="1445-mudata">1.4.4.5. MuData 视图<a class="headerlink" href="#1445-mudata" title="Permanent link">&para;</a></h3>
<p>类似于 AnnData 对象的行为，对 MuData 对象进行切片会返回原始数据的视图。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">view</span> <span class="o">=</span> <span class="n">mdata</span><span class="p">[:</span><span class="mi">100</span><span class="p">,</span> <span class="p">:</span><span class="mi">1000</span><span class="p">]</span>
<span class="linenos" data-linenos="2 "></span><span class="nb">print</span><span class="p">(</span><span class="n">view</span><span class="o">.</span><span class="n">is_view</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">view</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">is_view</span><span class="p">)</span>
</code></pre></div>
<p>对 MuData 对象进行子集化操作具有特殊性，因为它会跨模态进行切片。也就是说，对于一组 <code>obs_names</code> 和/或 <code>var_names</code> 的切片操作将对每个模态执行，而不仅仅是全局多模态注释。这种行为使得工作流在内存使用上更加高效，这在处理大数据集时尤为重要。然而，如果需要修改对象，应创建它的副本，这样它将不再是视图，也不再依赖于原始对象。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mdata_sub</span> <span class="o">=</span> <span class="n">view</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="2 "></span><span class="n">mdata_sub</span><span class="o">.</span><span class="n">is_view</span>
</code></pre></div>
<h3 id="1446">1.4.4.6. 共同的观测值<a class="headerlink" href="#1446" title="Permanent link">&para;</a></h3>
<p>虽然一个 MuData 对象包含了两个模态的相同观测值，但在现实世界中并非总是如此，有些数据可能会缺失。设计上，MuData 考虑了这些情况，因为没有保证观测值是相同的，甚至在一个 MuData 实例中是相交的。值得注意的是，其他工具可能会提供方便的函数来处理某些常见的缺失数据场景，例如 muon 中实现的 <code>intersect_obs()</code>。</p>
<h3 id="1447-mudata">1.4.4.7. MuData 对象的读取和写入<a class="headerlink" href="#1447-mudata" title="Permanent link">&para;</a></h3>
<p>与 AnnData 对象类似，MuData 对象被设计为可以序列化为基于 HDF5 的 <code>.h5mu</code> 文件。所有模态都存储在 <code>.h5mu</code> 文件的 <code>/mod</code> HDF5 组下的各自名称下。每个单独的模态，例如 <code>/mod/A</code>，以与存储在 <code>.h5ad</code> 文件中的方式相同的方式存储。MuData 对象可以按如下方式读取和写入：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mdata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;my_mudata.h5mu&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">mdata_r</span> <span class="o">=</span> <span class="n">md</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;my_mudata.h5mu&quot;</span><span class="p">,</span> <span class="n">backed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="n">mdata_r</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>MuData object with n_obs × n_vars = 1000 × 150 backed at &#39;my_mudata.h5mu&#39;
<span class="linenos" data-linenos="2 "></span>  2 modalities
<span class="linenos" data-linenos="3 "></span>    A:  1000 x 100
<span class="linenos" data-linenos="4 "></span>      uns:  &#39;misc&#39;
<span class="linenos" data-linenos="5 "></span>    B:  1000 x 50
</code></pre></div>
<p>各个模态也在 <code>.h5mu</code> 文件中进行备份。</p>
<p>如果原始对象已被备份，则必须在 <code>.copy()</code> 调用中提供文件名，结果对象将在新位置进行备份。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mdata_sub</span> <span class="o">=</span> <span class="n">mdata_r</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="s2">&quot;mdata_sub.h5mu&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata_sub</span><span class="o">.</span><span class="n">is_view</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata_sub</span><span class="o">.</span><span class="n">isbacked</span><span class="p">)</span>
</code></pre></div>
<h3 id="1448">1.4.4.8. 多模态方法<a class="headerlink" href="#1448" title="Permanent link">&para;</a></h3>
<p>当 MuData 对象准备好后，就可以使用多模态方法来理解数据。最简单和最直接的方法是连接来自多个模态的矩阵，以执行例如降维。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">])</span>
<span class="linenos" data-linenos="2 "></span><span class="n">x</span><span class="o">.</span><span class="n">shape</span>
</code></pre></div>
<p>我们可以编写一个简单的函数，在这样的连接矩阵上运行主成分分析（PCA）。MuData 对象提供了一个存储多模态嵌入的地方：MuData 的 <code>.obsm</code>。这类似于在单个模态上生成的嵌入的存储方式，只是这次它是存储在 MuData 对象内部，而不是在 AnnData 的 <code>.obsm</code> 中。</p>
<p>例如，为了计算模态联合值的主成分分析（PCA），我们水平堆叠存储在各个模态中的值，然后对堆叠的矩阵执行 PCA。这是可能的，因为各个模态的观测值数量是匹配的（请记住，每个模态的特征数量不必匹配）。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="k">def</span> <span class="nf">simple_pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">):</span>
<span class="linenos" data-linenos=" 2 "></span>    <span class="kn">from</span> <span class="nn">sklearn</span> <span class="kn">import</span> <span class="n">decomposition</span>
<span class="linenos" data-linenos=" 3 "></span>
<span class="linenos" data-linenos=" 4 "></span>    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">m</span><span class="o">.</span><span class="n">X</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
<span class="linenos" data-linenos=" 5 "></span>
<span class="linenos" data-linenos=" 6 "></span>    <span class="n">pca</span> <span class="o">=</span> <span class="n">decomposition</span><span class="o">.</span><span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span>    <span class="n">components</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="linenos" data-linenos=" 8 "></span>
<span class="linenos" data-linenos=" 9 "></span>    <span class="c1"># By default, methods operate in-place and embeddings are stored in the .obsm slot</span>
<span class="linenos" data-linenos="10 "></span>    <span class="n">mdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_pca&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">components</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">simple_pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>MuData object with n_obs × n_vars = 1000 × 150
<span class="linenos" data-linenos="2 "></span>  obsm: &#39;X_pca&#39;
<span class="linenos" data-linenos="3 "></span>  2 modalities
<span class="linenos" data-linenos="4 "></span>    A:  1000 x 100
<span class="linenos" data-linenos="5 "></span>      uns:  &#39;misc&#39;
<span class="linenos" data-linenos="6 "></span>    B:  1000 x 50
</code></pre></div>
<p>我们计算的主成分现在存储在 MuData 对象本身中，可以进行进一步的多模态转换。</p>
<p>然而，实际上，具有不同模态通常意味着它们之间的特征来自不同的生成过程，且不可比。这时，特殊的多模态集成方法就派上用场了。对于组学技术，这些方法通常称为多组学集成方法。在以下部分中，我们将介绍 muon，它不仅提供了 RNA-Seq 以外的单模态数据预处理工具，还提供了多组学集成方法。</p>
<h2 id="145-muon">1.4.5. 进阶：使用 muon 进行多模态数据分析<a class="headerlink" href="#145-muon" title="Permanent link">&para;</a></h2>
<p>尽管 scanpy 提供了通用的工具，例如 PCA、UMAP 和各种可视化，但它主要设计用于 RNA-Seq 数据分析。muon 填补了这一空白，提供了其他组学数据（如染色质可及性（ATAC）或蛋白质（CITE）数据）的预处理函数。如上所述，muon 还提供了运行多组学算法的工具，这些算法从联合模态中推断知识。例如，用户可以在单个模态上运行 PCA，但 muon 还提供了多组学因子分析算法，这些算法以多种模态作为输入。</p>
<p><img alt="" src="https://www.sc-best-practices.org/_images/muon.png" /></p>
<p>Fig. 4.5 muon overview. Image obtained from [<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#id125" title="Danila Bredikhin, Ilia Kats, and Oliver Stegle. Muon: multimodal omics analysis framework. Genome Biology, 23(1):42, Feb 2022. URL: https://doi.org/10.1186/s13059-021-02577-8, doi:10.1186/s13059-021-02577-8.">Bredikhin <em>et al.</em>, 2022</a>].<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#muon-fig" title="Link to this image">#</a></p>
<h3 id="1451">1.4.5.1. 安装<a class="headerlink" href="#1451" title="Permanent link">&para;</a></h3>
<p>muon 可以通过 PyPI 安装：</p>
<h3 id="1452-api">1.4.5.2. API 概览<a class="headerlink" href="#1452-api" title="Permanent link">&para;</a></h3>
<p>为了介绍 muon，我们将检查一个多模态数据集的 ATAC 数据。与 scanpy 章节类似，本章仅作为 muon 的快速演示和概览，而不是完整地分析数据集，更不用说提供最佳实践的多组学分析。请阅读相应的章节，了解如何正确地进行此类分析。</p>
<p>muon 以两种方式分隔其模块。首先，类似于 scanpy，通用和多模态函数被分组到预处理（<code>muon.pp</code>）、工具（<code>muon.tl</code>）和绘图（<code>muon.pl</code>）中。其次，单模态工具可以从 muon 的相应部分获得，这些部分再次分为预处理、工具和绘图。例如，所有 ATAC 预处理功能都被分组到 <code>muon.atac.pp</code> 中。这也适用于 CITE-Seq 预处理功能（<code>muon.prot.pp</code>）。</p>
<p><img alt="" src="https://www.sc-best-practices.org/_images/muon_api.png" /></p>
<p>Fig. 4.6 muon API overview. Modality specific functions are provided in correspondingly named modules. Functions for the joint analysis of modalities are available via the muon module directly.<a href="https://www.sc-best-practices.org/introduction/analysis_tools.html#muon-api" title="Link to this image">#</a></p>
<h3 id="1453-muon-api">1.4.5.3. muon API 演示<a class="headerlink" href="#1453-muon-api" title="Permanent link">&para;</a></h3>
<p>我们演示用的数据集是一个公开可用的 10x Genomics Multiome 数据集，针对人类 PBMCs。让我们导入 muon 并使用 mudatasets 获取该数据集。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">import</span> <span class="nn">mudatasets</span> <span class="k">as</span> <span class="nn">mds</span>
<span class="linenos" data-linenos="2 "></span><span class="kn">import</span> <span class="nn">muon</span> <span class="k">as</span> <span class="nn">mu</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mdata</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">mds</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;pbmc10k_multiome&quot;</span><span class="p">,</span> <span class="n">with_info</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">full</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">mdata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span>■ File filtered_feature_bc_matrix.h5 from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/filtered_feature_bc_matrix.h5
<span class="linenos" data-linenos=" 2 "></span>■ Checksum is validated (md5) for filtered_feature_bc_matrix.h5
<span class="linenos" data-linenos=" 3 "></span>■ File atac_fragments.tsv.gz from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz
<span class="linenos" data-linenos=" 4 "></span>■ Checksum is validated (md5) for atac_fragments.tsv.gz
<span class="linenos" data-linenos=" 5 "></span>■ File atac_fragments.tsv.gz.tbi from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz.tbi
<span class="linenos" data-linenos=" 6 "></span>■ Checksum is validated (md5) for atac_fragments.tsv.gz.tbi
<span class="linenos" data-linenos=" 7 "></span>■ File atac_peaks.bed from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_peaks.bed
<span class="linenos" data-linenos=" 8 "></span>■ Checksum is validated (md5) for atac_peaks.bed
<span class="linenos" data-linenos=" 9 "></span>■ File atac_peak_annotation.tsv from pbmc10k_multiome has been found at /home/zeth/mudatasets/pbmc10k_multiome/atac_peak_annotation.tsv
<span class="linenos" data-linenos="10 "></span>■ Checksum is validated (md5) for atac_peak_annotation.tsv
<span class="linenos" data-linenos="11 "></span>■ Loading filtered_feature_bc_matrix.h5...
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>/home/zeth/miniconda3/envs/bp2/lib/python3.9/site-packages/mudatasets/core.py:203: UserWarning: Dataset is in the 10X .h5 format and can&#39;t be loaded as backed.
<span class="linenos" data-linenos="2 "></span>  warn(&quot;Dataset is in the 10X .h5 format and can&#39;t be loaded as backed.&quot;)
<span class="linenos" data-linenos="3 "></span>Variable names are not unique. To make them unique, call `.var_names_make_unique`.
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Added `interval` annotation for features from /home/zeth/mudatasets/pbmc10k_multiome/filtered_feature_bc_matrix.h5
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Variable names are not unique. To make them unique, call `.var_names_make_unique`.
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Added peak annotation from /home/zeth/mudatasets/pbmc10k_multiome/atac_peak_annotation.tsv to .uns[&#39;atac&#39;][&#39;peak_annotation&#39;]
<span class="linenos" data-linenos="2 "></span>Added gene names to peak annotation in .uns[&#39;atac&#39;][&#39;peak_annotation&#39;]
<span class="linenos" data-linenos="3 "></span>Located fragments file: /home/zeth/mudatasets/pbmc10k_multiome/atac_fragments.tsv.gz
</code></pre></div>
<p>作为第一步，我们将数据集子集化为 ATAC 模态。</p>
<p>虽然我们现在不处理 RNA-Seq 数据，但可以使用一些 scanpy 的预处理功能，这些功能也可以用于 ATAC 数据。这是因为两种模态具有相似的分布和质量问题。在这里需要记住的是，在具有 ATAC-seq 数据的 AnnData 对象中，基因意味着一个峰。之后，可以使用 muon 的 ATAC 模块进行特定于 ATAC 的预处理。</p>
<p>让我们从一些质量控制开始，通过过滤掉具有太少峰的细胞和在太少细胞中检测到的峰来进行质量控制。现在，我们将过滤掉未通过质量控制的细胞。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>/home/zeth/miniconda3/envs/bp2/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.
<span class="linenos" data-linenos="2 "></span>  c.reorder_categories(natsorted(c.categories), inplace=True)
<span class="linenos" data-linenos="3 "></span>... storing &#39;feature_types&#39; as categorical
<span class="linenos" data-linenos="4 "></span>/home/zeth/miniconda3/envs/bp2/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.
<span class="linenos" data-linenos="5 "></span>  c.reorder_categories(natsorted(c.categories), inplace=True)
<span class="linenos" data-linenos="6 "></span>... storing &#39;genome&#39; as categorical
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/af874602e780c9e686a842b4d95c729ef75d817a01e4e613f1973ddb2d92370b.png" /></p>
<p>过滤掉未检测到表达的峰。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_var</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;n_cells_by_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">10</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="c1"># This is analogous to</span>
<span class="linenos" data-linenos="3 "></span><span class="c1">#   sc.pp.filter_genes(rna, min_cells=10)</span>
<span class="linenos" data-linenos="4 "></span><span class="c1"># but does in-place filtering and avoids copying the object</span>
</code></pre></div>
<p>我们也会过滤掉不合格的细胞。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">2000</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">15000</span><span class="p">))</span>
<span class="linenos" data-linenos="2 "></span><span class="c1"># This is analogous to</span>
<span class="linenos" data-linenos="3 "></span><span class="c1">#   sc.pp.filter_cells(atac, max_genes=15000)</span>
<span class="linenos" data-linenos="4 "></span><span class="c1">#   sc.pp.filter_cells(atac, min_genes=2000)</span>
<span class="linenos" data-linenos="5 "></span><span class="c1"># but does in-place filtering avoiding copying the object</span>
<span class="linenos" data-linenos="6 "></span>
<span class="linenos" data-linenos="7 "></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">4000</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="mi">40000</span><span class="p">))</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/03e1c22ce484a20491d297966cb3f717ce5bf756df418bcca5453152410a458c.png" /></p>
<p>muon 还提供了直方图，可以从不同角度查看指标。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">])</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/750ad0deb64d8e0242a27beccf8b02431853a720d5cc0d67b1c8f1a206febc30.png" /></p>
<p>现在我们已经初步过滤掉了具有太少峰的细胞和在太少细胞中检测到的峰，我们可以开始使用 muon 进行 ATAC 特定的质量控制。muon 在相应的模块中具有特定模态的预处理功能。我们导入 ATAC 模块以访问 ATAC 特定的预处理功能。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">from</span> <span class="nn">muon</span> <span class="kn">import</span> <span class="n">atac</span> <span class="k">as</span> <span class="n">ac</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># Perform rudimentary quality control with muon&#39;s ATAC module</span>
<span class="linenos" data-linenos="2 "></span><span class="n">atac</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;NS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
<span class="linenos" data-linenos="3 "></span><span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">nucleosome_signal</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span><span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">get_gene_annotation_from_rna</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="linenos" data-linenos="5 "></span><span class="n">tss</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tss_enrichment</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">n_tss</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="linenos" data-linenos="6 "></span><span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tss_enrichment</span><span class="p">(</span><span class="n">tss</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Reading Fragments: 100%|██████████| 1000000/1000000 [00:05&lt;00:00, 185917.24it/s]
<span class="linenos" data-linenos="2 "></span>Fetching Regions...: 100%|██████████| 1000/1000 [00:19&lt;00:00, 50.03it/s]
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/55eb6874ce8bf52d43523d5aa4299bed653f9aceb75c2bdf65a51812ecfac499.png" /></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># Save original counts, normalize data and select highly variable genes with scanpy</span>
<span class="linenos" data-linenos="2 "></span><span class="n">atac</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atac</span><span class="o">.</span><span class="n">X</span>
<span class="linenos" data-linenos="3 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">counts_per_cell_after</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span>
<span class="linenos" data-linenos="5 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="linenos" data-linenos="6 "></span><span class="n">atac</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">atac</span>
</code></pre></div>
<p>虽然 PCA 也常用于 ATAC 数据，但潜在语义索引（LSI）是另一种流行的选择。它在 muon 的 ATAC 模块中实现。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_lsi&quot;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/c805123a0db86b8944bcbab0b09263875ddaa4dfd88127632b1966b06a6d15c4.png" /></p>
<p>我们可以使用 muon 的 ATAC 模块的功能，根据对应于特定基因的峰值的切值来为图表着色。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">ac</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;KLF4&quot;</span><span class="p">],</span> <span class="n">average</span><span class="o">=</span><span class="s2">&quot;peak_type&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="https://www.sc-best-practices.org/_images/0d8d7f7863bfa9193357387e7e542d3a8c1a59323687d3270e30e58a84a78cef.png" /></p>
<p>有关 muon 所有可用功能的更多详细信息，请阅读 muon API 参考文档：<a href="https://muon.readthedocs.io/en/latest/api/index.html">https://muon.readthedocs.io/en/latest/api/index.html</a> 和 muon 教程：<a href="https://muon-tutorials.readthedocs.io/en/latest/">https://muon-tutorials.readthedocs.io/en/latest/</a>。</p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../1.2%20Raw%20Data%20Processing/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 1.2 Raw data processing">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                1.2 Raw data processing
              </div>
            </div>
          </a>
        
        
          
          <a href="../2.1%20quality_control/" class="md-footer__link md-footer__link--next" aria-label="Next: 2.1 Quality Control">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                2.1 Quality Control
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.code.select", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.instant.prefetch", "navigation.instant.progress", "navigation.path", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "header.autohide"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.525ec568.min.js"></script>
      
        <script src="../mathjax-config.js"></script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../extra.js"></script>
      
        <script src="../js/toggle_sidebar.js"></script>
      
    
  </body>
</html>