
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://localhost:8000/Autograd%20mechanics%20%E2%80%94%20PyTorch%202_3%20documentation/">
      
      
        <link rel="prev" href="../Attention%20Generation%20Work%20Flow/">
      
      
        <link rel="next" href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/">
      
      
      <link rel="icon" href="../img/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>Autograd mechanics — PyTorch 2 3 documentation - One Last Kiss</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lxgw-wenkai-webfont@1.1.0/style.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-M0R7V8QGQ7"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-M0R7V8QGQ7",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-M0R7V8QGQ7",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#autograd" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="One Last Kiss" class="md-header__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            One Last Kiss
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Autograd mechanics — PyTorch 2 3 documentation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
    
  
  🏠 Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Mygo/" class="md-tabs__link">
          
  
    
  
  ⛏️ CV and Multimodels

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../Session%201/" class="md-tabs__link">
          
  
    
  
  🚀 Re:0 Road to CPath & Visualization

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-tabs__link">
          
  
    
  
  💤 Re:0 Road to scRNA-seq Analysis

        </a>
      </li>
    
  

    
  

    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-tabs__link">
          
  
    
  
  🧬 scRNA Analysis

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-tabs__link">
          
  
    
  
  📐 Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-tabs__link">
          
  
    
  
  💡 Reviews

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="One Last Kiss" class="md-nav__button md-logo" aria-label="One Last Kiss" data-md-component="logo">
      
  <img src="../img/icon.png" alt="logo">

    </a>
    One Last Kiss
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/aletolia/aletolia.github.io.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    aletolia/aletolia.github.io
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href=".." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    🏠 Home
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            🏠 Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⛏️ CV and Multimodels
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ⛏️ CV and Multimodels
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mygo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MyGO: Discrete Modality Information as Fine-Grained Tokens for Multi-modal Knowledge Graph Completion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../GMM/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GMM
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../DINO/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    DINO
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mamba/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mamba
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ViT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ViT
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCo/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Self-Supervised%20Learning%20MoCoV2%2C3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MoCo V2
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%80%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (一)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%20%28%E4%BA%8C%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (二)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E4%B8%89%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer  (三)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E5%9B%9B%29%20%28TNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (四)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28%E7%9F%A5%E8%AF%86%E8%92%B8%E9%A6%8F%20%E9%A2%84%E8%AE%AD%E7%BB%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer (知识蒸馏 预训练）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Transformer%20%28PyramidTNT%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Transformer(Pyramid TNT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%82%85%E9%87%8C%E5%8F%B6%E7%BA%A7%E6%95%B0%E5%92%8C%E5%82%85%E9%87%8C%E5%8F%B6%E5%8F%98%E6%8D%A2%E4%B8%93%E9%A2%98/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    傅里叶级数和傅里叶变换专题
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Python%20%E5%AD%98%E5%82%A8%E4%B8%8E%E8%AF%BB%E5%8F%96%20HDF5%20%E6%96%87%E4%BB%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Python 存储与读取 HDF5 文件
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Segment%20Anything%20in%20Medical%20Images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Segment Anything in Medical Images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Vision%20Mamba%20Efficient%20Visual%20Representation%20Learning%20with%20Bidirectional%20State%20Space%20Model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vision Mamba Efficient Visual Representation Learning with Bidirectional State Space Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_18" >
        
          
          <label class="md-nav__link" for="__nav_2_18" id="__nav_2_18_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🚑 医学影像相关
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_18_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_18">
            <span class="md-nav__icon md-icon"></span>
            🚑 医学影像相关
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Why%20is%20the%20winner%20the%20best/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Why is the winner the best
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PORPOISE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PORPOISE
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../HIPT/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Scaling Vision Transformers to Gigapixel Images via Hierarchical Self-Supervised Learning(HIPT)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Data-efficient%20and%20weakly%20supervised%20computational%20pathology%20on%20whole-slide%20images%28CLAM%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data efficient and weakly supervised computational pathology on whole slide images(CLAM)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Towards%20a%20general-purpose%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Towards a general purpose foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20whole-slide%20foundation%20model%20for%20digital%20pathology%20from%20real-world%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A whole slide foundation model for digital pathology from real world data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20visual-language%20foundation%20model%20for%20computational%20pathology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A visual language foundation model for computational pathology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Recurrent%20memory%20with%20optimal%20polynomial%20projections./" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Recurrent memory with optimal polynomial projections.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Modeling%20Dense%20Multimodal%20Interactions%20Between%20Biological%20Pathways%20and%20Histology%20for%20Survival%20Prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modeling Dense Multimodal Interactions Between Biological Pathways and Histology for Survival Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Inferring%20super-resolution%20tissue%20architecture%20by%20integrating%20spatial%20transcriptomics%20with%20histology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inferring super resolution tissue architecture by integrating spatial transcriptomics with histology
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Prediction%20of%20recurrence%20risk%20in%20endometrial%20cancer%20with%20multimodal%20deep%20learning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction of recurrence risk in endometrial cancer with multimodal deep learning
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    🚀 Re:0 Road to CPath & Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            🚀 Re:0 Road to CPath & Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%201/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    理解全幅切片图像
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%202/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    在计算病理学中用于癌症诊断的弱监督深度学习
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Session%203/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    基于注意力的多示例学习可解释性
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../How%20to%20visualize%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to visualize Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../AttentionViz%20A%20Global%20View%20of%20Transformer%20Attention/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AttentionViz A Global View of Transformer Attention
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Transformer%20Interpretability%20Beyond%20Attention%20Visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transformer Interpretability Beyond Attention Visualization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Visualization%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualization Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Attention%20Generation%20Work%20Flow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Attention Generation Work Flow
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_9" checked>
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="">
            
  
  <span class="md-ellipsis">
    🪛 PyTorch
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            🪛 PyTorch
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Autograd mechanics — PyTorch 2 3 documentation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Autograd mechanics — PyTorch 2 3 documentation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#autograd" class="md-nav__link">
    <span class="md-ellipsis">
      Autograd 如何记录历史
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      已保存的张量
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      非可微函数的梯度
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      局部禁用梯度计算
    </span>
  </a>
  
    <nav class="md-nav" aria-label="局部禁用梯度计算">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#requires_grad" class="md-nav__link">
    <span class="md-ellipsis">
      设置 requires_grad
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grad-modes" class="md-nav__link">
    <span class="md-ellipsis">
      Grad Modes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      默认模式（梯度模式）
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      无梯度模式
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      推理模式
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#nnmoduleeval" class="md-nav__link">
    <span class="md-ellipsis">
      评估模式 (nn.Module.eval())
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#autograd_1" class="md-nav__link">
    <span class="md-ellipsis">
      使用 Autograd 的就地操作
    </span>
  </a>
  
    <nav class="md-nav" aria-label="使用 Autograd 的就地操作">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_7" class="md-nav__link">
    <span class="md-ellipsis">
      就地正确性检查
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#autograd_2" class="md-nav__link">
    <span class="md-ellipsis">
      多线程 Autograd
    </span>
  </a>
  
    <nav class="md-nav" aria-label="多线程 Autograd">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cpu" class="md-nav__link">
    <span class="md-ellipsis">
      CPU 并发
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_8" class="md-nav__link">
    <span class="md-ellipsis">
      不确定性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_9" class="md-nav__link">
    <span class="md-ellipsis">
      保留计算图
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#autograd_3" class="md-nav__link">
    <span class="md-ellipsis">
      Autograd 节点上的线程安全
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#c" class="md-nav__link">
    <span class="md-ellipsis">
      C++ 钩子上的线程不安全
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#autograd_4" class="md-nav__link">
    <span class="md-ellipsis">
      复数的 Autograd
    </span>
  </a>
  
    <nav class="md-nav" aria-label="复数的 Autograd">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_10" class="md-nav__link">
    <span class="md-ellipsis">
      什么是复数导数？
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_11" class="md-nav__link">
    <span class="md-ellipsis">
      维尔廷格微积分登场
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_12" class="md-nav__link">
    <span class="md-ellipsis">
      维尔廷格微积分在优化中的作用
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pytorch" class="md-nav__link">
    <span class="md-ellipsis">
      PyTorch 如何计算共轭维尔廷格导数？
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_13" class="md-nav__link">
    <span class="md-ellipsis">
      如何为复数函数编写导数公式？
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_14" class="md-nav__link">
    <span class="md-ellipsis">
      跨域函数呢？
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_15" class="md-nav__link">
    <span class="md-ellipsis">
      已保存张量的钩子
    </span>
  </a>
  
    <nav class="md-nav" aria-label="已保存张量的钩子">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_16" class="md-nav__link">
    <span class="md-ellipsis">
      反向传播钩子的执行
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_17" class="md-nav__link">
    <span class="md-ellipsis">
      钩子是否会触发
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_18" class="md-nav__link">
    <span class="md-ellipsis">
      钩子触发顺序
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_19" class="md-nav__link">
    <span class="md-ellipsis">
      特殊钩子
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_20" class="md-nav__link">
    <span class="md-ellipsis">
      修改张量时的钩子行为
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tensorregister_hookhook" class="md-nav__link">
    <span class="md-ellipsis">
      Tensor.register_hook(hook)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyTorch Autograd Explained   In depth Tutorial
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💤 Re:0 Road to scRNA-seq Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            💤 Re:0 Road to scRNA-seq Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏳ CHAPTER 1 简介
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ⏳ CHAPTER 1 简介
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20Single-cell%20RNA%20sequencing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1.The building block of life
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20Raw%20Data%20Processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 Raw data processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Analysis%20frameworks%20and%20tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Analysis frameworks and tools
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🌏 CHAPTER 2 数据准备与可视化
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            🌏 CHAPTER 2 数据准备与可视化
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20normalization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 Normalization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3_feature_selection/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 Feature selection
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.4_dimensionality_reduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.4 Dimensionality Reduction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_3" >
        
          
          <label class="md-nav__link" for="__nav_4_1_3" id="__nav_4_1_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🍀 CHAPTER 3 识别细胞结构
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_3">
            <span class="md-nav__icon md-icon"></span>
            🍀 CHAPTER 3 识别细胞结构
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 Data integration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_4" >
        
          
          <label class="md-nav__link" for="__nav_4_1_4" id="__nav_4_1_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⏱️ CHAPTER 4 推断细胞轨迹
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_4">
            <span class="md-nav__icon md-icon"></span>
            ⏱️ CHAPTER 4 推断细胞轨迹
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotemporal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 Pseudotemporal ordering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../rna_velocity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 RNA velocity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../lineage_tracing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 Lineage tracing
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_5" >
        
          
          <label class="md-nav__link" for="__nav_4_1_5" id="__nav_4_1_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🎁 CHAPTER 5 应对特殊情况
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_5">
            <span class="md-nav__icon md-icon"></span>
            🎁 CHAPTER 5 应对特殊情况
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../differential_gene_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 Differential gene expression analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compositional/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 Compositional analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gsea_pathway/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 Gene set enrichment and pathway analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../perturbation_modeling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 Perturbation modeling
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_6" >
        
          
          <label class="md-nav__link" for="__nav_4_1_6" id="__nav_4_1_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ⚙️ CHAPTER 6 机制建模
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_6">
            <span class="md-nav__icon md-icon"></span>
            ⚙️ CHAPTER 6 机制建模
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_regulatory_networks/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 Gene regulatory networks
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cell_cell_communication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 Cell-cell communication
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1_7" >
        
          
          <label class="md-nav__link" for="__nav_4_1_7" id="__nav_4_1_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🔓 CHAPTER 7 反卷积
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_7">
            <span class="md-nav__icon md-icon"></span>
            🔓 CHAPTER 7 反卷积
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bulk_deconvolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 Bulk deconvolution
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    简明易懂的单细胞分析流程（使用 R 语言）
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            简明易懂的单细胞分析流程（使用 R 语言）
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E5%89%8D%E8%A8%80/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    前言
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    环境配置
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.1%20%E6%9E%84%E5%BB%BA%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E5%88%86%E6%9E%90%E7%8E%AF%E5%A2%83/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.1 构建单细胞 RNA seq 分析环境
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.2%20R%20%E7%8E%AF%E5%A2%83%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.2 R 环境安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.3%20Docker%20%E5%AE%89%E8%A3%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.3 Docker 安装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.4%20%E5%9C%A8%20Docker%20%E4%B8%AD%E5%90%AF%E5%8A%A8%20Rstudio/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.4 在 Docker 中启动 Rstudio
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1.5%20%E6%9C%89%E5%85%B3%20Bioconductor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1.5 有关 Bioconductor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.1%20R%20%E7%9A%84%E5%9F%BA%E6%9C%AC%E7%94%A8%E6%B3%95/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.1 R 的基本用法
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.2%20%E5%A6%82%E4%BD%95%E5%A4%84%E7%90%86%20DataFrame%EF%BC%9F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.2 如何处理 DataFrame？
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2.3%20%E4%BD%BF%E7%94%A8%20ggplot2%20%E5%8F%AF%E8%A7%86%E5%8C%96%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2.3 使用 ggplot2 可视化数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.1%20PCA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.1 PCA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.2%20%E4%BD%BF%E7%94%A8%20clusterPlofiler%20%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.2 使用 clusterPlofiler 进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E5%9F%BA%E5%9B%A0%E6%9C%AC%E4%BD%93%EF%BC%88GO%EF%BC%89%E4%B8%AD%E5%90%84%20BP%E3%80%81MF%20%E5%92%8C%20CC%20%E7%9A%84%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 基因本体（GO）中各 BP、MF 和 CC 的富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.3%20%E4%BD%BF%E7%94%A8%20DEseq2%20%E6%8F%90%E5%8F%96%20RNA-seq%20%E5%88%86%E6%9E%90%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%20DEG/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.3 使用 DEseq2 提取 RNA seq 分析数据中的 DEG
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3.4%20%E4%BD%BF%E7%94%A8%E5%85%B6%E4%BB%96%E6%95%B0%E6%8D%AE%E5%BA%93%EF%BC%88KEGG%E3%80%81Reactome%20%E7%AD%89%EF%BC%89%E8%BF%9B%E8%A1%8C%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3.4 使用其他数据库（KEGG、Reactome 等）进行富集分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.1%20scRNA-seq%20%E6%A6%82%E8%A6%81/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.1 scRNA seq 概要
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.2%20%E5%87%86%E5%A4%87%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.2 准备使用 Seurat 进行分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.3%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.3 使用 Seurat 进行数据预处理
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4.4%20%E4%BD%BF%E7%94%A8%20Seurat%20%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E7%9A%84%E4%B8%80%E8%88%AC%E9%A1%BA%E5%BA%8F/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4.4 使用 Seurat 进行数据分析的一般顺序
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.1%20%E5%A6%82%E4%BD%95%E6%9F%A5%E6%89%BE%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.1 如何查找单细胞 RNA seq 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.2%20%E4%B8%BA%E6%AF%8F%E4%B8%AA%E7%B0%87%E8%87%AA%E5%8A%A8%E6%B7%BB%E5%8A%A0%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E6%A0%87%E7%AD%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.2 为每个簇自动添加细胞注释标签
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.3%20%E4%BD%BF%E7%94%A8%20FeaturePlot%20%E6%A3%80%E6%9F%A5%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.3 使用 FeaturePlot 检查基因表达
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.4%20%E5%B0%86%E5%8D%95%E7%BB%86%E8%83%9E%20RNA-seq%20%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9B%B8%E4%BA%92%E7%BB%93%E5%90%88/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.4 将单细胞 RNA seq 数据集相互结合
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.5%20%E7%BA%A0%E6%AD%A3%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94%EF%BC%8C%E6%AF%94%E8%BE%83%E5%AF%B9%E7%85%A7%E7%BB%84%E5%92%8C%E5%88%BA%E6%BF%80%E7%BB%84/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.5 纠正批次效应，比较对照组和刺激组
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5.6%20%E7%A1%AE%E5%AE%9A%E6%AF%8F%E4%B8%AA%E7%BE%A4%E7%BB%84%E7%9A%84%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5.6 确定每个群组的标记基因
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.1%20%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E7%BB%9F%E8%AE%A1%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E4%B8%AD%E7%9A%84%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E6%B0%B4%E5%B9%B3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.1 如何使用 Cell Ranger 统计 scRNA seq 数据中的基因表达水平
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6.2%20%E4%BD%BF%E7%94%A8%20Cell%20Ranger%20%E5%A4%84%E7%90%86%20NCBI%20Sequence%20Read%20Archive%20%28SRA%29%20%E6%95%B0%E6%8D%AE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6.2 使用 Cell Ranger 处理 NCBI Sequence Read Archive (SRA) 数据
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.1%20%E5%85%B3%E4%BA%8E%20scRNA-seq%20%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.1 关于 scRNA seq 数据的质量控制
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.2%20%E4%BD%BF%E7%94%A8%20DoubleFinder%20%E6%A3%80%E6%B5%8B%E5%8F%8C%E9%87%8D%E7%BB%86%E8%83%9E/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.2 使用 DoubleFinder 检测双重细胞
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../7.3%20%E7%94%A8%20CellBender%20%E5%8E%BB%E9%99%A4%E8%83%8C%E6%99%AF%20RNA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7.3 用 CellBender 去除背景 RNA
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧬 scRNA Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            🧬 scRNA Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🐍 Using Python
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            🐍 Using Python
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Large-scale%20foundation%20model%20on%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Large scale foundation model on single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scGPT%20toward%20building%20a%20foundation%20model%20for%20single-cell%20multi-omics%20using%20generative%20AI/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scGPT toward building a foundation model for single cell multi omics using generative AI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scBERT%20as%20a%20large-scale%20pretrained%20deep%20language%20model%20for%20cell%20type%20annotation%20of%20single-cell%20RNA-seq%20data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scBERT as a large scale pretrained deep language model for cell type annotation of single cell RNA seq data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../RNA%20velocity%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    RNA velocity of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TFvelo_gene%20regulation%20inspired%20RNA%20velocity%20estimation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TFvelo gene regulation inspired RNA velocity estimation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Disentanglement%20of%20single-cell%20data%20with%20biolord/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Disentanglement of single cell data with biolord
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E2%AD%90%E2%AD%90%E2%AD%90C5aR1%20inhibition%20reprograms%20tumor%20associated%20macrophages%20and%20reverses%20PARP%20inhibitor%20resistance%20in%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ⭐⭐⭐C5aR1 inhibition reprograms tumor associated macrophages and reverses PARP inhibitor resistance in breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Spatial%20transition%20tensor%20of%20single%20cells/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Spatial transition tensor of single cells
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Statistical%20method%20scDEED%20for%20detecting%20dubious%202D%20single-cell%20embeddings%20and%20optimizing%20t-SNE%20and%20UMAP%20hyperparameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Statistical method scDEED for detecting dubious 2D single cell embeddings and optimizing t SNE and UMAP hyperparameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRank%20infers%20drug-responsive%20cell%20types%20from%20untreated%20scRNA-seq%20data%20using%20a%20target-perturbed%20gene%20regulatory%20network/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRank infers drug responsive cell types from untreated scRNA seq data using a target perturbed gene regulatory network
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📔 文献
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            📔 文献
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ZNF689%20deficiency%20promotes%20intratumor%20heterogeneity%20and%20immunotherapy%20resistance%20in%20triple-negative%20breast%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ZNF689 deficiency promotes intratumor heterogeneity and immunotherapy resistance in triple negative breast cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Tumour%20vasculature%20at%20single-cell%20resolution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tumour vasculature at single-cell resolution
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scRNA-seq%20of%20gastric%20tumor%20shows%20complex%20intercellular%20interaction%20with%20an%20alternative%20T%20cell%20exhaustion%20trajectory%EF%BC%88%E5%8F%82%E8%80%83%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scRNA seq of gastric tumor shows complex intercellular interaction with an alternative T cell exhaustion trajectory（参考）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20reveals%20new%20evolutionary%20complexity%20in%20uveal%20melanoma%EF%BC%88%E5%8F%82%E8%80%83%202%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis reveals new evolutionary complexity in uveal melanoma（参考 2）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20characterization%20of%20macrophages%20in%20uveal%20melanoma%20uncovers%20transcriptionally%20heterogeneous%20subsets%20conferring%20poor%20prognosis%20and%20aggressive%20behavior%EF%BC%88%E8%8C%83%E4%BE%8B%201%EF%BC%89/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell characterization of macrophages in uveal melanoma uncovers transcriptionally heterogeneous subsets conferring poor prognosis and aggressive behavior（范例 1）
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ScRNA-seq%20of%20gastric%20cancer%20tissues%20reveals%20differences%20in%20the%20immune%20microenvironment%20of%20primary%20tumors%20and%20metastases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ScRNA seq of gastric cancer tissues reveals differences in the immune microenvironment of primary tumors and metastases
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Blimp-1%20and%20c-Maf%20regulate%20immune%20gene%20networks%20to%20protect%20against%20distinct%20pathways%20of%20pathobiont-induced%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Blimp 1 and c Maf regulate immune gene networks to protect against distinct pathways of pathobiont induced colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single%20cell%20analysis%20unveils%20B%20cell-dominated%20immune%20subtypes%20in%20HNSCC%20for%20enhanced%20prognostic%20and%20therapeutic%20stratification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis unveils B cell dominated immune subtypes in HNSCC for enhanced prognostic and therapeutic stratification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20spatial%20multi-omics%20and%20deep%20learning%20dissect%20enhancer-driven%20gene%20regulatory%20networks%20in%20liver%20zonation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell spatial multi omics and deep learning dissect enhancer driven gene regulatory networks in liver zonation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Neutrophil%20profiling%20illuminates%20anti-tumor%20antigen%20presenting%20potency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Neutrophil profiling illuminates anti tumor antigen presenting potency
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20blueprint%20for%20tumor-infiltrating%20B%20cells%20across%20human%20cancers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A blueprint for tumor infiltrating B cells across human cancers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20transcriptomic%20analyses%20reveal%20distinct%20immune%20cell%20contributions%20to%20epithelial%20barrier%20dysfunction%20in%20checkpoint%20inhibitor%20colitis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell transcriptomic analyses reveal distinct immune cell contributions to epithelial barrier dysfunction in checkpoint inhibitor colitis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Systematic%20dissection%20of%20tumor-normal%20single-cell%20ecosystems%20across%20a%20thousand%20tumors%20of%2030%20cancer%20types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Systematic dissection of tumor normal single cell ecosystems across a thousand tumors of 30 cancer types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../The%20aged%20tumor%20microenvironment%20limits%20T%20cell%20control%20of%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The aged tumor microenvironment limits T cell control of cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../TIM-3%2B%20CD8%20T%20cells%20with%20a%20terminally%20exhausted%20phenotype%20retain%20functional%20capacity%20in%20hematological%20malignancies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TIM 3+ CD8 T cells with a terminally exhausted phenotype retain functional capacity in hematological malignancies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20lactate-SREBP2%20signaling%20axis%20drives%20tolerogenic%20dendritic%20cell%20maturation%20and%20promotes%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A lactate SREBP2 signaling axis drives tolerogenic dendritic cell maturation and promotes cancer progression
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20human%20omentum-specific%20mesothelial-like%20stromal%20population%20inhibits%20adipogenesis%20through%20IGFBP2%20secretion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A human omentum specific mesothelial like stromal population inhibits adipogenesis through IGFBP2 secretion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Interferon-stimulated%20neutrophils%20as%20a%20predictor%20of%20immunotherapy%20response/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interferon stimulated neutrophils as a predictor of immunotherapy response
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Repression%20of%20latent%20NF-%CE%BAB%20enhancers%20by%20PDX1%20regulates%20%CE%B2%20cell%20functional%20heterogeneity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Repression of latent NF κB enhancers by PDX1 regulates β cell functional heterogeneity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20systematic%20pan-cancer%20study%20on%20deep%20learning-based%20prediction%20of%20multi-omic%20biomarkers%20from%20routine%20pathology%20images/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A systematic pan cancer study on deep learning based prediction of multi omic biomarkers from routine pathology images
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20and%20spatial%20transcriptomics%20analysis%20of%20non-small%20cell%20lung%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell and spatial transcriptomics analysis of non small cell lung cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Seeing%20data%20as%20t-SNE%20and%20UMAP%20do/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Seeing data as t SNE and UMAP do
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Mapping%20the%20cellular%20biogeography%20of%20human%20bone%20marrow%20niches%20using%20single-cell%20transcriptomics%20and%20proteomic%20imaging/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mapping the cellular biogeography of human bone marrow niches using single cell transcriptomics and proteomic imaging
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Onco-fetal%20Reprogramming%20of%20Endothelial%20Cells%20Drives%20Immunosuppressive%20Macrophages%20in%20Hepatocellular%20Carcinoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Onco fetal Reprogramming of Endothelial Cells Drives Immunosuppressive Macrophages in Hepatocellular Carcinoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20analysis%20of%20anti-BCMA%20CAR%20T%20cell%20therapy%20in%20patients%20with%20central%20nervous%20system%20autoimmunity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell analysis of anti BCMA CAR T cell therapy in patients with central nervous system autoimmunity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Single-cell%20sequencing%20depicts%20tumor%20architecture%20and%20empowers%20clinical%20decision%20in%20metastatic%20conjunctival%20melanoma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell sequencing depicts tumor architecture and empowers clinical decision in metastatic conjunctival melanoma
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../PR-SET7%20epigenetically%20restrains%20uterine%20interferon%20response%20and%20cell%20death%20governing%20proper%20postnatal%20stromal%20development/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PR SET7 epigenetically restrains uterine interferon response and cell death governing proper postnatal stromal development
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Clonal%20associations%20between%20lymphocyte%20subsets%20and%20functional%20states%20in%20rheumatoid%20arthritis%20synovium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clonal associations between lymphocyte subsets and functional states in rheumatoid arthritis synovium
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    📐 Data Structures
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            📐 Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%201%20%E7%AB%A0%20python%20%E5%85%A5%E9%97%A8/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 1 章 python 入门
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%202%20%E7%AB%A0%20%E9%9D%A2%E5%90%91%E5%AF%B9%E8%B1%A1%E7%BC%96%E7%A8%8B/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 2 章 面向对象编程
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%203%20%E7%AB%A0%20%E7%AE%97%E6%B3%95%E5%88%86%E6%9E%90/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 3 章 算法分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%204%20%E7%AB%A0%20%E9%80%92%E5%BD%92/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 4 章 递归
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%205%20%E7%AB%A0%20%E5%9F%BA%E4%BA%8E%E6%95%B0%E7%BB%84%E7%9A%84%E5%BA%8F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 5 章 基于数组的序列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%206%20%E7%AB%A0%20%E6%A0%88%E3%80%81%E9%98%9F%E5%88%97%E5%92%8C%E5%8F%8C%E7%AB%AF%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 6 章 栈、队列和双端队列
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%207%20%E7%AB%A0%20%E9%93%BE%E8%A1%A8%28Linked%20Lists%29/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 7 章 链表(Linked Lists)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%208%20%E7%AB%A0%20%E6%A0%91%20%E2%98%85/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 8 章 树 ★
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../%E7%AC%AC%209%20%E7%AB%A0%20%E4%BC%98%E5%85%88%E7%BA%A7%E9%98%9F%E5%88%97/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    第 9 章 优先级队列
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    💡 Reviews
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            💡 Reviews
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_1" >
        
          
          <label class="md-nav__link" for="__nav_7_1" id="__nav_7_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🧭 scRNA analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_1">
            <span class="md-nav__icon md-icon"></span>
            🧭 scRNA analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Best%20practices%20for%20single-cell%20analysis%20across%20modalities/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Best practices for single cell analysis across modalities
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Multi-omics%20integration%20in%20biomedical%20research%20%E2%80%93%20A%20metabolomics-centric%20review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multi omics integration in biomedical research – A metabolomics centric review
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Museum%20of%20spatial%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Museum of spatial transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../A%20guide%20to%20artificial%20intelligence%20for%20cancer%20researchers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    A guide to artificial intelligence for cancer researchers
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_2" >
        
          
          <label class="md-nav__link" for="__nav_7_2" id="__nav_7_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    🦠 Metabolism
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_2">
            <span class="md-nav__icon md-icon"></span>
            🦠 Metabolism
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20analysis%20as%20a%20driver%20for%20discovery%2C%20diagnosis%2C%20and%20therapy/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic analysis as a driver for discovery, diagnosis, and therapy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20heterogeneity%20in%20cancer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic heterogeneity in cancer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Toward%20modeling%20metabolic%20state%20from%20single-cell%20transcriptomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Toward modeling metabolic state from single cell transcriptomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metabolic%20reprogramming%20and%20cancer%20progression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolic reprogramming and cancer progression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7_3" >
        
          
          <label class="md-nav__link" for="__nav_7_3" id="__nav_7_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Protein
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7_3">
            <span class="md-nav__icon md-icon"></span>
            Protein
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Proteomic%20and%20interactomic%20insights%20into%20the%20molecular%20basis%20of%20cell%20functional%20diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Proteomic and interactomic insights into the molecular basis of cell functional diversity
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#autograd" class="md-nav__link">
    <span class="md-ellipsis">
      Autograd 如何记录历史
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    <span class="md-ellipsis">
      已保存的张量
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      非可微函数的梯度
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      局部禁用梯度计算
    </span>
  </a>
  
    <nav class="md-nav" aria-label="局部禁用梯度计算">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#requires_grad" class="md-nav__link">
    <span class="md-ellipsis">
      设置 requires_grad
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grad-modes" class="md-nav__link">
    <span class="md-ellipsis">
      Grad Modes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      默认模式（梯度模式）
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      无梯度模式
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      推理模式
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#nnmoduleeval" class="md-nav__link">
    <span class="md-ellipsis">
      评估模式 (nn.Module.eval())
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#autograd_1" class="md-nav__link">
    <span class="md-ellipsis">
      使用 Autograd 的就地操作
    </span>
  </a>
  
    <nav class="md-nav" aria-label="使用 Autograd 的就地操作">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_7" class="md-nav__link">
    <span class="md-ellipsis">
      就地正确性检查
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#autograd_2" class="md-nav__link">
    <span class="md-ellipsis">
      多线程 Autograd
    </span>
  </a>
  
    <nav class="md-nav" aria-label="多线程 Autograd">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cpu" class="md-nav__link">
    <span class="md-ellipsis">
      CPU 并发
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_8" class="md-nav__link">
    <span class="md-ellipsis">
      不确定性
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_9" class="md-nav__link">
    <span class="md-ellipsis">
      保留计算图
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#autograd_3" class="md-nav__link">
    <span class="md-ellipsis">
      Autograd 节点上的线程安全
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#c" class="md-nav__link">
    <span class="md-ellipsis">
      C++ 钩子上的线程不安全
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#autograd_4" class="md-nav__link">
    <span class="md-ellipsis">
      复数的 Autograd
    </span>
  </a>
  
    <nav class="md-nav" aria-label="复数的 Autograd">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_10" class="md-nav__link">
    <span class="md-ellipsis">
      什么是复数导数？
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_11" class="md-nav__link">
    <span class="md-ellipsis">
      维尔廷格微积分登场
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_12" class="md-nav__link">
    <span class="md-ellipsis">
      维尔廷格微积分在优化中的作用
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pytorch" class="md-nav__link">
    <span class="md-ellipsis">
      PyTorch 如何计算共轭维尔廷格导数？
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_13" class="md-nav__link">
    <span class="md-ellipsis">
      如何为复数函数编写导数公式？
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_14" class="md-nav__link">
    <span class="md-ellipsis">
      跨域函数呢？
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_15" class="md-nav__link">
    <span class="md-ellipsis">
      已保存张量的钩子
    </span>
  </a>
  
    <nav class="md-nav" aria-label="已保存张量的钩子">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#_16" class="md-nav__link">
    <span class="md-ellipsis">
      反向传播钩子的执行
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_17" class="md-nav__link">
    <span class="md-ellipsis">
      钩子是否会触发
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_18" class="md-nav__link">
    <span class="md-ellipsis">
      钩子触发顺序
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_19" class="md-nav__link">
    <span class="md-ellipsis">
      特殊钩子
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_20" class="md-nav__link">
    <span class="md-ellipsis">
      修改张量时的钩子行为
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tensorregister_hookhook" class="md-nav__link">
    <span class="md-ellipsis">
      Tensor.register_hook(hook)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/aletolia/aletolia.github.io.git/edit/master/docs/Autograd mechanics — PyTorch 2_3 documentation.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/aletolia/aletolia.github.io.git/raw/master/docs/Autograd mechanics — PyTorch 2_3 documentation.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


  <h1>Autograd mechanics — PyTorch 2 3 documentation</h1>

<p>这篇说明将概述 autograd 如何运行以及记录操作。虽然没有必要掌握所有细节，但我们建议您熟悉这些内容，因为这有助于编写更高效、清晰的程序，并能在调试中提供帮助。</p>
<h3 id="autograd">Autograd 如何记录历史<a class="headerlink" href="#autograd" title="Permanent link">&para;</a></h3>
<p>Autograd 是一种反向自动微分系统。从概念上讲，autograd 在您执行操作的同时会记录一张包含所有创建数据操作的图，并形成一个有向无环图（DAG），其叶节点是输入张量，根节点是输出张量。通过从根节点追溯到叶节点，您可以自动使用链式法则计算梯度。</p>
<p>在内部，autograd 使用 <code>Function</code> 对象（实际是表达式）来表示这张图，通过 <code>apply()</code> 方法可以计算该图的输出。在计算正向传递的同时，autograd 会执行所请求的计算并构建一张图，用于计算梯度（每个 <a href="https://pytorch.org/docs/stable/tensors.html#torch.Tensor" title="torch.Tensor"><code>torch.Tensor</code></a> 的 <code>.grad_fn</code> 属性就是该图的入口）。当正向传递完成后，我们会在反向传递中评估这张图以计算梯度。</p>
<p>值得注意的是，这张图在每次迭代时都会从头开始重建，这使得我们可以使用任意的 Python 控制流语句，从而在每次迭代时更改图的整体形状和大小。您无需在启动训练之前预先编码所有可能的路径——运行的内容就是您需要微分的内容。</p>
<h3 id="_1">已保存的张量<a class="headerlink" href="#_1" title="Permanent link">&para;</a></h3>
<p>某些操作在正向传递期间需要保存中间结果，以在反向传递中执行。例如，函数 <span class="arithmatex">\( x \mapsto x^2 \)</span> 会保存输入 <span class="arithmatex">\( x \)</span> 以计算梯度。</p>
<p>在定义自定义 Python <a href="https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function" title="torch.autograd.Function"><code>Function</code></a> 时，可以使用 <code>save_for_backward()</code> 在正向传递中保存张量，并通过 <code>saved_tensors</code> 在反向传递中检索它们。有关详细信息，请参阅 <a href="https://pytorch.org/docs/stable/notes/extending.html">扩展 PyTorch</a>。</p>
<p>对于 PyTorch 定义的操作（如 <a href="https://pytorch.org/docs/stable/generated/torch.pow.html#torch.pow" title="torch.pow"><code>torch.pow()</code></a>），需要时会自动保存张量。为了学习或调试的目的，可以通过查找其以 <code>_saved</code> 为前缀的属性，查看特定 <code>grad_fn</code> 保存了哪些张量。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">equal</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">grad_fn</span><span class="o">.</span><span class="n">_saved_self</span><span class="p">))</span>  <span class="c1"># True</span>
<span class="linenos" data-linenos="4 "></span><span class="nb">print</span><span class="p">(</span><span class="n">x</span> <span class="ow">is</span> <span class="n">y</span><span class="o">.</span><span class="n">grad_fn</span><span class="o">.</span><span class="n">_saved_self</span><span class="p">)</span>  <span class="c1"># True</span>
</code></pre></div>
<p>在上面的代码中，<code>y.grad_fn._saved_self</code> 指向与 <code>x</code> 相同的张量对象，但情况并非总是如此。例如：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">exp</span><span class="p">()</span>
<span class="linenos" data-linenos="3 "></span><span class="nb">print</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">equal</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">grad_fn</span><span class="o">.</span><span class="n">_saved_result</span><span class="p">))</span>  <span class="c1"># True</span>
<span class="linenos" data-linenos="4 "></span><span class="nb">print</span><span class="p">(</span><span class="n">y</span> <span class="ow">is</span> <span class="n">y</span><span class="o">.</span><span class="n">grad_fn</span><span class="o">.</span><span class="n">_saved_result</span><span class="p">)</span>  <span class="c1"># False</span>
</code></pre></div>
<p>在内部，PyTorch 保存张量时使用<strong>打包</strong>机制以防止引用循环，然后在读取时<strong>解包</strong>为不同的张量。在此例中，<code>y.grad_fn._saved_result</code> 中获取的张量对象与 <code>y</code> 并不相同（但它们共享相同的存储）。</p>
<p>张量是否会被打包成不同的张量对象取决于它是否是其自身 <code>grad_fn</code> 的输出，这是实现细节，可能会发生变化，不应作为依赖。</p>
<p>您可以通过 <a href="https://pytorch.org/docs/stable/notes/autograd.html#saved-tensors-hooks-doc">已保存张量的钩子</a> 来控制 PyTorch 如何打包/解包。</p>
<h3 id="_2">非可微函数的梯度<a class="headerlink" href="#_2" title="Permanent link">&para;</a></h3>
<p>自动微分中的梯度计算仅在每个使用的基本函数均可微时才有效。然而，许多实际使用的函数不具有这一性质（例如 <code>relu</code> 或 <code>sqrt</code> 在 0 处）。为了尽量减少非可微函数的影响，我们按照以下规则定义基本操作的梯度：</p>
<ol>
<li>
<p>如果函数在当前点可微并且存在梯度，则直接使用该梯度。</p>
</li>
<li>
<p>如果函数是凸的（至少局部凸），使用最小范数的次梯度（这是最陡下降方向）。</p>
</li>
<li>
<p>如果函数是凹的（至少局部凹），使用最小范数的上梯度（考虑 -f(x) 并应用前一点）。</p>
</li>
<li>
<p>如果函数定义明确，通过连续性在当前点定义梯度（此处可能为无穷，例如 <code>sqrt(0)</code>）。若存在多个可能值，则随机选取一个。</p>
</li>
<li>
<p>如果函数未定义（例如 <code>sqrt(-1)</code>、<code>log(-1)</code> 或在输入为 <code>NaN</code> 时的大多数函数），所用的梯度值是任意的（我们可能会抛出错误，但不保证一定如此）。大多数函数会将梯度设为 <code>NaN</code>，但出于性能原因，部分函数会使用其他值（例如 <code>log(-1)</code>）。</p>
</li>
<li>
<p>如果函数不是确定性映射（即不是一个<a href="https://en.wikipedia.org/wiki/Function_(mathematics)" title="数学函数">数学函数</a>），则会被标记为不可微分。如果在需要梯度的张量上使用此函数且不在 <code>no_grad</code> 环境中，它会在反向传播时抛出错误。</p>
</li>
</ol>
<h3 id="_3">局部禁用梯度计算<a class="headerlink" href="#_3" title="Permanent link">&para;</a></h3>
<p>在 Python 中有几种机制可以局部禁用梯度计算：</p>
<p>为了禁用整个代码块的梯度计算，可以使用无梯度模式（no-grad mode）和推理模式（inference mode）这类上下文管理器。对于更精细地从梯度计算中排除子图，可以设置张量的 <code>requires_grad</code> 属性。</p>
<p>除了讨论上述机制外，我们还描述了评估模式（<code>nn.Module.eval()</code>）。尽管它不用于禁用梯度计算，但其名称常与其他三者混淆。</p>
<h4 id="requires_grad">设置 <code>requires_grad</code><a class="headerlink" href="#requires_grad" title="Permanent link">&para;</a></h4>
<p><code>requires_grad</code> 是一个标志，默认情况下为 <code>False</code>（除非包含在 <code>nn.Parameter</code> 中），用于细粒度地将子图从梯度计算中排除。它在正向和反向传递中都会生效：</p>
<p>在正向传递中，只有当操作的至少一个输入张量需要梯度时，才会记录该操作。在反向传递（<code>.backward()</code>）中，只有 <code>requires_grad=True</code> 的叶张量才会将梯度累积到它们的 <code>.grad</code> 字段。</p>
<p>需要注意的是，即使每个张量都有此标志，<strong>设置</strong>它仅对叶张量（没有 <code>grad_fn</code> 的张量，例如 <code>nn.Module</code> 的参数）有效。非叶张量（具有 <code>grad_fn</code> 的张量）关联了反向传播图，因此它们的梯度是计算需要梯度的叶张量的梯度的中间结果。从定义上看，所有非叶张量都自动具有 <code>requires_grad=True</code>。</p>
<p>设置 <code>requires_grad</code> 应该是您控制模型的哪些部分参与梯度计算的主要方式。例如，如果需要在模型微调期间冻结预训练模型的部分参数，可以直接将不希望更新的参数设为 <code>.requires_grad_(False)</code>。正如上文所述，由于这些参数作为输入的计算不会记录在正向传递中，因此在反向传递中它们的 <code>.grad</code> 字段不会更新，因为它们最初就不在反向传播图中。</p>
<p>由于这种模式很常见，<code>requires_grad</code> 也可以在模块级别通过 <code>nn.Module.requires_grad_()</code> 设置。对模块应用 <code>.requires_grad_()</code> 会影响该模块的所有参数（默认情况下 <code>requires_grad=True</code>）。</p>
<h3 id="grad-modes">Grad Modes<a href="https://pytorch.org/docs/stable/notes/autograd.html#grad-modes"></a><a class="headerlink" href="#grad-modes" title="Permanent link">&para;</a></h3>
<p>Apart from setting <code>requires_grad</code> there are also three grad modes that can be selected from Python that can affect how computations in PyTorch are processed by autograd internally: default mode (grad mode), no-grad mode, and inference mode, all of which can be togglable via context managers and decorators.</p>
<table><colgroup><col> <col> <col> <col> <col> </colgroup><thead><tr><th><p>Mode</p></th><th><p>Excludes operations from being recorded in backward graph</p></th><th><p>Skips additional autograd tracking overhead</p></th><th><p>Tensors created while the mode is enabled can be used in grad-mode later</p></th><th><p>Examples</p></th></tr></thead><tbody><tr><td><p>default</p></td><td></td><td></td><td><p>✓</p></td><td><p>Forward pass</p></td></tr><tr><td><p>no-grad</p></td><td><p>✓</p></td><td></td><td><p>✓</p></td><td><p>Optimizer updates</p></td></tr><tr><td><p>inference</p></td><td><p>✓</p></td><td><p>✓</p></td><td></td><td><p>Data processing, model evaluation</p></td></tr></tbody></table>
<h3 id="_4">默认模式（梯度模式）<a class="headerlink" href="#_4" title="Permanent link">&para;</a></h3>
<p>“默认模式”是我们在未启用无梯度模式或推理模式时所处的模式。与“无梯度模式”相对比，默认模式有时也称为“梯度模式”。</p>
<p>默认模式最重要的一点是，这是唯一一个 <code>requires_grad</code> 属性生效的模式。在无梯度模式和推理模式中，<code>requires_grad</code> 都会被强制设置为 <code>False</code>。</p>
<h3 id="_5">无梯度模式<a class="headerlink" href="#_5" title="Permanent link">&para;</a></h3>
<p>在无梯度模式中，计算会表现得好像没有输入需要梯度一样。换句话说，即使输入张量设置了 <code>requires_grad=True</code>，在无梯度模式中的计算也不会记录在反向传播图中。</p>
<p>如果需要执行一些操作，这些操作不应由 autograd 记录，但您仍想在稍后以梯度模式使用这些计算的输出，此时可以启用无梯度模式。这个上下文管理器方便地让您禁用一段代码或函数的梯度，而无需临时将张量的 <code>requires_grad</code> 属性设置为 <code>False</code> 再恢复为 <code>True</code>。</p>
<p>例如，无梯度模式在编写优化器时很有用：执行训练更新时，您希望对参数进行就地更新，而无需让 autograd 记录该更新。您还打算在下一次前向传递中以梯度模式使用更新后的参数。</p>
<p>在 <a href="https://pytorch.org/docs/stable/nn.init.html#nn-init-doc">torch.nn.init</a> 中的实现也依赖于无梯度模式，以避免在初始化参数时，由 autograd 跟踪就地更新的已初始化参数。</p>
<h3 id="_6">推理模式<a class="headerlink" href="#_6" title="Permanent link">&para;</a></h3>
<p>推理模式是无梯度模式的增强版本。与无梯度模式一样，推理模式中的计算不会记录在反向传播图中，但启用推理模式可以进一步加快模型的运行速度。这个性能提升带来了一个缺点：在推理模式中创建的张量在退出推理模式后无法用于由 autograd 记录的计算中。</p>
<p>如果您在执行的计算不需要记录在反向传播图中，且您不打算在后续的任何由 autograd 记录的计算中使用推理模式创建的张量，那么可以启用推理模式。</p>
<p>建议您在代码中不需要 autograd 跟踪的部分（例如数据处理和模型评估）尝试使用推理模式。如果您的用例直接奏效，那就是免费的性能提升。如果启用推理模式后出现错误，请检查您是否在退出推理模式后，将推理模式创建的张量用于由 autograd 记录的计算中。如果无法避免这种情况，可以随时切换回无梯度模式。</p>
<p>有关推理模式的详细信息，请参阅 <a href="https://pytorch.org/cppdocs/notes/inference_mode.html">推理模式</a>。</p>
<p>有关推理模式的实现细节，请参阅 <a href="https://github.com/pytorch/rfcs/pull/17">RFC-0011-InferenceMode</a>。</p>
<h3 id="nnmoduleeval">评估模式 (<code>nn.Module.eval()</code>)<a class="headerlink" href="#nnmoduleeval" title="Permanent link">&para;</a></h3>
<p>评估模式并不是一种局部禁用梯度计算的机制，但仍在此列出，因为它有时会被误认为是禁用梯度的方式。</p>
<p>从功能上讲，<code>module.eval()</code>（或等价的 <code>module.train(False)</code>）与无梯度模式和推理模式完全无关。<code>model.eval()</code> 如何影响您的模型，完全取决于模型中使用的特定模块以及它们是否定义了训练模式相关的行为。</p>
<p>如果您的模型依赖于可能根据训练模式有不同行为的模块（例如 <a href="https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html#torch.nn.Dropout" title="torch.nn.Dropout"><code>torch.nn.Dropout</code></a> 和 <a href="https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html#torch.nn.BatchNorm2d" title="torch.nn.BatchNorm2d"><code>torch.nn.BatchNorm2d</code></a>），请确保自行调用 <code>model.eval()</code> 和 <code>model.train()</code>，以避免在验证数据上更新 BatchNorm 的运行统计信息。</p>
<p>即使您不确定模型是否具有训练模式相关的行为，建议您在训练时始终使用 <code>model.train()</code>，在评估时（验证/测试）使用 <code>model.eval()</code>。因为您使用的模块可能会更新，从而在训练和评估模式中表现不同。</p>
<h2 id="autograd_1">使用 Autograd 的就地操作<a class="headerlink" href="#autograd_1" title="Permanent link">&para;</a></h2>
<p>在 autograd 中支持就地操作非常困难，在大多数情况下我们不建议使用。autograd 具有积极的缓冲区释放和重用策略，使其非常高效，几乎没有情况能够通过就地操作显著降低内存使用。除非在极高的内存压力下操作，否则几乎不需要使用它们。</p>
<p>就地操作适用范围受限的两个主要原因是：</p>
<ol>
<li>
<p>就地操作可能会覆盖计算梯度所需的值。</p>
</li>
<li>
<p>每个就地操作都要求实现重写计算图。非就地版本会分配新对象并保留对旧图的引用，而就地操作则要求更改代表该操作的 <code>Function</code> 所有输入的创建者。这可能很棘手，尤其是当许多张量共享相同的存储空间时（例如，通过索引或转置创建）。如果其他任何 <code>Tensor</code> 引用了已修改输入的存储空间，就地函数会引发错误。</p>
</li>
</ol>
<h3 id="_7">就地正确性检查<a class="headerlink" href="#_7" title="Permanent link">&para;</a></h3>
<p>每个张量都有一个版本计数器，在任何操作中标记为已更改时，该计数器都会递增。当 <code>Function</code> 保存任何张量用于反向传播时，会同时保存其包含张量的版本计数器。一旦您访问 <code>self.saved_tensors</code>，它就会被检查，如果其值大于已保存的值，将引发错误。这确保了如果您使用就地函数且未看到任何错误，可以确信计算的梯度是正确的。</p>
<h2 id="autograd_2">多线程 Autograd<a class="headerlink" href="#autograd_2" title="Permanent link">&para;</a></h2>
<p>autograd 引擎负责执行所有反向操作以完成反向传播。本节将描述在多线程环境中如何最好地利用它的所有细节（这仅适用于 PyTorch 1.6 及以上版本，因为之前版本的行为不同）。</p>
<p>用户可以使用多线程代码（例如 Hogwild 训练）训练模型，并且不会被并发的反向计算阻塞，示例代码如下：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 定义一个可在不同线程中使用的训练函数</span>
<span class="linenos" data-linenos=" 2 "></span><span class="k">def</span> <span class="nf">train_fn</span><span class="p">():</span>
<span class="linenos" data-linenos=" 3 "></span>    <span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos=" 4 "></span>    <span class="c1"># 正向</span>
<span class="linenos" data-linenos=" 5 "></span>    <span class="n">y</span> <span class="o">=</span> <span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">4</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.5</span>
<span class="linenos" data-linenos=" 6 "></span>    <span class="c1"># 反向</span>
<span class="linenos" data-linenos=" 7 "></span>    <span class="n">y</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">backward</span><span class="p">()</span>
<span class="linenos" data-linenos=" 8 "></span>    <span class="c1"># 可能的优化器更新</span>
<span class="linenos" data-linenos=" 9 "></span>
<span class="linenos" data-linenos="10 "></span><span class="c1"># 用户编写自己的线程代码来驱动 train_fn</span>
<span class="linenos" data-linenos="11 "></span><span class="n">threads</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="12 "></span><span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
<span class="linenos" data-linenos="13 "></span>    <span class="n">p</span> <span class="o">=</span> <span class="n">threading</span><span class="o">.</span><span class="n">Thread</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">train_fn</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">())</span>
<span class="linenos" data-linenos="14 "></span>    <span class="n">p</span><span class="o">.</span><span class="n">start</span><span class="p">()</span>
<span class="linenos" data-linenos="15 "></span>    <span class="n">threads</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="linenos" data-linenos="16 "></span>
<span class="linenos" data-linenos="17 "></span><span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">threads</span><span class="p">:</span>
<span class="linenos" data-linenos="18 "></span>    <span class="n">p</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
</code></pre></div>
<p>需要注意用户应该了解以下行为：</p>
<h3 id="cpu">CPU 并发<a class="headerlink" href="#cpu" title="Permanent link">&para;</a></h3>
<p>在 CPU 上通过 Python 或 C++ API 以多线程方式运行 <code>backward()</code> 或 <code>grad()</code> 时，应看到更多的并发执行，而不是按特定顺序将所有反向调用串行化（这是 PyTorch 1.6 之前的行为）。</p>
<h3 id="_8">不确定性<a class="headerlink" href="#_8" title="Permanent link">&para;</a></h3>
<p>如果您在多线程环境中并发调用 <code>backward()</code> 并共享输入（即 Hogwild CPU 训练），则应预期出现不确定性。这种情况发生的原因在于参数在线程间自动共享，因此多个线程可能会尝试访问并累积相同的 <code>.grad</code> 属性。这在技术上并不安全，可能导致竞争条件，使结果无效。</p>
<p>开发多线程模型并共享参数的用户，应牢记多线程模型并了解上述问题。</p>
<p>可以使用功能性 API <a href="https://pytorch.org/docs/stable/generated/torch.autograd.grad.html"><code>torch.autograd.grad()</code></a> 来计算梯度，以避免不确定性，而非使用 <code>backward()</code>。</p>
<h3 id="_9">保留计算图<a class="headerlink" href="#_9" title="Permanent link">&para;</a></h3>
<p>如果 autograd 图的一部分在线程之间共享，例如，先在单线程中运行正向传播的第一部分，然后在多线程中运行后续部分，则第一部分图会被共享。在这种情况下，不同的线程在同一个图上执行 <code>grad()</code> 或 <code>backward()</code> 可能会遇到一个线程在执行时销毁图的问题，导致其他线程崩溃。autograd 会向用户报告错误，类似于未设置 <code>retain_graph=True</code> 而多次调用 <code>backward()</code>，并提醒用户应使用 <code>retain_graph=True</code>。</p>
<h3 id="autograd_3">Autograd 节点上的线程安全<a class="headerlink" href="#autograd_3" title="Permanent link">&para;</a></h3>
<p>由于 Autograd 允许调用线程驱动其反向执行以实现潜在的并行性，因此确保在 CPU 上以并行方式调用 <code>backward()</code> 时线程安全性很重要，特别是共享部分或整个 GraphTask 时。</p>
<p>自定义 Python <code>autograd.Function</code> 自动因 GIL 而线程安全。对于内置的 C++ Autograd 节点（如 AccumulateGrad、CopySlices）和自定义 <code>autograd::Function</code>，Autograd 引擎使用线程互斥锁来确保可能存在状态读/写的 autograd 节点上的线程安全。</p>
<h3 id="c">C++ 钩子上的线程不安全<a class="headerlink" href="#c" title="Permanent link">&para;</a></h3>
<p>Autograd 依赖用户编写线程安全的 C++ 钩子。如果希望钩子在多线程环境中正确应用，需要编写适当的线程锁定代码，确保钩子线程安全。</p>
<h2 id="autograd_4">复数的 Autograd<a class="headerlink" href="#autograd_4" title="Permanent link">&para;</a></h2>
<p>简要说明：</p>
<ul>
<li>
<p>当您使用 PyTorch 对任意复数域和/或复数值域的函数 <span class="arithmatex">\(f(z)\)</span> 求导时，梯度是基于该函数作为更大实值损失函数 <span class="arithmatex">\(g(\)</span> input <span class="arithmatex">\()=L\)</span> 的一部分来计算的。计算的梯度是 <span class="arithmatex">\(\frac{\partial L}{\partial z^*}\)</span>（注意 <span class="arithmatex">\(z\)</span> 的共轭），其负值正是梯度下降算法中最陡下降的方向。因此，所有现有的优化器都可以直接处理复数参数。</p>
</li>
<li>
<p>这种方式与 TensorFlow 的复数微分惯例相符，但与 JAX 不同（JAX 计算的是 <span class="arithmatex">\(\frac{\partial L}{\partial z}\)</span>）。</p>
</li>
<li>
<p>如果有一个实到实的函数，内部使用复数运算，此处的惯例无关紧要：结果将始终与仅使用实数运算的实现相同。</p>
</li>
</ul>
<p>如果您对数学细节感兴趣，或者想知道如何在 PyTorch 中定义复数导数，请继续阅读。</p>
<h3 id="_10">什么是复数导数？<a class="headerlink" href="#_10" title="Permanent link">&para;</a></h3>
<p>复数可微分的数学定义取导数的极限定义，并将其推广到复数。考虑函数 <span class="arithmatex">\(f:\mathbb{C} \rightarrow \mathbb{C}\)</span>，</p>
<div class="arithmatex">\[
f(z = x + yj) = u(x, y) + v(x, y)j
\]</div>
<p>其中 <span class="arithmatex">\(u\)</span> 和 <span class="arithmatex">\(v\)</span> 是两个变量的实值函数，<span class="arithmatex">\(j\)</span> 是虚数单位。</p>
<p>使用导数定义，我们可以写出：</p>
<div class="arithmatex">\[
f^{\prime}(z) = \lim_{h \rightarrow 0, h \in C}\frac{f(z+h) - f(z)}{h}
\]</div>
<p>为了使此极限存在，<span class="arithmatex">\(u\)</span> 和 <span class="arithmatex">\(v\)</span> 不仅必须是实可微的，<span class="arithmatex">\(f\)</span> 还必须满足柯西-黎曼方程。换句话说，按实数和虚数步长 <span class="arithmatex">\((h)\)</span> 计算的极限必须相等。这是一个更为严格的条件。</p>
<p>复数可微的函数通常被称为全纯函数。它们性质良好，具有实数可微函数的所有优良属性，但在优化领域几乎无用。对于优化问题，研究界通常只使用实值目标函数，因为复数不是有序域的一部分，因此使用复数值损失没有多大意义。</p>
<p>此外，没有有趣的实值目标函数符合柯西-黎曼方程。因此，全纯函数的理论不能用于优化，大多数人转而使用维尔廷格微积分。</p>
<h3 id="_11">维尔廷格微积分登场<a class="headerlink" href="#_11" title="Permanent link">&para;</a></h3>
<p>我们拥有复数可微性和全纯函数的理论，但由于许多常用函数并非全纯，无法充分利用这一理论。那么数学家 Wirtinger 是怎么做的呢？他发现，即使 <span class="arithmatex">\(f(z)\)</span> 不是全纯的，也可以将其重写为双变量函数 <span class="arithmatex">\(f(z, z^*)\)</span>，该函数总是全纯的。这是因为 <span class="arithmatex">\(z\)</span> 的实部和虚部可以通过 <span class="arithmatex">\(z\)</span> 和 <span class="arithmatex">\(z^*\)</span> 表示为：</p>
<div class="arithmatex">\[
\begin{aligned}
\operatorname{Re}(z) = \frac{z + z^*}{2} \\
\operatorname{Im}(z) = \frac{z - z^*}{2j}
\end{aligned}
\]</div>
<p>维尔廷格微积分建议研究 <span class="arithmatex">\(f(z, z^*)\)</span>，如果 <span class="arithmatex">\(f\)</span> 实可微，则该函数可保证全纯（另一种看法是这是坐标系的改变，从 <span class="arithmatex">\(f(x, y)\)</span> 变为 <span class="arithmatex">\(f(z, z^*)\)</span>）。该函数具有偏导数 <span class="arithmatex">\(\frac{\partial}{\partial z}\)</span> 和 <span class="arithmatex">\(\frac{\partial}{\partial z^*}\)</span>。我们可以使用链式法则建立这些偏导数与 <span class="arithmatex">\(z\)</span> 的实部和虚部的偏导数之间的关系：</p>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial}{\partial x} = \frac{\partial z}{\partial x} \cdot \frac{\partial}{\partial z} + \frac{\partial z^*}{\partial x} \cdot \frac{\partial}{\partial z^*} \\
\frac{\partial}{\partial y} = \frac{\partial z}{\partial y} \cdot \frac{\partial}{\partial z} + \frac{\partial z^*}{\partial y} \cdot \frac{\partial}{\partial z^*}
\end{aligned}
\]</div>
<p>从上述方程可得：</p>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial}{\partial z} = \frac{1}{2}\left(\frac{\partial}{\partial x} - 1j \cdot \frac{\partial}{\partial y}\right) \\
\frac{\partial}{\partial z^*} = \frac{1}{2}\left(\frac{\partial}{\partial x} + 1j \cdot \frac{\partial}{\partial y}\right)
\end{aligned}
\]</div>
<p>这就是您在维基百科上找到的维尔廷格微积分的经典定义。</p>
<p>这个改变带来了许多有趣的结果：</p>
<ul>
<li>
<p>首先，柯西-黎曼方程简单地翻译成 <span class="arithmatex">\(\frac{\partial f}{\partial z^*} = 0\)</span>（即函数 <span class="arithmatex">\(f\)</span> 可以完全用 <span class="arithmatex">\(z\)</span> 表示，不必涉及 <span class="arithmatex">\(z^*\)</span>）。</p>
</li>
<li>
<p>另一个重要（且看似反直觉）的结果是，当我们对实值损失进行优化时，更新变量的步骤由 <span class="arithmatex">\(\frac{\partial \text{Loss}}{\partial z^*}\)</span> 决定（而不是 <span class="arithmatex">\(\frac{\partial \text{Loss}}{\partial z}\)</span>）。</p>
</li>
</ul>
<p>如需了解更多信息，请查阅：<a href="https://arxiv.org/pdf/0906.4835.pdf">https://arxiv.org/pdf/0906.4835.pdf</a></p>
<h3 id="_12">维尔廷格微积分在优化中的作用<a class="headerlink" href="#_12" title="Permanent link">&para;</a></h3>
<p>音频等领域的研究者通常使用梯度下降法，通过复变量来优化实值损失函数。这些研究者通常将实部和虚部分开处理，视为可独立更新的两个通道。对于步长 <span class="arithmatex">\(\alpha / 2\)</span> 和损失 <span class="arithmatex">\(L\)</span>，我们可以在 <span class="arithmatex">\(\mathbb{R}^2\)</span> 中写出以下公式：</p>
<div class="arithmatex">\[
\begin{aligned}
x_{n+1} &amp; = x_n - (\alpha / 2) \cdot \frac{\partial L}{\partial x} \\
y_{n+1} &amp; = y_n - (\alpha / 2) \cdot \frac{\partial L}{\partial y}
\end{aligned}
\]</div>
<p>这些公式在复数空间 <span class="arithmatex">\(\mathbb{C}\)</span> 中如何转换呢？</p>
<div class="arithmatex">\[
\begin{aligned}
z_{n+1} &amp; = x_n - (\alpha / 2) \cdot \frac{\partial L}{\partial x} + 1j \cdot \left(y_n - (\alpha / 2) \cdot \frac{\partial L}{\partial y}\right) \\
&amp; = z_n - \alpha \cdot (1 / 2) \cdot \left(\frac{\partial L}{\partial x} + j \frac{\partial L}{\partial y}\right) \\
&amp; = z_n - \alpha \cdot \frac{\partial L}{\partial z^*}
\end{aligned}
\]</div>
<p>这里发生了一件非常有趣的事情：维尔廷格微积分告诉我们可以将上述复变量更新公式简化，只引用共轭的维尔廷格导数 <span class="arithmatex">\(\frac{\partial L}{\partial z^*}\)</span>，这样我们在优化中就可以得到正确的更新步骤。</p>
<p>由于共轭维尔廷格导数能够给出实值损失函数的精确更新步骤，当使用 PyTorch 对具有实值损失的函数求导时，您将获得此导数。</p>
<h3 id="pytorch">PyTorch 如何计算共轭维尔廷格导数？<a class="headerlink" href="#pytorch" title="Permanent link">&para;</a></h3>
<p>通常情况下，我们的导数公式将 <code>grad_output</code> 作为输入，它代表已计算的向量-雅可比积，也即 <span class="arithmatex">\(\frac{\partial L}{\partial s^*}\)</span>，其中 <span class="arithmatex">\(L\)</span> 是整个计算的损失（产生实值损失），<span class="arithmatex">\(s\)</span> 是函数的输出。此处的目标是计算 <span class="arithmatex">\(\frac{\partial L}{\partial z^*}\)</span>，其中 <span class="arithmatex">\(z\)</span> 是函数的输入。实际上，对于实值损失来说，我们只需计算 <span class="arithmatex">\(\frac{\partial L}{\partial s^*}\)</span>，尽管链式法则表明我们还需要获取 <span class="arithmatex">\(\frac{\partial L}{\partial s}\)</span>。如果您想跳过此推导，可以直接查看本节的最后一个公式，然后跳到下一节。</p>
<p>继续研究 <span class="arithmatex">\(f: \mathbb{C} \to \mathbb{C}\)</span>，定义为 <span class="arithmatex">\(f(z) = f(x + yj) = u(x, y) + v(x, y)j\)</span>。如上所述，autograd 的梯度规范围绕实值损失函数的优化，因此假设 <span class="arithmatex">\(f\)</span> 是更大实值损失函数 <span class="arithmatex">\(g\)</span> 的一部分。使用链式法则，我们可以写出：</p>
<div class="arithmatex">\[
\frac{\partial L}{\partial z^*} = \frac{\partial L}{\partial u} \cdot \frac{\partial u}{\partial z^*} + \frac{\partial L}{\partial v} \cdot \frac{\partial v}{\partial z^*}
\]</div>
<p>现在使用维尔廷格导数的定义，可以写出：</p>
<div class="arithmatex">\[
\begin{aligned}
&amp; \frac{\partial L}{\partial s} = \frac{1}{2}\left(\frac{\partial L}{\partial u} - \frac{\partial L}{\partial v}j\right) \\
&amp; \frac{\partial L}{\partial s^*} = \frac{1}{2}\left(\frac{\partial L}{\partial u} + \frac{\partial L}{\partial v}j\right)
\end{aligned}
\]</div>
<p>需要注意的是，由于 <span class="arithmatex">\(u\)</span> 和 <span class="arithmatex">\(v\)</span> 是实函数，而假设 <span class="arithmatex">\(f\)</span> 是实值函数的一部分，因此 <span class="arithmatex">\(L\)</span> 也是实数。我们有：</p>
<div class="arithmatex">\[
\left(\frac{\partial L}{\partial s}\right)^* = \frac{\partial L}{\partial s^*}
\]</div>
<p>也就是说，<span class="arithmatex">\(\frac{\partial L}{\partial s}\)</span> 等于 <code>grad_output</code> 的共轭。</p>
<p>解出上述关于 <span class="arithmatex">\(\frac{\partial L}{\partial u}\)</span> 和 <span class="arithmatex">\(\frac{\partial L}{\partial v}\)</span> 的方程，我们得到：</p>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial L}{\partial u} = \frac{\partial L}{\partial s} + \frac{\partial L}{\partial s^*} \\
\frac{\partial L}{\partial v} = -1j \cdot \left(\frac{\partial L}{\partial s} - \frac{\partial L}{\partial s^*}\right)
\end{aligned}
\]</div>
<p>将公式 (3) 代入 (1)，得到：</p>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial L}{\partial z^*} &amp; = \left(\frac{\partial L}{\partial s} + \frac{\partial L}{\partial s^*}\right) \cdot \frac{\partial u}{\partial z^*} - 1j \cdot \left(\frac{\partial L}{\partial s} - \frac{\partial L}{\partial s^*}\right) \cdot \frac{\partial v}{\partial z^*} \\
&amp; = \frac{\partial L}{\partial s} \cdot \left(\frac{\partial u}{\partial z^*} + \frac{\partial v}{\partial z^*}j\right) + \frac{\partial L}{\partial s^*} \cdot \left(\frac{\partial u}{\partial z^*} - \frac{\partial v}{\partial z^*}j\right) \\
&amp; = \frac{\partial L}{\partial s^*} \cdot \frac{\partial (u + vj)}{\partial z^*} + \frac{\partial L}{\partial s} \cdot \frac{\partial (u + vj)^*}{\partial z^*} \\
&amp; = \frac{\partial L}{\partial s} \cdot \frac{\partial s}{\partial z^*} + \frac{\partial L}{\partial s^*} \cdot \frac{\partial s^*}{\partial z^*}
\end{aligned}
\]</div>
<p>根据公式 (2)，我们得到：</p>
<p><img alt="image.png" src="https://raw.githubusercontent.com/aletolia/Pictures/main/202405051505485.png" /></p>
<p>最后一个公式非常重要，它将我们的导数公式分解为一个更简单、易于手动计算的形式。</p>
<h3 id="_13">如何为复数函数编写导数公式？<a class="headerlink" href="#_13" title="Permanent link">&para;</a></h3>
<p>上面框中的方程为复数函数的所有导数提供了通用公式，但我们仍需计算 <span class="arithmatex">\(\frac{\partial s}{\partial z}\)</span> 和 <span class="arithmatex">\(\frac{\partial s}{\partial z^*}\)</span>。以下是两种方法：</p>
<ul>
<li>
<p>第一种方法是直接使用维尔廷格导数的定义，通过 <span class="arithmatex">\(\frac{\partial s}{\partial x}\)</span> 和 <span class="arithmatex">\(\frac{\partial s}{\partial y}\)</span>（可以以常规方式计算）计算 <span class="arithmatex">\(\frac{\partial s}{\partial z}\)</span> 和 <span class="arithmatex">\(\frac{\partial s}{\partial z^*}\)</span>。</p>
</li>
<li>
<p>第二种方法是使用变量替换，将 <span class="arithmatex">\(f(z)\)</span> 重写为双变量函数 <span class="arithmatex">\(f(z, z^*)\)</span>，通过将 <span class="arithmatex">\(z\)</span> 和 <span class="arithmatex">\(z^*\)</span> 视为独立变量来计算共轭维尔廷格导数。这通常更容易；例如，如果函数全纯，则只使用 <span class="arithmatex">\(z\)</span>（且 <span class="arithmatex">\(\frac{\partial s}{\partial z^*}\)</span> 为零）。</p>
</li>
</ul>
<p>让我们以函数 <span class="arithmatex">\(f(z = x + yj) = c \cdot z = c \cdot (x + yj)\)</span> 为例，其中 <span class="arithmatex">\(c \in \mathbb{R}\)</span>。</p>
<p>使用第一种方法计算维尔廷格导数：</p>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial s}{\partial z} &amp; = \frac{1}{2}\left(\frac{\partial s}{\partial x} - \frac{\partial s}{\partial y}j\right) \\
&amp; = \frac{1}{2}(c - (c \cdot 1j) \cdot 1j) \\
&amp; = c
\end{aligned}
\]</div>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial s}{\partial z^*} &amp; = \frac{1}{2}\left(\frac{\partial s}{\partial x} + \frac{\partial s}{\partial y}j\right) \\
&amp; = \frac{1}{2}(c + (c \cdot 1j) \cdot 1j) \\
&amp; = 0
\end{aligned}
\]</div>
<p>根据公式 (4) 并设 <code>grad_output = 1.0</code>（这是在 PyTorch 中对标量输出调用 <code>backward()</code> 时的默认梯度输出值），我们得到：</p>
<div class="arithmatex">\[
\frac{\partial L}{\partial z^*} = 1 \cdot 0 + 1 \cdot c = c
\]</div>
<p>使用第二种方法计算维尔廷格导数，可以直接得到：</p>
<div class="arithmatex">\[
\begin{aligned}
\frac{\partial s}{\partial z} &amp; = \frac{\partial (c \cdot z)}{\partial z} \\
&amp; = c \\
\frac{\partial s}{\partial z^*} &amp; = \frac{\partial (c \cdot z)}{\partial z^*} \\
&amp; = 0
\end{aligned}
\]</div>
<p>再次使用公式 (4)，我们得到 <span class="arithmatex">\(\frac{\partial L}{\partial z^*} = c\)</span>。由此可见，第二种方法计算更少，更方便于快速计算。</p>
<h3 id="_14">跨域函数呢？<a class="headerlink" href="#_14" title="Permanent link">&para;</a></h3>
<p>有些函数从复数输入映射到实数输出，或反之亦然。这些函数是公式 (4) 的特殊情况，可以通过链式法则推导：</p>
<ul>
<li>对于 <span class="arithmatex">\(f: \mathbb{C} \to \mathbb{R}\)</span>，我们得到：</li>
</ul>
<div class="arithmatex">\[
\frac{\partial L}{\partial z^*} = 2 \cdot \text{grad\_output} \cdot \frac{\partial s}{\partial z^*}
\]</div>
<ul>
<li>对于 <span class="arithmatex">\(f: \mathbb{R} \to \mathbb{C}\)</span>，我们得到：</li>
</ul>
<div class="arithmatex">\[
\frac{\partial L}{\partial z^*} = 2 \cdot \operatorname{Re}\left(\text{grad\_output}^* \cdot \frac{\partial s}{\partial z^*}\right)
\]</div>
<h2 id="_15">已保存张量的钩子<a class="headerlink" href="#_15" title="Permanent link">&para;</a></h2>
<p>可以通过定义 <code>pack_hook</code> / <code>unpack_hook</code> 钩子对来控制已保存张量的打包/解包方式。<code>pack_hook</code> 函数应以张量作为唯一参数，但可以返回任何 Python 对象（例如另一个张量、元组，甚至包含文件名的字符串）。<code>unpack_hook</code> 函数的唯一参数是 <code>pack_hook</code> 的输出，并应返回用于反向传播的张量。<code>unpack_hook</code> 返回的张量只需与传递给 <code>pack_hook</code> 的张量内容相同，特别是任何与 autograd 相关的元数据都可以忽略，因为它们会在解包过程中被覆盖。</p>
<p>这对钩子的一个示例是：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="n">y</span><span class="o">.</span><span class="n">grad_fn</span><span class="o">.</span><span class="n">_raw_saved_self</span><span class="o">.</span><span class="n">register_hooks</span><span class="p">(</span><span class="n">pack_hook</span><span class="p">,</span> <span class="n">unpack_hook</span><span class="p">)</span>
</code></pre></div>
<p>在注册成对的钩子时立即调用 <code>pack_hook</code> 方法，而每次需要访问保存的张量时（通过 <code>y.grad_fn._saved_self</code> 或在反向传播期间），都会调用 <code>unpack_hook</code> 方法。</p>
<p><strong>警告</strong></p>
<p>如果在保存的张量已被释放（即反向传播调用后）仍保留对 <code>SavedTensor</code> 的引用，则禁止调用其 <code>register_hooks()</code>。PyTorch 通常会抛出错误，但在某些情况下可能不会抛错，并可能出现未定义行为。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1"># 仅将大小 &gt;= 1000 的张量保存在磁盘上</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">SAVE_ON_DISK_THRESHOLD</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="linenos" data-linenos=" 3 "></span>
<span class="linenos" data-linenos=" 4 "></span><span class="k">def</span> <span class="nf">pack_hook</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="linenos" data-linenos=" 5 "></span>    <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">SAVE_ON_DISK_THRESHOLD</span><span class="p">:</span>
<span class="linenos" data-linenos=" 6 "></span>        <span class="k">return</span> <span class="n">x</span>
<span class="linenos" data-linenos=" 7 "></span>    <span class="n">temp_file</span> <span class="o">=</span> <span class="n">SelfDeletingTempFile</span><span class="p">()</span>
<span class="linenos" data-linenos=" 8 "></span>    <span class="n">torch</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">tensor</span><span class="p">,</span> <span class="n">temp_file</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span>    <span class="k">return</span> <span class="n">temp_file</span>
<span class="linenos" data-linenos="10 "></span>
<span class="linenos" data-linenos="11 "></span><span class="k">def</span> <span class="nf">unpack_hook</span><span class="p">(</span><span class="n">tensor_or_sctf</span><span class="p">):</span>
<span class="linenos" data-linenos="12 "></span>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">tensor_or_sctf</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
<span class="linenos" data-linenos="13 "></span>        <span class="k">return</span> <span class="n">tensor_or_sctf</span>
<span class="linenos" data-linenos="14 "></span>    <span class="k">return</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">tensor_or_sctf</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="linenos" data-linenos="15 "></span>
<span class="linenos" data-linenos="16 "></span><span class="k">class</span> <span class="nc">Model</span><span class="p">(</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
<span class="linenos" data-linenos="17 "></span>    <span class="k">def</span> <span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
<span class="linenos" data-linenos="18 "></span>        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">autograd</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">saved_tensors_hooks</span><span class="p">(</span><span class="n">pack_hook</span><span class="p">,</span> <span class="n">unpack_hook</span><span class="p">):</span>
<span class="linenos" data-linenos="19 "></span>            <span class="c1"># 计算输出</span>
<span class="linenos" data-linenos="20 "></span>            <span class="n">output</span> <span class="o">=</span> <span class="n">x</span>
<span class="linenos" data-linenos="21 "></span>        <span class="k">return</span> <span class="n">output</span>
<span class="linenos" data-linenos="22 "></span>
<span class="linenos" data-linenos="23 "></span><span class="n">model</span> <span class="o">=</span> <span class="n">Model</span><span class="p">()</span>
<span class="linenos" data-linenos="24 "></span><span class="n">net</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">DataParallel</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
</code></pre></div>
<p>使用该上下文管理器定义的钩子是线程局部的。因此，以下代码不会产生预期的效果，因为钩子无法通过 <code>DataParallel</code>：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># 示例：不要这样做</span>
<span class="linenos" data-linenos="2 "></span>
<span class="linenos" data-linenos="3 "></span><span class="n">net</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">DataParallel</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span><span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">autograd</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">saved_tensors_hooks</span><span class="p">(</span><span class="n">pack_hook</span><span class="p">,</span> <span class="n">unpack_hook</span><span class="p">):</span>
<span class="linenos" data-linenos="5 "></span>    <span class="n">output</span> <span class="o">=</span> <span class="n">net</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
</code></pre></div>
<p>请注意，使用这些钩子会禁用为减少 Tensor 对象创建而进行的优化。例如：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">autograd</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">saved_tensors_hooks</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">):</span>
<span class="linenos" data-linenos="2 "></span>    <span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span>    <span class="n">y</span> <span class="o">=</span> <span class="n">x</span> <span class="o">*</span> <span class="n">x</span>
</code></pre></div>
<p>在没有钩子的情况下，<code>x</code>、<code>y.grad_fn._saved_self</code> 和 <code>y.grad_fn._saved_other</code> 都引用同一个张量对象。但有钩子时，PyTorch 会将 <code>x</code> 打包和解包成两个新的张量对象，它们与原始的 <code>x</code> 共享相同的存储（不会执行复制操作）。</p>
<h3 id="_16">反向传播钩子的执行<a class="headerlink" href="#_16" title="Permanent link">&para;</a></h3>
<p>本节将讨论何时触发不同的钩子，何时不会触发，然后讨论触发顺序。涵盖的钩子有：通过 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html" title="torch.Tensor.register_hook"><code>torch.Tensor.register_hook()</code></a> 注册到张量的反向钩子，通过 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.register_post_accumulate_grad_hook.html" title="torch.Tensor.register_post_accumulate_grad_hook"><code>torch.Tensor.register_post_accumulate_grad_hook()</code></a> 注册到张量的梯度累积后钩子，通过 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.graph.Node.register_hook.html" title="torch.autograd.graph.Node.register_hook"><code>torch.autograd.graph.Node.register_hook()</code></a> 注册到节点的后钩子，以及通过 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.graph.Node.register_prehook.html" title="torch.autograd.graph.Node.register_prehook"><code>torch.autograd.graph.Node.register_prehook()</code></a> 注册到节点的前钩子。</p>
<h3 id="_17">钩子是否会触发<a class="headerlink" href="#_17" title="Permanent link">&para;</a></h3>
<p>通过 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html" title="torch.Tensor.register_hook"><code>torch.Tensor.register_hook()</code></a> 注册到张量的钩子，会在为该张量计算梯度时执行（注意，这不要求执行该张量的 <code>grad_fn</code>。例如，如果该张量作为 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.grad.html" title="torch.autograd.grad"><code>torch.autograd.grad()</code></a> 的 <code>inputs</code> 参数的一部分传递，可能不会执行张量的 <code>grad_fn</code>，但该张量的钩子始终会被执行）。</p>
<p>通过 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.register_post_accumulate_grad_hook.html" title="torch.Tensor.register_post_accumulate_grad_hook"><code>torch.Tensor.register_post_accumulate_grad_hook()</code></a> 注册到张量的钩子，在该张量的梯度累积之后执行，这意味着该张量的 <code>grad</code> 字段已经设置。通过 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html" title="torch.Tensor.register_hook"><code>torch.Tensor.register_hook()</code></a> 注册的钩子在计算梯度时执行，而通过 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.register_post_accumulate_grad_hook.html" title="torch.Tensor.register_post_accumulate_grad_hook"><code>torch.Tensor.register_post_accumulate_grad_hook()</code></a> 注册的钩子只会在反向传播结束时，autograd 更新张量的 <code>grad</code> 字段后触发。因此，梯度累积后钩子只能为叶张量注册。如果尝试在非叶张量上注册，会抛出错误，即使 <code>backward(retain_graph=True)</code> 被调用也是如此。</p>
<p>通过 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.graph.Node.register_hook.html" title="torch.autograd.graph.Node.register_hook"><code>torch.autograd.graph.Node.register_hook()</code></a> 或 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.graph.Node.register_prehook.html" title="torch.autograd.graph.Node.register_prehook"><code>torch.autograd.graph.Node.register_prehook()</code></a> 注册到 <code>torch.autograd.graph.Node</code> 的钩子，只有在节点执行时才会触发。</p>
<p>特定节点是否执行取决于反向传播是否通过 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.grad.html" title="torch.autograd.grad"><code>torch.autograd.grad()</code></a> 或 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.backward.html" title="torch.autograd.backward"><code>torch.autograd.backward()</code></a> 调用。特别是，如果您在对应于传递给 <code>inputs</code> 参数的张量的节点上注册钩子，应了解这两者之间的差异。</p>
<p>如果使用 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.backward.html" title="torch.autograd.backward"><code>torch.autograd.backward()</code></a>，上述所有钩子都会执行，无论是否指定了 <code>inputs</code> 参数。这是因为 <code>.backward()</code> 会执行所有节点，即使它们对应于作为输入指定的张量。（注意，执行额外的节点通常是不必要的，但仍会这样做。这种行为可能会发生变化，因此不应依赖它。）</p>
<p>另一方面，如果使用 <a href="https://pytorch.org/docs/stable/generated/torch.autograd.grad.html" title="torch.autograd.grad"><code>torch.autograd.grad()</code></a>，传递给 <code>input</code> 的张量所对应节点上的反向钩子可能不会执行，因为除非有另一个输入依赖于该节点的梯度结果，否则不会执行这些节点。</p>
<h3 id="_18">钩子触发顺序<a class="headerlink" href="#_18" title="Permanent link">&para;</a></h3>
<p>执行顺序如下：</p>
<ol>
<li>钩子注册到张量时执行</li>
<li>钩子注册到节点的前钩子时执行（如果节点执行）</li>
<li>为保留梯度的张量更新 <code>.grad</code> 字段</li>
<li>节点执行（受上述规则约束）</li>
<li>梯度累积的叶张量执行梯度累积后钩子</li>
<li>钩子注册到节点的后钩子时执行（如果节点执行）</li>
</ol>
<p>如果同类型的多个钩子注册到同一个张量或节点，将按注册顺序执行。后执行的钩子可以看到之前钩子对梯度的修改。</p>
<h3 id="_19">特殊钩子<a class="headerlink" href="#_19" title="Permanent link">&para;</a></h3>
<p><a href="https://pytorch.org/docs/stable/autograd.html#torch.autograd.graph.register_multi_grad_hook" title="torch.autograd.graph.register_multi_grad_hook"><code>torch.autograd.graph.register_multi_grad_hook()</code></a> 是使用注册到张量的钩子实现的。每个单独的张量钩子按照上述顺序执行，已注册的多重梯度钩子会在计算最后一个张量的梯度时触发。</p>
<p><a href="https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_full_backward_hook.html" title="torch.nn.modules.module.register_module_full_backward_hook"><code>torch.nn.modules.module.register_module_full_backward_hook()</code></a> 是使用注册到节点的钩子实现的。计算正向传播时，钩子会注册到模块的输入和输出的 <code>grad_fn</code>。由于模块可能有多个输入并返回多个输出，因此自定义的 <code>autograd</code> 函数首先应用于模块的输入，再应用于模块的输出，以确保这些张量共享一个 <code>grad_fn</code>，从而可以将钩子附加到其中。</p>
<h3 id="_20">修改张量时的钩子行为<a class="headerlink" href="#_20" title="Permanent link">&para;</a></h3>
<p>通常，注册到张量的钩子会接收相对于该张量的输出梯度，此时张量的值是执行反向传播时的值。</p>
<p>但是，如果您在注册钩子后修改张量，修改前注册的钩子也会接收相对于该张量的输出梯度，但此时张量的值是修改前的值。</p>
<p>如果更喜欢前一种情况的行为，应在修改张量后再注册钩子。例如：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">t</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="mf">1.</span><span class="p">,</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">sin</span><span class="p">()</span>
<span class="linenos" data-linenos="2 "></span><span class="n">t</span><span class="o">.</span><span class="n">cos_</span><span class="p">()</span>
<span class="linenos" data-linenos="3 "></span><span class="n">t</span><span class="o">.</span><span class="n">register_hook</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span><span class="n">t</span><span class="o">.</span><span class="n">backward</span><span class="p">()</span>
</code></pre></div>
<p>此外，需要知道注册到张量的钩子实际上永久绑定到该张量的 <code>grad_fn</code> 上。因此，如果修改张量，即使该张量现在有一个新的 <code>grad_fn</code>，修改前注册的钩子仍会与旧的 <code>grad_fn</code> 关联。例如，当 autograd 引擎在计算图中达到该张量的旧 <code>grad_fn</code> 时，它们将会触发。</p>
<h3 id="tensorregister_hookhook"><code>Tensor.register_hook(hook)</code><a class="headerlink" href="#tensorregister_hookhook" title="Permanent link">&para;</a></h3>
<p>注册一个反向传播钩子。</p>
<p>每次计算该张量的梯度时都会调用该钩子。钩子的签名应如下：</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">hook</span><span class="p">(</span><span class="n">grad</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tensor</span> <span class="ow">or</span> <span class="kc">None</span>
</code></pre></div>
<p>钩子不应修改其参数，但可以选择性地返回一个新的梯度，用以替换原来的 <a href="https://pytorch.org/docs/stable/generated/torch.Tensor.grad.html" title="torch.Tensor.grad"><code>grad</code></a>。</p>
<p>该函数返回一个句柄，具有 <code>handle.remove()</code> 方法，可以用来将钩子从模块中移除。</p>
<p><strong>注意</strong></p>
<p>有关此钩子何时执行以及其与其他钩子的执行顺序，请参阅<a href="https://pytorch.org/docs/stable/notes/autograd.html#backward-hooks-execution">反向传播钩子的执行</a>。</p>
<p><strong>示例：</strong></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">v</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">([</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">],</span> <span class="n">requires_grad</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">h</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">register_hook</span><span class="p">(</span><span class="k">lambda</span> <span class="n">grad</span><span class="p">:</span> <span class="n">grad</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>  <span class="c1"># 将梯度翻倍</span>
<span class="linenos" data-linenos="3 "></span><span class="n">v</span><span class="o">.</span><span class="n">backward</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">([</span><span class="mf">1.</span><span class="p">,</span> <span class="mf">2.</span><span class="p">,</span> <span class="mf">3.</span><span class="p">]))</span>
<span class="linenos" data-linenos="4 "></span><span class="n">v</span><span class="o">.</span><span class="n">grad</span>
<span class="linenos" data-linenos="5 "></span><span class="c1"># 输出：</span>
<span class="linenos" data-linenos="6 "></span><span class="c1"># tensor([2., 4., 6.])</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">h</span><span class="o">.</span><span class="n">remove</span><span class="p">()</span>  <span class="c1"># 移除钩子</span>
</code></pre></div>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../Attention%20Generation%20Work%20Flow/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Attention Generation Work Flow">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Attention Generation Work Flow
              </div>
            </div>
          </a>
        
        
          
          <a href="../PyTorch%20Autograd%20Explained%20-%20In-depth%20Tutorial/" class="md-footer__link md-footer__link--next" aria-label="Next: PyTorch Autograd Explained   In depth Tutorial">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                PyTorch Autograd Explained   In depth Tutorial
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.code.select", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.instant.prefetch", "navigation.instant.progress", "navigation.path", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "header.autohide"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.525ec568.min.js"></script>
      
        <script src="../mathjax-config.js"></script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../extra.js"></script>
      
        <script src="../js/toggle_sidebar.js"></script>
      
    
  </body>
</html>